From 0052ca84e7ee88adf1cbfe84f4b7fcbcb75266be Mon Sep 17 00:00:00 2001
From: Ross Lagerwall <ross.lagerwall@citrix.com>
Date: Fri, 6 May 2016 15:02:45 +0100
Subject: [PATCH 1/5] lib: Add a generic implementation of current_text_addr()

Remove the unused x86 implementation.

Signed-off-by: Ross Lagerwall <ross.lagerwall@citrix.com>
diff --git a/xen/include/xen/lib.h b/xen/include/xen/lib.h
index 394319c81863..ced4783e1cfa 100644
--- a/xen/include/xen/lib.h
+++ b/xen/include/xen/lib.h
@@ -158,6 +158,15 @@ void cf_check dump_execstate(const struct cpu_user_regs *regs);
 
 void init_constructors(void);
 
+static inline void *current_text_addr(void)
+{
+    void *ip;
+
+    asm volatile ("lea 0(%%rip), %0" : "=r" (ip));
+
+    return ip;
+}
+
 /*
  * bsearch - binary search an array of elements
  * @key: pointer to item being searched for
