From d1d9ca85c856edd84ed57eb3f0e38d40d162c91f Mon Sep 17 00:00:00 2001
From: Sergey Dyasli <sergey.dyasli@citrix.com>
Date: Tue, 9 Apr 2019 15:21:12 +0100
Subject: [PATCH] microcode: remove panic() calls

Signed-off-by: Sergey Dyasli <sergey.dyasli@citrix.com>
diff --git a/xen/arch/x86/microcode.c b/xen/arch/x86/microcode.c
index e7b7ba14c8..3e328bd612 100644
--- a/xen/arch/x86/microcode.c
+++ b/xen/arch/x86/microcode.c
@@ -398,8 +398,11 @@ static int primary_thread_fn(const struct microcode_patch *patch)
         done = atomic_read(&cpu_out);
         if ( wait_for_condition(wait_cpu_callout, (done + 1),
                                 MICROCODE_UPDATE_TIMEOUT_US) )
-            panic("Timeout during sequential microcode update (finished %d/%d)",
-                  done, nr_cores);
+        {
+            printk("Timeout during sequential microcode update (finished %d/%d)",
+                   done, nr_cores);
+            return -EBUSY;
+        }
     }
 
     ret = microcode_ops->apply_microcode(patch);
@@ -477,8 +480,11 @@ static int control_thread_fn(const struct microcode_patch *patch)
          */
         if ( wait_for_condition(wait_cpu_callout, (done + 1),
                                 MICROCODE_UPDATE_TIMEOUT_US) )
-            panic("Timeout when finished updating microcode (finished %u/%u)",
-                  done, nr_cores);
+        {
+            printk("Timeout when finished updating microcode (finished %u/%u)",
+                   done, nr_cores);
+            return -EBUSY;
+        }
 
         /* Print warning message once if long time is spent here */
         if ( tick && rdtsc_ordered() - tick >= cpu_khz * 1000 )
