From 968c822e91a5ca1b42fe49fb58eedc240c6734e6 Mon Sep 17 00:00:00 2001
From: Andrew Cooper <andrew.cooper3@citrix.com>
Date: Thu, 31 Oct 2019 20:27:11 +0000
Subject: x86/msr: Expose cpu_has_tsx_ctrl via MSR_ARCH_CAPS

Signed-off-by: Andrew Cooper <andrew.cooper3@citrix.com>

diff --git a/xen/lib/x86/msr.c b/xen/lib/x86/msr.c
index 7d71e92a380a..058c7514f2ad 100644
--- a/xen/lib/x86/msr.c
+++ b/xen/lib/x86/msr.c
@@ -39,7 +39,17 @@ int x86_msr_copy_to_buffer(const struct msr_policy *p,
     })
 
     COPY_MSR(MSR_INTEL_PLATFORM_INFO, p->platform_info.raw);
-    COPY_MSR(MSR_ARCH_CAPABILITIES,   p->arch_caps.raw);
+
+#ifdef __XEN__
+    /*
+     * When requesting the Host MSR policy, bodge cpu_has_{hle,rtm} as
+     * TSX_CTRL too, to trigger xenopsd's compatiblity logic on all systems.
+     */
+    if ( p == &host_msr_policy && (cpu_has_hle || cpu_has_rtm) )
+        COPY_MSR(MSR_ARCH_CAPABILITIES, ARCH_CAPS_TSX_CTRL | p->arch_caps.raw);
+    else
+#endif
+        COPY_MSR(MSR_ARCH_CAPABILITIES,   p->arch_caps.raw);
 
 #undef COPY_MSR
 
