diff --git a/tools/config.h.in b/tools/config.h.in
index 9d839f8af1e0..043b680004ef 100644
--- a/tools/config.h.in
+++ b/tools/config.h.in
@@ -93,6 +93,9 @@
 /* Define extfs header to use */
 #undef INCLUDE_EXTFS_H
 
+/* Header path for INCLUDE_JSON_H */
+#undef INCLUDE_JSON_H
+
 /* IPXE path */
 #undef IPXE_PATH
 
diff --git a/tools/configure b/tools/configure
index 402364f4c4e3..1503abcfd23f 100755
--- a/tools/configure
+++ b/tools/configure
@@ -9711,6 +9711,45 @@ fi
 done
 
 
+ax_found=0
+
+    if test "$ax_found" = "0"; then :
+
+        ac_fn_c_check_header_mongrel "$LINENO" "json/json.h" "ac_cv_header_json_json_h" "$ac_includes_default"
+if test "x$ac_cv_header_json_json_h" = xyes; then :
+
+
+$as_echo "#define INCLUDE_JSON_H <json/json.h>" >>confdefs.h
+
+            ax_found=1
+fi
+
+
+
+fi
+
+    if test "$ax_found" = "0"; then :
+
+        ac_fn_c_check_header_mongrel "$LINENO" "json-c/json.h" "ac_cv_header_json_c_json_h" "$ac_includes_default"
+if test "x$ac_cv_header_json_c_json_h" = xyes; then :
+
+
+$as_echo "#define INCLUDE_JSON_H <json-c/json.h>" >>confdefs.h
+
+            ax_found=1
+fi
+
+
+
+fi
+
+if test "$ax_found" = "0"; then :
+
+    as_fn_error $? "No header found from list json/json.h json-c/json.h" "$LINENO" 5
+
+fi
+
+
 cat >confcache <<\_ACEOF
 # This file is a shell script that caches the results of configure
 # tests run on this system so they can be shared between configure
diff --git a/tools/configure.ac b/tools/configure.ac
index c1ea1e32ca58..ae2d37f87ad0 100644
--- a/tools/configure.ac
+++ b/tools/configure.ac
@@ -517,6 +517,7 @@ AS_IF([test "x$pvshim" = "xy"], [
 ])
 
 AX_FIND_HEADER([INCLUDE_ENDIAN_H], [endian.h sys/endian.h])
+AX_FIND_HEADER([INCLUDE_JSON_H], [json/json.h json-c/json.h])
 
 AC_CHECK_FUNCS([pipe2])
 
