diff --git a/tools/config.h.in b/tools/config.h.in
index 3071cb3998ce..39a115b6afef 100644
--- a/tools/config.h.in
+++ b/tools/config.h.in
@@ -90,6 +90,9 @@
 /* Define extfs header to use */
 #undef INCLUDE_EXTFS_H
 
+/* Header path for INCLUDE_JSON_H */
+#undef INCLUDE_JSON_H
+
 /* IPXE path */
 #undef IPXE_PATH
 
diff --git a/tools/configure b/tools/configure
index 52b4717d0119..5249e53de7af 100755
--- a/tools/configure
+++ b/tools/configure
@@ -10155,6 +10155,45 @@ if test "$ax_found" = "0"; then :
 fi
 
 
+ax_found=0
+
+    if test "$ax_found" = "0"; then :
+
+        ac_fn_c_check_header_mongrel "$LINENO" "json/json.h" "ac_cv_header_json_json_h" "$ac_includes_default"
+if test "x$ac_cv_header_json_json_h" = xyes; then :
+
+
+$as_echo "#define INCLUDE_JSON_H <json/json.h>" >>confdefs.h
+
+            ax_found=1
+fi
+
+
+
+fi
+
+    if test "$ax_found" = "0"; then :
+
+        ac_fn_c_check_header_mongrel "$LINENO" "json-c/json.h" "ac_cv_header_json_c_json_h" "$ac_includes_default"
+if test "x$ac_cv_header_json_c_json_h" = xyes; then :
+
+
+$as_echo "#define INCLUDE_JSON_H <json-c/json.h>" >>confdefs.h
+
+            ax_found=1
+fi
+
+
+
+fi
+
+if test "$ax_found" = "0"; then :
+
+    as_fn_error $? "No header found from list json/json.h json-c/json.h" "$LINENO" 5
+
+fi
+
+
 cat >confcache <<\_ACEOF
 # This file is a shell script that caches the results of configure
 # tests run on this system so they can be shared between configure
diff --git a/tools/configure.ac b/tools/configure.ac
index 3cccf4196085..729c5f8ddf68 100644
--- a/tools/configure.ac
+++ b/tools/configure.ac
@@ -516,5 +516,6 @@ AS_IF([test "x$pvshim" = "xy"], [
 ])
 
 AX_FIND_HEADER([INCLUDE_ENDIAN_H], [endian.h sys/endian.h])
+AX_FIND_HEADER([INCLUDE_JSON_H], [json/json.h json-c/json.h])
 
 AC_OUTPUT()
