diff --git a/tools/config.h.in b/tools/config.h.in
index 0bab3cb136e9..b03cbf7bbb54 100644
--- a/tools/config.h.in
+++ b/tools/config.h.in
@@ -96,6 +96,9 @@
 /* Define extfs header to use */
 #undef INCLUDE_EXTFS_H
 
+/* Header path for INCLUDE_JSON_H */
+#undef INCLUDE_JSON_H
+
 /* IPXE path */
 #undef IPXE_PATH
 
diff --git a/tools/configure b/tools/configure
index d160ca296202..2f3abc918caf 100755
--- a/tools/configure
+++ b/tools/configure
@@ -10464,6 +10464,48 @@ then :
 fi
 
 
+ax_found=0
+
+    if test "$ax_found" = "0"
+then :
+
+        ac_fn_c_check_header_compile "$LINENO" "json/json.h" "ac_cv_header_json_json_h" "$ac_includes_default"
+if test "x$ac_cv_header_json_json_h" = xyes
+then :
+
+
+printf "%s\n" "#define INCLUDE_JSON_H <json/json.h>" >>confdefs.h
+
+            ax_found=1
+fi
+
+
+fi
+
+    if test "$ax_found" = "0"
+then :
+
+        ac_fn_c_check_header_compile "$LINENO" "json-c/json.h" "ac_cv_header_json_c_json_h" "$ac_includes_default"
+if test "x$ac_cv_header_json_c_json_h" = xyes
+then :
+
+
+printf "%s\n" "#define INCLUDE_JSON_H <json-c/json.h>" >>confdefs.h
+
+            ax_found=1
+fi
+
+
+fi
+
+if test "$ax_found" = "0"
+then :
+
+    as_fn_error $? "No header found from list json/json.h json-c/json.h" "$LINENO" 5
+
+fi
+
+
 ac_fn_c_check_func "$LINENO" "pipe2" "ac_cv_func_pipe2"
 if test "x$ac_cv_func_pipe2" = xyes
 then :
diff --git a/tools/configure.ac b/tools/configure.ac
index be58f06be450..414df40f3346 100644
--- a/tools/configure.ac
+++ b/tools/configure.ac
@@ -542,6 +542,7 @@ AS_IF([test "x$pvshim" = "xy"], [
 ])
 
 AX_FIND_HEADER([INCLUDE_ENDIAN_H], [endian.h sys/endian.h])
+AX_FIND_HEADER([INCLUDE_JSON_H], [json/json.h json-c/json.h])
 
 AC_CHECK_FUNCS([pipe2])
 
