From 8b1ac353b4db7c5bb2f82cb6afee9cc641e756a4 Mon Sep 17 00:00:00 2001
From: Yann Dirson <yann.dirson@vates.fr>
Date: Tue, 9 May 2023 09:01:36 +0000
Subject: docs/man: fix xenstore-write synopsis

Reported-by: zithro <slack@rabbit.lu>
Signed-off-by: Yann Dirson <yann.dirson@vates.fr>
Reviewed-by: Andrew Cooper <andrew.cooper3@citrix.com>

diff --git a/docs/man/xenstore-write.1.pod b/docs/man/xenstore-write.1.pod
index d1b011236a57..c4e7f9181073 100644
--- a/docs/man/xenstore-write.1.pod
+++ b/docs/man/xenstore-write.1.pod
@@ -4,7 +4,7 @@ xenstore-write - write Xenstore values
 
 =head1 SYNOPSIS
 
-B<xenstore-read> [I<OPTION>]... I<PATH> I<VALUE>...
+B<xenstore-write> [I<OPTION>]... I<PATH> I<VALUE>...
 
 =head1 DESCRIPTION
 
