From 85d7fe2e14e562e9a006d19f5e165bc870639f0f Mon Sep 17 00:00:00 2001
From: Gerald Elder-Vass <gerald.elder-vass@cloud.com>
Date: Thu, 25 Sep 2025 16:01:36 +0200
Subject: efi: Fix line length in init_secure_boot_mode
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Commit cb41b4ce14a9 introduced init_secure_boot_mode but one line was
not wrapped appropriately.

Signed-off-by: Gerald Elder-Vass <gerald.elder-vass@cloud.com>
Reviewed-by: Alejandro Vallejo <alejandro.garciavallejo@amd.com>
Acked-by: Marek Marczykowski-Górecki <marmarek@invisiblethingslab.com>

diff --git a/xen/common/efi/boot.c b/xen/common/efi/boot.c
index 7615e47b8ff7..bfb55146217d 100644
--- a/xen/common/efi/boot.c
+++ b/xen/common/efi/boot.c
@@ -919,7 +919,8 @@ static void __init init_secure_boot_mode(void)
 
     if ( status == EFI_NOT_FOUND ||
          (status == EFI_SUCCESS &&
-          attr == (EFI_VARIABLE_BOOTSERVICE_ACCESS | EFI_VARIABLE_RUNTIME_ACCESS) &&
+          attr == (EFI_VARIABLE_BOOTSERVICE_ACCESS |
+                   EFI_VARIABLE_RUNTIME_ACCESS) &&
           size == 1 && data == 0) )
         /* Platform does not support Secure Boot or it's disabled. */
         efi_secure_boot = false;
