diff --git a/xen/arch/x86/cpu-policy.c b/xen/arch/x86/cpu-policy.c
index 22a1940fde19..b92e47efc4cc 100644
--- a/xen/arch/x86/cpu-policy.c
+++ b/xen/arch/x86/cpu-policy.c
@@ -454,10 +454,23 @@ static void __init guest_common_max_feature_adjustments(uint32_t *fs)
              raw_cpu_policy.feat.clwb )
             __set_bit(X86_FEATURE_CLWB, fs);
     }
+
+    /*
+     * This is AMD-specific, but past mistakes made the bit surface on
+     * existing non-AMD/Hygon domains. The side effects are benign, as the
+     * affected CPUID processor enumeration fields are reserved and unused
+     * on other vendors, so we can just leave it in the max policy.
+     */
+     __set_bit(X86_FEATURE_CMP_LEGACY, fs);
 }
 
 static void __init guest_common_default_feature_adjustments(uint32_t *fs)
 {
+    /* CMP_LEGACY is AMD/Hygon-only. Strip it from CPUs that don't have it */
+    __clear_bit(X86_FEATURE_CMP_LEGACY, fs);
+    if ( cpu_has_cmp_legacy )
+        __set_bit(X86_FEATURE_CMP_LEGACY, fs);
+
     if ( boot_cpu_data.x86_vendor == X86_VENDOR_INTEL )
     {
         /*
@@ -824,12 +837,11 @@ void recalculate_cpuid_policy(struct domain *d)
     }
 
     /*
-     * Allow the toolstack to set HTT, X2APIC and CMP_LEGACY.  These bits
+     * Allow the toolstack to set HTT and X2APIC.  These bits
      * affect how to interpret topology information in other cpuid leaves.
      */
     __set_bit(X86_FEATURE_HTT, max_fs);
     __set_bit(X86_FEATURE_X2APIC, max_fs);
-    __set_bit(X86_FEATURE_CMP_LEGACY, max_fs);
 
     /*
      * 32bit PV domains can't use any Long Mode features, and cannot use
