diff --git a/tools/pygrub/src/pygrub b/tools/pygrub/src/pygrub
index 9cc06859e32a..d61cb3834639 100755
--- a/tools/pygrub/src/pygrub
+++ b/tools/pygrub/src/pygrub
@@ -1105,6 +1105,11 @@ if __name__ == "__main__":
     if list_entries:
         sys.exit(0)
 
+    # XenMobile controller has a broken bootloader config, fix it up
+    if chosencfg["kernel"] and not fs.file_exists(chosencfg["kernel"]) \
+            and fs.file_exists("/kernel-xenserver"):
+        chosencfg["kernel"] = "/kernel-xenserver"
+
     # Did looping through partitions find us a kernel?
     if fs is None:
         raise RuntimeError("Unable to find partition containing kernel")
