From 5921e50d79cd0ec2b20d74ab90fb0a425fa8f8e4 Mon Sep 17 00:00:00 2001
From: Marcus Granado <marcus.granado@citrix.com>
Date: Thu, 4 Dec 2025 13:20:00 +0000
Subject: Add XEN_DOMCTL_numa_op and bindings
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Marcus Granado <marcus.granado@citrix.com>
Signed-off-by: Bernhard Kaindl <bernhard.kaindl@citrix.com>
Signed-off-by: Roger Pau Monné <roger.pau@citrix.com>

diff --git a/tools/flask/policy/modules/dom0.te b/tools/flask/policy/modules/dom0.te
index a16e847da1cc..c3ca28db4853 100644
--- a/tools/flask/policy/modules/dom0.te
+++ b/tools/flask/policy/modules/dom0.te
@@ -98,6 +98,7 @@ allow dom0_t dom0_t:domain2 {
 	gettsc
 	settsc
 	get_runstate_info
+	numa_op
 	setscheduler
 	set_vnumainfo
 	get_vnumainfo
diff --git a/tools/flask/policy/modules/xen.if b/tools/flask/policy/modules/xen.if
index 997d55260a87..ce32666acc2d 100644
--- a/tools/flask/policy/modules/xen.if
+++ b/tools/flask/policy/modules/xen.if
@@ -88,6 +88,7 @@ define(`create_domain_common', `
 		setscheduler
 		setclaim
 		get_runstate_info
+		numa_op
 		set_vnumainfo
 		get_vnumainfo
 		cacheflush
diff --git a/tools/include/xenctrl.h b/tools/include/xenctrl.h
index fb789296035a..0d2885f6782f 100644
--- a/tools/include/xenctrl.h
+++ b/tools/include/xenctrl.h
@@ -1203,6 +1203,11 @@ int xc_domain_getvnuma(xc_interface *xch,
                        unsigned int *vdistance,
                        unsigned int *vcpu_to_vnode);
 
+int xc_domain_numa_get_node_pages(xc_interface *xch,
+                                  uint32_t domid,
+                                  uint32_t *nr_nodes,
+                                  uint64_t *tot_pages_per_node);
+
 int xc_domain_soft_reset(xc_interface *xch,
                          uint32_t domid);
 
diff --git a/tools/libs/ctrl/xc_domain.c b/tools/libs/ctrl/xc_domain.c
index 763bc9cca50d..5cdd23202f60 100644
--- a/tools/libs/ctrl/xc_domain.c
+++ b/tools/libs/ctrl/xc_domain.c
@@ -2263,6 +2263,53 @@ int xc_domain_getvnuma(xc_interface *xch,
     return rc;
 }
 
+int xc_domain_numa_get_node_pages(xc_interface *xch,
+                                  uint32_t domid,
+                                  uint32_t *nr_nodes,
+                                  uint64_t *node_tot_pages)
+{
+    struct xen_domctl domctl = {
+        .cmd = XEN_DOMCTL_numa_op,
+        .domain = domid,
+        .u.numa_op = {
+            .op = XEN_DOMCTL_NUMA_OP_GET_NODE_PAGES,
+            .u.node_pages.nr_nodes = *nr_nodes,
+        },
+    };
+    DECLARE_HYPERCALL_BOUNCE(node_tot_pages,
+                             *nr_nodes * sizeof(*node_tot_pages),
+                             XC_HYPERCALL_BUFFER_BOUNCE_OUT);
+    int rc = -1;
+    bool fetch_size = !*nr_nodes;
+
+    if ( fetch_size && node_tot_pages )
+    {
+        errno = EINVAL;
+        return -1;
+    }
+
+    if ( xc_hypercall_bounce_pre(xch, node_tot_pages) )
+        return rc;
+
+    set_xen_guest_handle(
+        domctl.u.numa_op.u.node_pages.node_tot_pages,
+        node_tot_pages);
+
+    rc = do_domctl(xch, &domctl);
+    if ( rc )
+        return rc;
+
+    *nr_nodes = domctl.u.numa_op.u.node_pages.nr_nodes;
+    if ( fetch_size )
+        return 0;
+
+    HYPERCALL_BOUNCE_SET_SIZE(node_tot_pages,
+                              *nr_nodes * sizeof(*node_tot_pages));
+    xc_hypercall_bounce_post(xch, node_tot_pages);
+
+    return rc;
+}
+
 int xc_domain_soft_reset(xc_interface *xch,
                          uint32_t domid)
 {
diff --git a/tools/ocaml/libs/xc/xenctrl.ml b/tools/ocaml/libs/xc/xenctrl.ml
index f92e9d552e3d..3760b16b14c8 100644
--- a/tools/ocaml/libs/xc/xenctrl.ml
+++ b/tools/ocaml/libs/xc/xenctrl.ml
@@ -291,6 +291,12 @@ external domain_getinfo: handle -> domid -> domaininfo= "stub_xc_domain_getinfo"
 external domain_get_vcpuinfo: handle -> int -> int -> vcpuinfo
   = "stub_xc_vcpu_getinfo"
 
+type domain_numainfo_node_pages = {
+  tot_pages_per_node : int64 array;
+}
+external domain_get_numa_info_node_pages : handle -> int -> domain_numainfo_node_pages
+  = "stub_xc_domain_numa_get_node_pages"
+
 external domain_ioport_permission: handle -> domid -> int -> int -> bool -> unit
   = "stub_xc_domain_ioport_permission"
 external domain_iomem_permission: handle -> domid -> nativeint -> nativeint -> bool -> unit
diff --git a/tools/ocaml/libs/xc/xenctrl.mli b/tools/ocaml/libs/xc/xenctrl.mli
index 92bdd3e8d958..85b4d7d15163 100644
--- a/tools/ocaml/libs/xc/xenctrl.mli
+++ b/tools/ocaml/libs/xc/xenctrl.mli
@@ -209,6 +209,13 @@ end
 
 include module type of Runstateinfo.V1
 
+(* tot_pages_per_node are counted as 4k pages *)
+type domain_numainfo_node_pages = {
+  tot_pages_per_node : int64 array;
+}
+external domain_get_numa_info_node_pages : handle -> int -> domain_numainfo_node_pages
+  = "stub_xc_domain_numa_get_node_pages"
+
 external domain_sethandle : handle -> domid -> string -> unit = "stub_xc_domain_sethandle"
 external domain_max_vcpus : handle -> domid -> int -> unit
   = "stub_xc_domain_max_vcpus"
diff --git a/tools/ocaml/libs/xc/xenctrl_stubs.c b/tools/ocaml/libs/xc/xenctrl_stubs.c
index c2d68da52603..d82ca3ddaef5 100644
--- a/tools/ocaml/libs/xc/xenctrl_stubs.c
+++ b/tools/ocaml/libs/xc/xenctrl_stubs.c
@@ -641,6 +641,61 @@ static int get_cpumap_len(xc_interface *xch, value cpumap)
 		return xc_len;
 }
 
+/*
+ * Retrieve the total number of pages per NUMA node for a given domain.
+ *
+ * This function will return the total number of pages for each NUMA node
+ * associated with the specified domain ID. The information is retrieved
+ * from the Xen hypervisor and returned as an array of integers, where each
+ * element corresponds to a NUMA node.
+ */
+CAMLprim value stub_xc_domain_numa_get_node_pages(value xch_val, value domid)
+{
+	CAMLparam2(xch_val, domid);
+	CAMLlocal2(o_result, o_result_array);
+	xc_interface *xch = xch_of_val(xch_val);
+	uint64_t *c_node_tot_pages;
+	uint32_t c_nr_nodes = 0, c_domid = Int_val(domid);
+	unsigned int node;
+	int retval;
+
+	/* First call to get the number of nodes to be returned by the hypercall */
+	caml_enter_blocking_section();
+	retval = xc_domain_numa_get_node_pages(xch, c_domid, &c_nr_nodes, NULL);
+	caml_leave_blocking_section();
+
+	if (retval < 0)
+		failwith_xc(xch);
+
+	/* Now get the actual data from xc_domain_numa_get_node_pages() */
+	c_node_tot_pages = malloc(c_nr_nodes * sizeof(*c_node_tot_pages));
+	if (!c_node_tot_pages)
+		caml_raise_out_of_memory();
+
+	caml_enter_blocking_section();
+	retval = xc_domain_numa_get_node_pages(xch, c_domid,
+					       &c_nr_nodes, c_node_tot_pages);
+	caml_leave_blocking_section();
+
+	/* c_nr_nodes has been reduced by the hypercall to the #nodes of the host */
+	if (retval < 0) {
+		free(c_node_tot_pages);
+		failwith_xc(xch);
+	}
+
+	o_result_array = caml_alloc(c_nr_nodes, 0);
+	for (node = 0; node < c_nr_nodes; node++)
+		Store_field(o_result_array, node,
+			    caml_copy_int64(c_node_tot_pages[node]));
+
+	free(c_node_tot_pages);
+
+	o_result = caml_alloc_tuple(1);
+	Store_field(o_result, 0, o_result_array);
+
+	CAMLreturn(o_result);
+}
+
 CAMLprim value stub_xc_vcpu_setaffinity(value xch_val, value domid,
                                         value vcpu, value cpumap)
 {
diff --git a/xen/common/domctl.c b/xen/common/domctl.c
index 156752c90c06..d12362850320 100644
--- a/xen/common/domctl.c
+++ b/xen/common/domctl.c
@@ -603,6 +603,10 @@ long do_domctl(XEN_GUEST_HANDLE_PARAM(xen_domctl_t) u_domctl)
         copyback = true;
         goto domctl_out_unlock_domonly;
 
+    case XEN_DOMCTL_numa_op:
+        ret = numa_domctl(d, &op->u.numa_op, &copyback);
+        goto domctl_out_unlock_domonly;
+
     default:
         /* Everything else handled further up or further down. */
         break;
diff --git a/xen/common/numa.c b/xen/common/numa.c
index 904ad766706b..19c1d09f722e 100644
--- a/xen/common/numa.c
+++ b/xen/common/numa.c
@@ -5,6 +5,7 @@
  */
 
 #include <xen/init.h>
+#include <xen/guest_access.h>
 #include <xen/keyhandler.h>
 #include <xen/mm.h>
 #include <xen/nodemask.h>
@@ -824,3 +825,69 @@ static int __init cf_check register_numa_trigger(void)
     return 0;
 }
 __initcall(register_numa_trigger);
+
+/**
+ * @brief Copy d->node_tot_pages[<node>] to guest memory.
+ * @param d The domain whose NUMA info is being retrieved.
+ * @param nodes The data structure to fill with total pages per node.
+ * @return 0 on success, negative error code on failure.
+ *
+ * This function copies the total number of pages per NUMA node
+ * from the host domain structure to the guest memory specified
+ * in the `node_pages` structure.
+ *
+ * On success, it updates `nr_nodes` to reflect the actual
+ * number of nodes copied. When the caller provided nr_nodes as 0
+ * to query the number of nodes, it just sets `nr_nodes` and returns 0.
+ * Otherwise, if the provided `nr_nodes` is less than the number of online
+ * nodes, it returns -ERANGE.
+ */
+static int domain_get_node_tot_pages(struct domain *d, struct node_pages *nodes)
+{
+    const nodeid_t online_nodes = last_node(node_online_map) + 1;
+    nodeid_t node;
+    uint64_t node_tot_pages[MAX_NUMNODES];
+
+    if ( nodes->nr_nodes == 0 )
+    {
+        /* For callers who want to know the number of nodes to allocate for */
+        nodes->nr_nodes = online_nodes;
+        return 0;
+    }
+
+    if ( online_nodes > nodes->nr_nodes )
+        return -ERANGE;
+
+    nrspin_lock(&d->page_alloc_lock);
+    for ( node = 0; node < online_nodes; node++ )
+        node_tot_pages[node] = d->node_tot_pages[node];
+    nrspin_unlock(&d->page_alloc_lock);
+
+    if ( copy_to_guest(nodes->node_tot_pages, node_tot_pages, online_nodes) )
+    {
+        nodes->nr_nodes = 0;
+        return -EFAULT;
+    }
+    nodes->nr_nodes = online_nodes;
+    return 0;
+}
+
+/* Domain NUMA operations */
+int numa_domctl(
+    struct domain *d, struct xen_domctl_numa_op *numa, bool *copyback)
+{
+    int ret;
+
+    switch ( numa->op )
+    {
+    case XEN_DOMCTL_NUMA_OP_GET_NODE_PAGES:
+        ret = domain_get_node_tot_pages(d, &numa->u.node_pages);
+        *copyback = true;
+        break;
+
+    default:
+        ret = -EOPNOTSUPP;
+    }
+
+    return ret;
+}
diff --git a/xen/include/public/domctl.h b/xen/include/public/domctl.h
index 3ae7b9158684..fb759b3f09b7 100644
--- a/xen/include/public/domctl.h
+++ b/xen/include/public/domctl.h
@@ -1069,6 +1069,23 @@ struct xen_domctl_vcpu_msrs {
 };
 #endif
 
+/*
+ * Domain NUMA operations
+ * GET_NODE_PAGES: get the number of pages per NUMA node
+ */
+#define XEN_DOMCTL_NUMA_OP_GET_NODE_PAGES       0
+
+/* XEN_DOMCTL_numa_op */
+struct xen_domctl_numa_op {
+    uint32_t op; /* IN */
+    union {
+        struct node_pages {
+            uint32_t nr_nodes; /* IN/OUT: number of nodes for node_tot_pages */
+            XEN_GUEST_HANDLE_64(uint64) node_tot_pages; /* OUT */
+        } node_pages;
+    } u;
+};
+
 /* XEN_DOMCTL_setvnumainfo: specifies a virtual NUMA topology for the guest */
 struct xen_domctl_vnuma {
     /* IN: number of vNUMA nodes to setup. Shall be greater than 0 */
@@ -1337,6 +1354,7 @@ struct xen_domctl {
 #define XEN_DOMCTL_shadow_op                     10
 #define XEN_DOMCTL_max_mem                       11
 #define XEN_DOMCTL_get_runstate_info           1099
+#define XEN_DOMCTL_numa_op                     1101
 #define XEN_DOMCTL_setvcpucontext                12
 #define XEN_DOMCTL_getvcpucontext                13
 #define XEN_DOMCTL_getvcpuinfo                   14
@@ -1428,6 +1446,7 @@ struct xen_domctl {
         struct xen_domctl_getpageframeinfo3 getpageframeinfo3;
         struct xen_domctl_nodeaffinity      nodeaffinity;
         struct xen_domctl_runstate_info_ext domain_runstate;
+        struct xen_domctl_numa_op           numa_op;
         struct xen_domctl_vcpuaffinity      vcpuaffinity;
         struct xen_domctl_shadow_op         shadow_op;
         struct xen_domctl_max_mem           max_mem;
diff --git a/xen/include/xen/numa.h b/xen/include/xen/numa.h
index f6c1f27ca105..6737ade8640a 100644
--- a/xen/include/xen/numa.h
+++ b/xen/include/xen/numa.h
@@ -127,6 +127,8 @@ extern bool numa_memblks_available(void);
 extern bool numa_update_node_memblks(nodeid_t node, unsigned int arch_nid,
                                      paddr_t start, paddr_t size, bool hotplug);
 extern void numa_set_processor_nodes_parsed(nodeid_t node);
+extern int numa_domctl(
+    struct domain *d, struct xen_domctl_numa_op *numa, bool *copyback);
 
 #else
 
@@ -148,6 +150,12 @@ static inline nodeid_t mfn_to_nid(mfn_t mfn)
     return 0;
 }
 
+static inline int numa_domctl(
+    struct domain *d, struct xen_domctl_numa_op *numa, bool *copyback)
+{
+    return -EOPNOTSUPP;
+}
+
 #endif
 
 #define page_to_nid(pg) mfn_to_nid(page_to_mfn(pg))
diff --git a/xen/xsm/flask/hooks.c b/xen/xsm/flask/hooks.c
index 0c13667c228b..67ffe31d08ac 100644
--- a/xen/xsm/flask/hooks.c
+++ b/xen/xsm/flask/hooks.c
@@ -722,6 +722,9 @@ static int cf_check flask_domctl(struct domain *d, struct xen_domctl *op)
     case XEN_DOMCTL_get_runstate_info:
         return current_has_perm(d, SECCLASS_DOMAIN2, DOMAIN2__GET_RUNSTATE_INFO);
 
+    case XEN_DOMCTL_numa_op:
+        return current_has_perm(d, SECCLASS_DOMAIN2, DOMAIN2__NUMA_OP);
+
     case XEN_DOMCTL_setvcpuaffinity:
     case XEN_DOMCTL_setnodeaffinity:
         return current_has_perm(d, SECCLASS_DOMAIN, DOMAIN__SETAFFINITY);
diff --git a/xen/xsm/flask/policy/access_vectors b/xen/xsm/flask/policy/access_vectors
index b175fdb8d53d..7503f06386d0 100644
--- a/xen/xsm/flask/policy/access_vectors
+++ b/xen/xsm/flask/policy/access_vectors
@@ -221,6 +221,8 @@ class domain2
     settsc
 # XEN_DOMCTL_get_runstate_info
     get_runstate_info
+# XEN_DOMCTL_numa_op
+    numa_op
 # XEN_DOMCTL_scheduler_op with XEN_DOMCTL_SCHEDOP_putinfo
     setscheduler
 # XENMEM_claim_pages
