diff --git a/xen/arch/x86/spec_ctrl.c b/xen/arch/x86/spec_ctrl.c
index dcd9a78ef4f4..82a4fb359a92 100644
--- a/xen/arch/x86/spec_ctrl.c
+++ b/xen/arch/x86/spec_ctrl.c
@@ -1074,7 +1074,7 @@ static void __init ibpb_calculations(void)
     }
 
     if ( opt_ibpb_entry_pv == -1 )
-        opt_ibpb_entry_pv = IS_ENABLED(CONFIG_PV) && def_ibpb_entry;
+        opt_ibpb_entry_pv = IS_ENABLED(CONFIG_PV) && def_ibpb_entry && pv_shim;
     if ( opt_ibpb_entry_hvm == -1 )
         opt_ibpb_entry_hvm = IS_ENABLED(CONFIG_HVM) && def_ibpb_entry;
 
@@ -1584,7 +1584,7 @@ void spec_ctrl_init_domain(struct domain *d)
     bool pv = is_pv_domain(d);
 
     bool verw = ((pv ? opt_verw_pv : opt_verw_hvm) ||
-                 (opt_verw_mmio && is_iommu_enabled(d)));
+                 (opt_verw_mmio && is_iommu_enabled(d) && !is_hardware_domain(d)));
 
     bool ibpb = ((pv ? opt_ibpb_entry_pv : opt_ibpb_entry_hvm) &&
                  (d->domain_id != 0 || opt_ibpb_entry_dom0));
