diff --git a/tools/libxc/include/xenctrl.h b/tools/libxc/include/xenctrl.h
index 18e9675878..2be871123a 100644
--- a/tools/libxc/include/xenctrl.h
+++ b/tools/libxc/include/xenctrl.h
@@ -46,6 +46,7 @@
 #include <xen/xsm/flask_op.h>
 #include <xen/kexec.h>
 #include <xen/platform.h>
+#include <xen/pv-iommu.h>
 
 #include "xentoollog.h"
 
@@ -2448,6 +2449,8 @@ int xc_kexec_unload(xc_interface *xch, int type);
  */
 int xc_kexec_status(xc_interface *xch, int type);
 
+int xc_iommu_op(xc_interface *xch, struct pv_iommu_op *ops, unsigned int count);
+
 typedef xenpf_resource_entry_t xc_resource_entry_t;
 
 /*
diff --git a/tools/libxc/xc_misc.c b/tools/libxc/xc_misc.c
index b44f12e533..6e867728d4 100644
--- a/tools/libxc/xc_misc.c
+++ b/tools/libxc/xc_misc.c
@@ -919,6 +919,31 @@ int xc_livepatch_replace(xc_interface *xch, char *name, uint32_t timeout)
     return _xc_livepatch_action(xch, name, LIVEPATCH_ACTION_REPLACE, timeout);
 }
 
+int xc_iommu_op(xc_interface *xch, struct pv_iommu_op *ops, unsigned int count)
+{
+    DECLARE_HYPERCALL_BOUNCE(ops, count * sizeof(*ops),
+                             XC_HYPERCALL_BUFFER_BOUNCE_BOTH);
+    int ret;
+
+    if ( xc_hypercall_bounce_pre(xch, ops) )
+    {
+        PERROR("Could not bounce memmory for IOMMU hypercall");
+        return -1;
+    }
+
+    ret = xencall2(xch->xcall, __HYPERVISOR_iommu_op,
+                   HYPERCALL_BUFFER_AS_ARG(ops), count);
+
+    xc_hypercall_bounce_post(xch, ops);
+
+    if ( ret < 0 )
+    {
+        errno = -ret;
+        ret = -1;
+    }
+    return ret;
+}
+
 /*
  * Local variables:
  * mode: C
