From 8182d8d0d819ead81e70265a4ab79c9e74cdc43c Mon Sep 17 00:00:00 2001
From: Andrew Cooper <andrew.cooper3@citrix.com>
Date: Tue, 22 Jul 2014 14:09:53 +0100
Subject: [PATCH 1/3] tools/xenconsoled: Newline on 'Logfile Opened' messages

Signed-off-by: Andrew Cooper <andrew.cooper3@citrix.com>
CC: Ian Campbell <Ian.Campbell@citrix.com>
CC: Ian Jackson <Ian.Jackson@eu.citrix.com>
---
 tools/console/daemon/io.c |    8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/tools/console/daemon/io.c b/tools/console/daemon/io.c
index 46719a8..313b256 100644
--- a/tools/console/daemon/io.c
+++ b/tools/console/daemon/io.c
@@ -276,8 +276,8 @@ static int create_hv_log(void)
 		dolog(LOG_ERR, "Failed to open log %s: %d (%s)",
 		      logfile, errno, strerror(errno));
 	if (fd != -1 && log_time_hv) {
-		if (write_with_timestamp(fd, "Logfile Opened",
-					 strlen("Logfile Opened"),
+		if (write_with_timestamp(fd, "Logfile Opened\n",
+					 strlen("Logfile Opened\n"),
 					 &log_time_hv_needts) < 0) {
 			dolog(LOG_ERR, "Failed to log opening timestamp "
 				       "in %s: %d (%s)", logfile, errno,
@@ -322,8 +322,8 @@ static int create_domain_log(struct domain *dom)
 		dolog(LOG_ERR, "Failed to open log %s: %d (%s)",
 		      logfile, errno, strerror(errno));
 	if (fd != -1 && log_time_guest) {
-		if (write_with_timestamp(fd, "Logfile Opened",
-					 strlen("Logfile Opened"),
+		if (write_with_timestamp(fd, "Logfile Opened\n",
+					 strlen("Logfile Opened\n"),
 					 &log_time_guest_needts) < 0) {
 			dolog(LOG_ERR, "Failed to log opening timestamp "
 				       "in %s: %d (%s)", logfile, errno,
-- 
1.7.10.4

