From d94c6e1a4904406c97b54bf3f8158aa011c0a31f Mon Sep 17 00:00:00 2001
From: Rob Hoes <rob.hoes@citrix.com>
Date: Thu, 20 Jun 2013 15:48:52 +0100
Subject: [PATCH 32/34] libxl: define OCAML_READY

Signed-off-by: Rob Hoes <rob.hoes@citrix.com>
---
 tools/libxl/libxl.h |    2 ++
 1 file changed, 2 insertions(+)

diff --git a/tools/libxl/libxl.h b/tools/libxl/libxl.h
index 37e4d82..9970482 100644
--- a/tools/libxl/libxl.h
+++ b/tools/libxl/libxl.h
@@ -13,6 +13,8 @@
  * GNU Lesser General Public License for more details.
  */
 
+#define OCAML_READY
+
 /*
  * libxl API compatibility
  *
-- 
1.7.10.4

