# HG changeset patch
# Parent 2073131e3d9a9bf3508aac0b5b2730dbd57a8db5
diff -r 2073131e3d9a -r 43c4f1a142e6 tools/ocaml/libs/xl/Makefile
--- a/tools/ocaml/libs/xl/Makefile
+++ b/tools/ocaml/libs/xl/Makefile
@@ -26,7 +26,7 @@
 
 all: $(INTF) $(LIBS)
 
-xenlight.ml: xenlight.ml.in _libxl_types.ml.in
+xenlight.ml: xenlight.ml.in _libxl_types.ml.in xenlight.mli
 	$(Q)sed -e '1i\
 (*\
  * AUTO-GENERATED FILE DO NOT EDIT\
