# HG changeset patch
# Parent 45824746d068fd1a09a8d74d2b59ab664e83fba8

diff -r db2b2daf0cfc xen/arch/x86/cpu/intel.c
--- a/xen/arch/x86/cpu/intel.c
+++ b/xen/arch/x86/cpu/intel.c
@@ -104,7 +104,7 @@ static void __devinit set_cpuidmask(cons
 			return;
 		extra = "xsave ";
 		break;
-	case 0x2a: case 0x2d:
+	case 0x2a: case 0x2d: case 0x3e:
 		wrmsr(MSR_INTEL_CPUID1_FEATURE_MASK_V2,
 		      opt_cpuid_mask_ecx,
 		      opt_cpuid_mask_edx);
