# HG changeset patch
# Parent 5fe6bd6ffcbae5821f19fe6f78753d5a2507ce22

diff -r 5fe6bd6ffcba xen/Makefile
--- a/xen/Makefile
+++ b/xen/Makefile
@@ -111,7 +111,7 @@ delete-unfresh-files:
 
 .banner: Makefile
 	$(MAKE) -C tools
-	@tools/figlet/figlet -d tools/figlet Xen $(XEN_FULLVERSION) 2>$@2 >$@1
+	@tools/figlet/figlet -d tools/figlet Xen $$(echo $(XEN_FULLVERSION) | sed "s/-xs.*//") 2>$@2 >$@1
 	@cat $@1 $@2 >$@
 	@rm -f $@1 $@2
 
