# HG changeset patch
# Parent bf4f5582358211ca603f39af02adbe07a5848379

diff -r bf4f55823582 mk/Makefile.uclibc
--- a/mk/Makefile.uclibc	Wed Jan 22 21:28:59 2014 +0000
+++ b/mk/Makefile.uclibc	Wed Jan 22 21:33:10 2014 +0000
@@ -1,12 +1,10 @@
 include $(B_BASE)/common.mk
 
-include $(PROJECT_OUTPUTDIR)/uclibc-toolchain/toolchain.mk
+CROSS_COMPILE := musl-
 
-CROSS_COMPILE := $(UCLIBC_TOOLCHAIN_PATH)/$(UCLIBC_TOOL_PREFIX)
+CROSS_MAKE := PATH=$(MY_OBJ_DIR)/bin:$$PATH $(MAKE) CC=musl-gcc ZLIB= EXTRA_CFLAGS_XEN_TOOLS="-fgnu89-inline" LDFLAGS="--static -s"
 
-CROSS_MAKE := $(MAKE) CROSS_COMPILE=$(CROSS_COMPILE) CROSS_SYS_ROOT=$(UCLIBC_TOOLCHAIN_PATH) ZLIB= EXTRA_CFLAGS_XEN_TOOLS="-fgnu89-inline" LDFLAGS="--static -s"
-
-STRIP := $(CROSS_COMPILE)strip --strip-unneeded
+STRIP := strip --strip-unneeded
 
 REPODEP := $(call pq_req,xen-4.3)
 REPO    := $(call pq_loc,xen-4.3)
@@ -16,39 +14,38 @@ SOURCE_DIR           := $(MY_OBJ_DIR)/uc
 SOURCE_STAMP         := $(MY_OBJ_DIR)/.source_stamp
 
 LIBXC_BUILD_STAMP    := $(MY_OBJ_DIR)/.libxc_build_stamp
-LIBXC_INSTALL_STAMP  := $(MY_OBJ_DIR)/.libxc_install_stamp
-LIBXC_INSTALL_DIR    := $(MY_OBJ_DIR)/libxc.install
 
 XENSTORE_BUILD_STAMP := $(MY_OBJ_DIR)/.xenstore_build_stamp
 
 BINARY_INSTALL_STAMP := $(MY_OBJ_DIR)/.binary_install_stamp
 BINARY_INSTALL_DIR   := $(MY_OBJ_DIR)/bin.install
 
-LIBXC_PKG            := $(MY_OUTPUT_DIR)/uclibc-libxc.tar.bz2
 BINARY_PKG           := $(MY_OUTPUT_DIR)/uclibc-binaries.tar.bz2
 SOURCE_PKG           := $(MY_OUTPUT_DIR)/uclibc-sources.tar.bz2
 
-INSTALLER_FILES_STAMP:= $(MY_OBJ_DIR)/.installer_stamp
-
 XEN_HG_INCLUDE := -I tools/libxc -I tools/xenstore -I tools/misc \
                   -I tools/include -I tools/check -I tools/Rules.mk \
                   -I tools/cross-install -I Config.mk -I config \
                   -I xen/include/public -I xen/common/libelf \
                   -I xen/include/xen -I COPYING
 
+MUSL_SRC	:= $(CARBON_DISTFILES)/musl/musl-0.9.14.tar.gz
+MUSL_STAMP	:= $(MY_OBJ_DIR)/.musl_stamp
+
+ZLIB_SRC	:= $(CARBON_DISTFILES)/musl/zlib-1.2.8.tar.gz
+ZLIB_STAMP	:= $(MY_OBJ_DIR)/.zlib_stamp
+
 .PHONY: build
-build: $(LIBXC_PKG) $(BINARY_PKG) $(INSTALLER_FILES_STAMP)
+build: $(BINARY_PKG)
 	@ :
 
 .PHONY: clean
 clean:
 	rm -f $(SOURCE_STAMP)
-	rm -f $(LIBXC_BUILD_STAMP) $(LIBXC_INSTALL_STAMP)
+	rm -f $(LIBXC_BUILD_STAMP)
 	rm -f $(XENSTORE_BUILD_STAMP)
 	rm -f $(BINARY_INSTALL_STAMP)
-	rm -f $(INSTALLER_FILES_STAMP)
-	rm -rf $(REPO)
-	rm -rf $(LIBXC_INSTALL_DIR)
+	rm -f $(MUSL_STAMP) $(ZLIB_STAMP)
 	rm -rf $(BINARY_INSTALL_DIR)
 
 $(SOURCE_PKG): $(REPODEP)
@@ -58,26 +55,34 @@ clean:
 	tar -C $(MY_OBJ_DIR) --extract --bzip2 --file $<
 	touch $@
 
-$(LIBXC_BUILD_STAMP): $(SOURCE_STAMP)
+$(MUSL_STAMP): $(MUSL_SRC)
+	cd $(MY_OBJ_DIR) && rm -rf bin lib include musl-*
+	tar oxf $< -C $(MY_OBJ_DIR)
+	set -e ;\
+	cd $(MY_OBJ_DIR)/musl-* ;\
+	./configure --prefix=$(MY_OBJ_DIR) --disable-shared ;\
+	make -j4 ;\
+	make install
+	rm -rf $(MY_OBJ_DIR)/musl-*
+	sed -ri 's,-isystem include%s,-isystem include -isystem /usr/include%s,' $(MY_OBJ_DIR)/lib/musl-gcc.specs
+	touch $@
+
+$(ZLIB_STAMP): $(ZLIB_SRC) $(MUSL_STAMP)
+	tar oxf $< -C $(MY_OBJ_DIR)
+	set -e ;\
+	cd $(MY_OBJ_DIR)/zlib-* ;\
+	export PATH=$(MY_OBJ_DIR)/bin:$$PATH ;\
+	CC=musl-gcc ./configure --prefix=$(MY_OBJ_DIR) ;\
+	make -j4 ;\
+	make install
+	rm -rf $(MY_OBJ_DIR)/zlib-*
+	touch $@
+
+$(LIBXC_BUILD_STAMP): $(SOURCE_STAMP) $(MUSL_STAMP) $(ZLIB_STAMP)
 	$(CROSS_MAKE) -C $(SOURCE_DIR)/tools/include
 	$(CROSS_MAKE) -C $(SOURCE_DIR)/tools/libxc LIBXC_SUPPORT_THREADING=no all
 	touch $@
 
-$(LIBXC_INSTALL_STAMP): $(LIBXC_BUILD_STAMP)
-	$(CROSS_MAKE) -C $(SOURCE_DIR)/tools/libxc LIBXC_SUPPORT_THREADING=no DESTDIR=$(LIBXC_INSTALL_DIR) install
-
-	mkdir -p $(LIBXC_INSTALL_DIR)$(UCLIBC_TOOLCHAIN)/lib
-	mv $(LIBXC_INSTALL_DIR)/usr/lib/* $(LIBXC_INSTALL_DIR)$(UCLIBC_TOOLCHAIN)/lib/
-	rmdir $(LIBXC_INSTALL_DIR)/usr/lib
-
-	mkdir -p $(LIBXC_INSTALL_DIR)$(UCLIBC_TOOLCHAIN)/include
-	mv $(LIBXC_INSTALL_DIR)/usr/include/* $(LIBXC_INSTALL_DIR)$(UCLIBC_TOOLCHAIN)/include/
-	rmdir $(LIBXC_INSTALL_DIR)/usr/include
-	touch $@
-
-$(LIBXC_PKG): $(LIBXC_INSTALL_STAMP)
-	tar --create --bzip2 --file $@ --directory $(LIBXC_INSTALL_DIR) .
-
 $(XENSTORE_BUILD_STAMP): $(LIBXC_BUILD_STAMP)
 	$(CROSS_MAKE) -C $(SOURCE_DIR)/tools/xenstore XENSTORE_STATIC_CLIENTS=y clients
 	touch $@
@@ -90,7 +95,3 @@ clean:
 
 $(BINARY_PKG): $(BINARY_INSTALL_STAMP)
 	tar --create --bzip2 --file $@ --directory $(BINARY_INSTALL_DIR) .
-
-$(INSTALLER_FILES_STAMP): $(BINARY_INSTALL_STAMP)
-	mkdir -p $(MY_INSTALLER_FILES)/usr/bin
-	touch $@
