# HG changeset patch
# Parent 3bd72d8c9c7a26d397900a68e052509b4fe75a32

diff -r 3bd72d8c9c7a xen/Makefile
--- a/xen/Makefile
+++ b/xen/Makefile
@@ -113,6 +113,7 @@ delete-unfresh-files:
 		rm -f include/xen/compile.h; \
 	fi
 
+.banner: XEN_FULLVERSION := $$(echo $(XEN_FULLVERSION) | sed "s/-xs.*//")
 .banner: Makefile
 	@if which figlet >/dev/null 2>&1 ; then \
 		echo " Xen $(XEN_FULLVERSION)" | figlet -f tools/xen.flf > $@.tmp; \
