# HG changeset patch
# Parent 09536eba67195633cbdea522ec766035a89d32f8

diff -r 09536eba6719 xen/Makefile
--- a/xen/Makefile
+++ b/xen/Makefile
@@ -109,6 +109,7 @@ delete-unfresh-files:
 		rm -f include/xen/compile.h; \
 	fi
 
+.banner: XEN_FULLVERSION := $$(echo $(XEN_FULLVERSION) | sed "s/-xs.*//")
 .banner: Makefile
 	$(MAKE) -C tools
 	@tools/figlet/figlet -d tools/figlet Xen $(XEN_FULLVERSION) 2>$@2 >$@1
