# HG changeset patch
# Parent f58d57edb121780d831a42fb4316783836d30a4e

diff -r f58d57edb121 docs/Makefile
--- a/docs/Makefile
+++ b/docs/Makefile
@@ -86,6 +86,12 @@ install: all
 	cp -R man5 $(DESTDIR)$(MANDIR)
 	[ ! -d html ] || cp -R html $(DESTDIR)$(DOCDIR)
 
+.PHONY: install-man-pages
+install-man-pages: man-pages
+	$(INSTALL_DIR) $(DESTDIR)$(MANDIR)
+	cp -R man1 $(DESTDIR)$(MANDIR)
+	cp -R man5 $(DESTDIR)$(MANDIR)
+
 html/index.html: $(DOC_HTML) $(CURDIR)/gen-html-index INDEX
 	$(PERL) -w -- $(CURDIR)/gen-html-index -i INDEX html $(DOC_HTML)
 
