# HG changeset patch
# Parent 8f673d55adebdbdd77881363275a035a812a6aca

diff -r 8f673d55adeb xen/arch/x86/mm/shadow/private.h
--- a/xen/arch/x86/mm/shadow/private.h
+++ b/xen/arch/x86/mm/shadow/private.h
@@ -552,7 +552,18 @@ static inline void sh_put_ref(struct vcp
     struct page_info *sp = mfn_to_page(smfn);
 
     ASSERT(mfn_valid(smfn));
-    ASSERT(sp->u.sh.head);
+    if ( unlikely(!sp->u.sh.head) )
+    {
+        unsigned long *ptr = (unsigned long *)sp;
+        unsigned i;
+        printk("CA-114235 - d%"PRId16"v%d, smfn %"PRI_mfn
+               ": dumping page_info (size %zu):\n",
+               v->domain->domain_id, v->vcpu_id, mfn_x(smfn), sizeof *sp);
+        for ( i = 0; i < sizeof *sp / sizeof (unsigned long); ++i )
+            printk("    0x%02x: %016lx\n", i, ptr[i]);
+
+        ASSERT(sp->u.sh.head);
+    }
     ASSERT(!(sp->count_info & PGC_count_mask));
 
     /* If this is the entry in the up-pointer, remove it */
