# HG changeset patch
# Parent 0fad04d96f730a2e69669ddb29110b8e64552c6b

diff -r 0fad04d96f73 xen/arch/x86/mm/shadow/common.c
--- a/xen/arch/x86/mm/shadow/common.c
+++ b/xen/arch/x86/mm/shadow/common.c
@@ -1367,6 +1367,8 @@ static void shadow_blow_tables(struct do
 
     ASSERT(v != NULL);
 
+    printk("DBG - shadow_blow_tables(domain %"PRId16")\n", d->domain_id);
+
     /* Pass one: unpin all pinned pages */
     foreach_pinned_shadow(d, sp, t)
     {
diff -r 0fad04d96f73 xen/arch/x86/mm/shadow/private.h
--- a/xen/arch/x86/mm/shadow/private.h
+++ b/xen/arch/x86/mm/shadow/private.h
@@ -552,7 +552,19 @@ static inline void sh_put_ref(struct vcp
     struct page_info *sp = mfn_to_page(smfn);
 
     ASSERT(mfn_valid(smfn));
-    ASSERT(sp->u.sh.head);
+    if ( unlikely(!sp->u.sh.head) )
+    {
+        unsigned long *ptr = (unsigned long *)sp;
+        unsigned i;
+        printk("CA-114235 - d%"PRId16"v%d, smfn %"PRI_mfn
+               ": dumping page_info (size %zu), sp %p:\n",
+               v->domain->domain_id, v->vcpu_id, mfn_x(smfn),
+               sizeof *sp, _p(sp));
+        for ( i = 0; i < sizeof *sp / sizeof (unsigned long); ++i )
+            printk("    0x%02lx: %016lx\n", i * sizeof(unsigned long), ptr[i]);
+
+        ASSERT(sp->u.sh.head);
+    }
     ASSERT(!(sp->count_info & PGC_count_mask));
 
     /* If this is the entry in the up-pointer, remove it */
