# HG changeset patch
# Parent 825205423ec04a2c21b415addcc3c797cd7c1ee3

diff -r 825205423ec0 xen/arch/x86/hvm/hvm.c
--- a/xen/arch/x86/hvm/hvm.c
+++ b/xen/arch/x86/hvm/hvm.c
@@ -1182,8 +1182,10 @@ static int hvm_load_cpu_ctxt(struct doma
 
     if ( ctxt.cr4 & HVM_CR4_GUEST_RESERVED_BITS(v) )
     {
-        printk(XENLOG_G_ERR "HVM%d restore: bad CR4 %#" PRIx64 "\n",
-               d->domain_id, ctxt.cr4);
+        printk(XENLOG_G_ERR "HVM%d restore: bad CR4 %#" PRIx64
+               " (reserved bits %#"PRIx64", xor %#"PRIx64")\n",
+               d->domain_id, ctxt.cr4, HVM_CR4_GUEST_RESERVED_BITS(v),
+               ctxt.cr4 ^HVM_CR4_GUEST_RESERVED_BITS(v));
         return -EINVAL;
     }
 
