# HG changeset patch
# Parent d856dff8effeed19e64db235ce71e04781f21f4f

diff -r d856dff8effe tools/firmware/hvmloader/acpi/static_tables.c
--- a/tools/firmware/hvmloader/acpi/static_tables.c
+++ b/tools/firmware/hvmloader/acpi/static_tables.c
@@ -157,7 +157,7 @@ struct acpi_20_waet Waet = {
         .creator_id   = ACPI_CREATOR_ID,
         .creator_revision = ACPI_CREATOR_REVISION
     },
-    .flags = ACPI_WAET_FLAGS
+    .flags = 0 //ACPI_WAET_FLAGS
 };
 
 /*
diff -r d856dff8effe xen/arch/x86/hvm/rtc.c
--- a/xen/arch/x86/hvm/rtc.c
+++ b/xen/arch/x86/hvm/rtc.c
@@ -51,7 +51,7 @@ enum rtc_mode {
 };
 
 /* This must be in sync with how hvmloader sets the ACPI WAET flags. */
-#define mode_is(d, m) ((void)(d), rtc_mode_##m == rtc_mode_no_ack)
+#define mode_is(d, m) ((void)(d), rtc_mode_##m == rtc_mode_strict)
 #define rtc_mode_is(s, m) mode_is(vrtc_domain(s), m)
 
 static void rtc_copy_date(RTCState *s);
