diff -r 9b4c1763bba7 xen/drivers/passthrough/vtd/qinval.c
--- a/xen/drivers/passthrough/vtd/qinval.c
+++ b/xen/drivers/passthrough/vtd/qinval.c
@@ -219,6 +219,7 @@ static int queue_invalidate_wait(struct 
         {
             if ( NOW() > (start_time + DMAR_OPERATION_TIMEOUT) )
             {
+                WARN();
                 print_qi_regs(iommu);
                 panic("queue invalidate wait descriptor was not executed\n");
             }
