# HG changeset patch
# Parent fdcd253e4e1c9e066dfc3db5745d3f3c7e29153b

diff -r fdcd253e4e1c tools/firmware/hvmloader/acpi/static_tables.c
--- a/tools/firmware/hvmloader/acpi/static_tables.c
+++ b/tools/firmware/hvmloader/acpi/static_tables.c
@@ -157,7 +157,7 @@ struct acpi_20_waet Waet = {
         .creator_id   = ACPI_CREATOR_ID,
         .creator_revision = ACPI_CREATOR_REVISION
     },
-    .flags = ACPI_WAET_FLAGS
+    .flags = 0 //ACPI_WAET_FLAGS
 };
 
 /*
