# HG changeset patch
# Parent 864a7f2e7b46b23060f48b211b9d5cec6bc38b12

diff -r 864a7f2e7b46 xen/arch/x86/hvm/hvm.c
--- a/xen/arch/x86/hvm/hvm.c
+++ b/xen/arch/x86/hvm/hvm.c
@@ -3966,6 +3966,10 @@ long do_hvm_op(unsigned long op, XEN_GUE
                     rc = -EINVAL;
 
                 break;
+            case HVM_PARAM_VPT_ALIGN:
+                gdprintk(XENLOG_WARNING,
+                         "**DBG: Setting VTP_ALIGN to %"PRIu64"\n", a.value);
+                break;
             case HVM_PARAM_ACPI_IOPORTS_LOCATION:
                 rc = pmtimer_change_ioport(d, a.value);
                 break;
