# HG changeset patch
# Parent b3b4f9178c9e53d7e2ef547b10ff24cb1ef36b91

diff -r b3b4f9178c9e tools/misc/xen-hvmctx.c
--- a/tools/misc/xen-hvmctx.c
+++ b/tools/misc/xen-hvmctx.c
@@ -307,6 +307,7 @@ static void dump_pit(void)
 
 static void dump_rtc(void)
 {
+    uint8_t v;
     HVM_SAVE_TYPE(RTC) r;
     READ(r);
     printf("    RTC: regs 0x%2.2x 0x%2.2x 0x%2.2x 0x%2.2x 0x%2.2x 0x%2.2x 0x%2.2x 0x%2.2x\n",
@@ -316,6 +317,68 @@ static void dump_rtc(void)
            r.cmos_data[8], r.cmos_data[9], r.cmos_data[10], r.cmos_data[11], 
            r.cmos_data[12], r.cmos_data[13], r.cmos_index);
 
+    if ( r.cmos_data[0xb] & 0x2 ) // Binary mode
+    {
+        printf("    RTC decode (binary)\n");
+        printf("         Current time:  %02d:%02d:%02d\n",
+               r.cmos_data[4], r.cmos_data[2], r.cmos_data[0]);
+        printf("         Current alarm: %02d:%02d:%02d\n",
+               r.cmos_data[5], r.cmos_data[3], r.cmos_data[1]);
+        printf("         Current date:  y %d, m %d, d %d, day of week %d\n",
+               r.cmos_data[9], r.cmos_data[8], r.cmos_data[7], r.cmos_data[6]);
+    }
+    else // BCD mode
+    {
+        printf("    RTC decode (bcd)\n");
+        printf("         Current time:  %x%x:%x%x:%x%x\n",
+               r.cmos_data[4] >> 4, r.cmos_data[4] & 0xf,
+               r.cmos_data[2] >> 4, r.cmos_data[2] & 0xf,
+               r.cmos_data[0] >> 4, r.cmos_data[0] & 0xf
+            );
+        printf("         Current alarm: %x%x:%x%x:%x%x\n",
+               r.cmos_data[5] >> 4, r.cmos_data[5] & 0xf,
+               r.cmos_data[3] >> 4, r.cmos_data[3] & 0xf,
+               r.cmos_data[1] >> 4, r.cmos_data[1] & 0xf
+            );
+        printf("         Current date:  y %x%x, m %x%x, d %x%x, day of week %x%x\n",
+               r.cmos_data[9] >> 4, r.cmos_data[9] & 0xf,
+               r.cmos_data[8] >> 4, r.cmos_data[8] & 0xf,
+               r.cmos_data[7] >> 4, r.cmos_data[7] & 0xf,
+               r.cmos_data[6] >> 4, r.cmos_data[6] & 0xf
+            );
+    }
+
+    v = r.cmos_data[0xa];
+    printf("         REG A: 0x%02x - UIP%c\n",
+           v, v & 0x80 ? '+' : '-');
+
+    v = r.cmos_data[0xb];
+    printf("         REG B: 0x%02x - %s %s %s %s %s %s %s %s\n", v,
+           v & 0x80 ? "SET" : "",
+           v & 0x40 ? "PIE" : "",
+           v & 0x20 ? "AIE" : "",
+           v & 0x10 ? "UIE" : "",
+           v & 0x08 ? "SQWE" : "",
+           v & 0x04 ? "DM" : "",
+           v & 0x02 ? "24" : "12",
+           v & 0x01 ? "DSE" : ""
+        );
+
+    v = r.cmos_data[0xc];
+    printf("         REG C: 0x%02x - %s %s %s %s %s\n", v,
+           v & 0x80 ? "IRQF" : "",
+           v & 0x40 ? "PF" : "",
+           v & 0x20 ? "AF" : "",
+           v & 0x10 ? "UF" : "",
+           v & 0x0f ? "RAZ invalid" : ""
+        );
+
+    v = r.cmos_data[0xd];
+    printf("         REG D: 0x%02x - %s %s\n", v,
+           v & 0x80 ? "VRT" : "",
+           v & 0x7f ? "RAZ invalid" : ""
+        );
+
 }
 
 static void dump_hpet(void)
diff -r b3b4f9178c9e xen/arch/x86/hvm/rtc.c
--- a/xen/arch/x86/hvm/rtc.c
+++ b/xen/arch/x86/hvm/rtc.c
@@ -467,6 +467,9 @@ static int rtc_ioport_write(void *opaque
             rtc_timer_update(s);
         break;
     case RTC_REG_B:
+        if ( (data ^ s->hw.cmos_data[RTC_REG_B]) & RTC_DM_BINARY )
+            gdprintk(XENLOG_INFO, "** Guest changed RTC[B].DM.  Now %d\n",
+                     !!(data & RTC_DM_BINARY));
         if ( data & RTC_SET )
         {
             /* set mode: reset UIP mode */
