# HG changeset patch
# Parent 93e6132491dd0f1916e7e13428677eb46107e3e8

diff -r 93e6132491dd tools/firmware/hvmloader/acpi/build.c
--- a/tools/firmware/hvmloader/acpi/build.c
+++ b/tools/firmware/hvmloader/acpi/build.c
@@ -186,20 +186,20 @@ static struct acpi_20_hpet *construct_hp
     return hpet;
 }
 
-static struct acpi_20_waet *construct_waet(void)
-{
-    struct acpi_20_waet *waet;
+/* static struct acpi_20_waet *construct_waet(void) */
+/* { */
+/*     struct acpi_20_waet *waet; */
 
-    waet = mem_alloc(sizeof(*waet), 16);
-    if (!waet) return NULL;
+/*     waet = mem_alloc(sizeof(*waet), 16); */
+/*     if (!waet) return NULL; */
 
-    memcpy(waet, &Waet, sizeof(*waet));
+/*     memcpy(waet, &Waet, sizeof(*waet)); */
 
-    waet->header.length = sizeof(*waet);
-    set_checksum(waet, offsetof(struct acpi_header, checksum), sizeof(*waet));
+/*     waet->header.length = sizeof(*waet); */
+/*     set_checksum(waet, offsetof(struct acpi_header, checksum), sizeof(*waet)); */
 
-    return waet;
-}
+/*     return waet; */
+/* } */
 
 static int construct_passthrough_tables(unsigned long *table_ptrs,
                                         int nr_tables)
@@ -253,7 +253,7 @@ static int construct_secondary_tables(un
     int nr_tables = 0;
     struct acpi_20_madt *madt;
     struct acpi_20_hpet *hpet;
-    struct acpi_20_waet *waet;
+    /* struct acpi_20_waet *waet; */
     struct acpi_20_tcpa *tcpa;
     unsigned char *ssdt;
     static const uint16_t tis_signature[] = {0x0001, 0x0001, 0x0001};
@@ -278,9 +278,9 @@ static int construct_secondary_tables(un
     }
 
     /* WAET. */
-    waet = construct_waet();
-    if (!waet) return -1;
-    table_ptrs[nr_tables++] = (unsigned long)waet;
+    /* waet = construct_waet(); */
+    /* if (!waet) return -1; */
+    /* table_ptrs[nr_tables++] = (unsigned long)waet; */
 
     if ( battery_port_exists() )
     {
