# HG changeset patch
# Parent e43b5ebc85256baeab8b8ee56a3c8bf00e5e68b5

diff -r e43b5ebc8525 xen/tools/scmversion
--- a/xen/tools/scmversion
+++ b/xen/tools/scmversion
@@ -69,15 +69,11 @@ scm_version()
 
 	# Check for mercurial and a mercurial repo.
 	if test -d .hg && hgid=`hg id 2>/dev/null`; then
-		id=`printf '%s' "$hgid" | sed 's/[+ ].*//'`
-		date=`hg parents --template "{date|date}"`
-		printf '%s %s%s' "$date" hg: "$id"
+		TEMPLATE="{rev}:{node|short}"
+		BASE_CSET_STR="$(hg log -r qparent --template $TEMPLATE)"
+		PQ_CSET_STR="$(hg log -R .hg/patches -r tip --template $TEMPLATE)"
 
-		# Are there uncommitted changes?
-		# These are represented by + after the changeset id.
-		case "$hgid" in
-			*+|*+\ *) printf '%s' -dirty ;;
-		esac
+		printf '%s, pq %s' "$BASE_CSET_STR" "$PQ_CSET_STR"
 
 		# All done with mercurial
 		return
