# HG changeset patch
# Parent 6b81dcadbbd20b99993757aa0e05556863e7bca6

diff -r 6b81dcadbbd2 xen/arch/x86/crash.c
--- a/xen/arch/x86/crash.c	Thu Nov 07 16:03:23 2013 +0000
+++ b/xen/arch/x86/crash.c	Thu Nov 07 17:19:23 2013 +0000
@@ -121,6 +121,12 @@ static void nmi_shootdown_cpus(void)
     unsigned long msecs;
     int i, cpu = smp_processor_id();
 
+    /*
+     * About to change all the NMI trap handlers; no point in having
+     * the NMI watchdog running.
+     */
+    disable_lapic_nmi_watchdog();
+
     local_irq_disable();
 
     crashing_cpu = cpu;
diff -r 6b81dcadbbd2 xen/arch/x86/nmi.c
--- a/xen/arch/x86/nmi.c	Thu Nov 07 16:03:23 2013 +0000
+++ b/xen/arch/x86/nmi.c	Thu Nov 07 17:19:23 2013 +0000
@@ -165,7 +165,7 @@ static void nmi_timer_fn(void *unused)
     set_timer(&this_cpu(nmi_timer), NOW() + MILLISECS(1000));
 }
 
-static void disable_lapic_nmi_watchdog(void)
+void disable_lapic_nmi_watchdog(void)
 {
     if (nmi_active <= 0)
         return;
diff -r 6b81dcadbbd2 xen/include/asm-x86/apic.h
--- a/xen/include/asm-x86/apic.h	Thu Nov 07 16:03:23 2013 +0000
+++ b/xen/include/asm-x86/apic.h	Thu Nov 07 17:19:23 2013 +0000
@@ -212,6 +212,7 @@ extern void smp_local_timer_interrupt (s
 extern void setup_boot_APIC_clock (void);
 extern void setup_secondary_APIC_clock (void);
 extern void setup_apic_nmi_watchdog (void);
+extern void disable_lapic_nmi_watchdog(void);
 extern int reserve_lapic_nmi(void);
 extern void release_lapic_nmi(void);
 extern void self_nmi(void);
