# HG changeset patch
# Parent d1bd17dd311f21006b557f7d59049d37eeca8f4d

diff -r d1bd17dd311f xen/include/xen/domain_page.h
--- a/xen/include/xen/domain_page.h
+++ b/xen/include/xen/domain_page.h
@@ -108,6 +108,9 @@ domain_mmap_cache_destroy(struct domain_
 #define __map_domain_page(pg)               page_to_virt(pg)
 #define unmap_domain_page(va)               ((void)(va))
 #define domain_page_map_to_mfn(va)          virt_to_mfn((unsigned long)(va))
+#define clear_domain_page(mfn)              clear_page(mfn_to_virt(mfn))
+#define copy_domain_page(dmfn, smfn)        copy_page(mfn_to_virt(dmfn), \
+                                                      mfn_to_virt(smfn))
 
 #define map_domain_page_global(mfn)         mfn_to_virt(mfn)
 #define __map_domain_page_global(pg)        page_to_virt(pg)
