# HG changeset patch
# Parent 0d4d7b32bc1d2ce25b22539b5c4171120a92fb66

diff -r 0d4d7b32bc1d tools/firmware/hvmloader/acpi/build.c
--- a/tools/firmware/hvmloader/acpi/build.c	Wed Apr 24 14:34:44 2013 +0100
+++ b/tools/firmware/hvmloader/acpi/build.c	Wed Apr 24 17:10:44 2013 +0100
@@ -210,6 +210,7 @@ static int construct_secondary_tables(un
     static const uint16_t tis_signature[] = {0x0001, 0x0001, 0x0001};
     uint16_t *tis_hdr;
     void *lasa;
+    const char *xs_str;
 
     /* MADT. */
     if ( (hvm_info->nr_vcpus > 1) || hvm_info->apic_mode )
@@ -238,7 +239,8 @@ static int construct_secondary_tables(un
         table_ptrs[nr_tables++] = (unsigned long)ssdt;
     }
 
-    if ( !strncmp(xenstore_read("platform/acpi_s3", "1"), "1", 1) )
+    xs_str = xenstore_read("platform/acpi_s3", "1");
+    if ( !strncmp(xs_str, "1", 1) || !strncmp(xs_str, "true", 4) )
     {
         ssdt = mem_alloc(sizeof(ssdt_s3), 16);
         if (!ssdt) return -1;
@@ -248,7 +250,8 @@ static int construct_secondary_tables(un
         printf("S3 disabled\n");
     }
 
-    if ( !strncmp(xenstore_read("platform/acpi_s4", "1"), "1", 1) )
+    xs_str = xenstore_read("platform/acpi_s4", "1");
+    if ( !strncmp(xs_str, "1", 1) || !strncmp(xs_str, "true", 4) )
     {
         ssdt = mem_alloc(sizeof(ssdt_s4), 16);
         if (!ssdt) return -1;
diff -r 0d4d7b32bc1d tools/firmware/hvmloader/hvmloader.c
--- a/tools/firmware/hvmloader/hvmloader.c	Wed Apr 24 14:34:44 2013 +0100
+++ b/tools/firmware/hvmloader/hvmloader.c	Wed Apr 24 17:10:44 2013 +0100
@@ -268,6 +268,7 @@ static void acpi_enable_sci(void)
 int main(void)
 {
     const struct bios_config *bios;
+    const char * xs_str;
     int acpi_enabled;
 
     /* Initialise hypercall stubs with RET, rendering them no-ops. */
@@ -318,7 +319,8 @@ int main(void)
     if ( bios->load_roms )
         bios->load_roms();
 
-    acpi_enabled = !strncmp(xenstore_read("platform/acpi", "1"), "1", 1);
+    xs_str = xenstore_read("platform/acpi", "1");
+    acpi_enabled = !strncmp(xs_str, "1", 1) || !strncmp(xs_str, "true", 4);
 
     if ( acpi_enabled )
     {
