# HG changeset patch
# Parent 85f5671a420f418a37a5a7eb22f74e130abc7652

diff -r 85f5671a420f tools/firmware/hvmloader/acpi/build.c
--- a/tools/firmware/hvmloader/acpi/build.c
+++ b/tools/firmware/hvmloader/acpi/build.c
@@ -189,6 +189,7 @@ static struct acpi_20_hpet *construct_hp
 static struct acpi_20_waet *construct_waet(void)
 {
     struct acpi_20_waet *waet;
+    const char *s;
 
     waet = mem_alloc(sizeof(*waet), 16);
     if (!waet) return NULL;
@@ -196,6 +197,28 @@ static struct acpi_20_waet *construct_wa
     memcpy(waet, &Waet, sizeof(*waet));
 
     waet->header.length = sizeof(*waet);
+
+    s = xenstore_read("platform/dbg-waet-rtc", NULL);
+    if ( s )
+    {
+        if ( !strncmp(s, "1", 1) || !strncmp(s, "true", 4) )
+            waet->flags |= 1u;
+        else
+            waet->flags &= ~1u;
+    }
+
+    s = xenstore_read("platform/dbg-waet-pm", NULL);
+    if ( s )
+    {
+        if ( !strncmp(s, "1", 1) || !strncmp(s, "true", 4) )
+            waet->flags |= 2u;
+        else
+            waet->flags &= ~2u;
+    }
+
+    printf("WAET flags: RTC %d, PM %d\n",
+           !!(waet->flags & 1), !!(waet->flags & 2));
+
     set_checksum(waet, offsetof(struct acpi_header, checksum), sizeof(*waet));
 
     return waet;
