# HG changeset patch
# Parent 991aab9c104f37c4eb3ba6fea786d71e5030535d

diff -r 991aab9c104f xen/common/kexec.c
--- a/xen/common/kexec.c
+++ b/xen/common/kexec.c
@@ -821,21 +821,17 @@ static int kexec_load_slot(struct kexec_
 
     ret = machine_kexec_load(kimage);
     if ( ret < 0 )
-        goto error;
+        return ret;
 
     crash_save_vmcoreinfo();
 
     ret = kexec_swap_images(kimage->type, kimage, &old_kimage);
     if ( ret < 0 )
-        goto error;
+        return ret;
 
     kexec_unload_image(old_kimage);
 
     return 0;
-
-error:
-    kimage_free(kimage);
-    return ret;
 }
 
 static uint16_t kexec_load_v1_arch(void)
