CA-55274: Assume only one checkpoint unless told otherwise.

diff -r b4cb8880e8f5 tools/libxc/xc_domain_restore.c
--- a/tools/libxc/xc_domain_restore.c
+++ b/tools/libxc/xc_domain_restore.c
@@ -1339,6 +1339,10 @@ int xc_domain_restore(xc_interface *xch,
     ctx->superpages = superpages;
     ctx->hvm = hvm;
 
+    /* Assume first checkpoint is last, unless we get a record
+     * at the head saying otherwise */
+    ctx->last_checkpoint = 1;
+
     ctxt = xc_hypercall_buffer_alloc(xch, ctxt, sizeof(*ctxt));
 
     if ( ctxt == NULL )
