diff -r b44c73016c75 tools/libxl/Makefile
--- a/tools/libxl/Makefile
+++ b/tools/libxl/Makefile
@@ -188,6 +188,7 @@ testidl: testidl.o libxlutil.so libxenli
 
 .PHONY: install
 install: all
+	$(INSTALL_DIR) $(DESTDIR)$(XEN_CONFIG_DIR)
 	$(INSTALL_DIR) $(DESTDIR)$(SBINDIR)
 	$(INSTALL_DIR) $(DESTDIR)$(LIBDIR)
 	$(INSTALL_DIR) $(DESTDIR)$(INCLUDEDIR)
@@ -206,6 +207,7 @@ install: all
 	$(INSTALL_DATA) libxlutil.a $(DESTDIR)$(LIBDIR)
 	$(INSTALL_DATA) libxl.h libxl_event.h libxl_json.h _libxl_types.h _libxl_types_json.h _libxl_list.h libxl_utils.h libxl_uuid.h $(DESTDIR)$(INCLUDEDIR)
 	$(INSTALL_DATA) bash-completion $(DESTDIR)$(BASH_COMPLETION_DIR)/xl.sh
+	$(INSTALL_DATA) xl.conf $(DESTDIR)$(XEN_CONFIG_DIR)/xl.conf
 
 .PHONY: clean
 clean:
diff -r b44c73016c75 tools/libxl/xl.conf
--- /dev/null
+++ b/tools/libxl/xl.conf
@@ -0,0 +1,2 @@
+autoballoon = 0
+
