diff -r bcf9da5ec51d mk/xen-4.3.spec.in
--- a/mk/xen-4.3.spec.in
+++ b/mk/xen-4.3.spec.in
@@ -239,6 +239,7 @@ rm -rf $RPM_BUILD_ROOT
 %{_bindir}/xentrace*
 %{_bindir}/xen-detect
 %{_sysconfdir}/bash_completion.d/xl.sh
+%{_sysconfdir}/xen/xl.conf
 %{_sbindir}/xl
 %{_sbindir}/xen-hvmctx
 %{_sbindir}/kdd
diff -r bcf9da5ec51d tools/libxl/Makefile
--- a/tools/libxl/Makefile
+++ b/tools/libxl/Makefile
@@ -205,6 +205,7 @@ install: all
 	$(INSTALL_DIR) $(DESTDIR)$(BASH_COMPLETION_DIR)
 	$(INSTALL_DIR) $(DESTDIR)$(XEN_RUN_DIR)
 	$(INSTALL_DIR) $(DESTDIR)$(PRIVATE_BINDIR)
+	$(INSTALL_DIR) $(DESTDIR)$(XEN_CONFIG_DIR)
 	$(INSTALL_PROG) xl $(DESTDIR)$(SBINDIR)
 	$(INSTALL_PROG) libxl-save-helper $(DESTDIR)$(PRIVATE_BINDIR)
 	$(INSTALL_PROG) libxenlight.so.$(MAJOR).$(MINOR) $(DESTDIR)$(LIBDIR)
@@ -217,6 +218,7 @@ install: all
 	$(INSTALL_DATA) libxlutil.a $(DESTDIR)$(LIBDIR)
 	$(INSTALL_DATA) libxl.h libxl_event.h libxl_json.h _libxl_types.h _libxl_types_json.h _libxl_list.h libxl_utils.h libxl_uuid.h $(DESTDIR)$(INCLUDEDIR)
 	$(INSTALL_DATA) bash-completion $(DESTDIR)$(BASH_COMPLETION_DIR)/xl.sh
+	$(INSTALL_DATA) xl.conf $(DESTDIR)$(XEN_CONFIG_DIR)/xl.conf
 
 .PHONY: clean
 clean:
diff -r bcf9da5ec51d tools/libxl/xl.conf
--- /dev/null
+++ b/tools/libxl/xl.conf
@@ -0,0 +1,1 @@
+autoballoon = 0
