diff -r 1607ab7e53a6 mk/xen-4.4.spec.in
--- a/mk/xen-4.4.spec.in
+++ b/mk/xen-4.4.spec.in
@@ -324,6 +324,7 @@ rm -rf $RPM_BUILD_ROOT
 %{_sysconfdir}/rc.d/init.d/xen-watchdog
 %config %{_sysconfdir}/xen/oxenstored.conf
 %{_sysconfdir}/xen/scripts/hotplugpath.sh
+%config %{_sysconfdir}/xen/xl.conf
 /opt/xensource/bin/xenstored
 %{_bindir}/pygrub
 %{_bindir}/xen-detect
diff -r 1607ab7e53a6 tools/libxl/Makefile
--- a/tools/libxl/Makefile
+++ b/tools/libxl/Makefile
@@ -207,6 +207,7 @@ install: all
 	$(INSTALL_DIR) $(DESTDIR)$(BASH_COMPLETION_DIR)
 	$(INSTALL_DIR) $(DESTDIR)$(XEN_RUN_DIR)
 	$(INSTALL_DIR) $(DESTDIR)$(PRIVATE_BINDIR)
+	$(INSTALL_DIR) $(DESTDIR)$(XEN_CONFIG_DIR)
 	$(INSTALL_PROG) xl $(DESTDIR)$(SBINDIR)
 	$(INSTALL_PROG) libxl-save-helper $(DESTDIR)$(PRIVATE_BINDIR)
 	$(INSTALL_PROG) libxenlight.so.$(MAJOR).$(MINOR) $(DESTDIR)$(LIBDIR)
@@ -219,6 +220,7 @@ install: all
 	$(INSTALL_DATA) libxlutil.a $(DESTDIR)$(LIBDIR)
 	$(INSTALL_DATA) libxl.h libxl_event.h libxl_json.h _libxl_types.h _libxl_types_json.h _libxl_list.h libxl_utils.h libxl_uuid.h $(DESTDIR)$(INCLUDEDIR)
 	$(INSTALL_DATA) bash-completion $(DESTDIR)$(BASH_COMPLETION_DIR)/xl.sh
+	$(INSTALL_DATA) xl.conf $(DESTDIR)$(XEN_CONFIG_DIR)/xl.conf
 
 .PHONY: clean
 clean:
diff -r 1607ab7e53a6 tools/libxl/xl.conf
--- /dev/null
+++ b/tools/libxl/xl.conf
@@ -0,0 +1,1 @@
+autoballoon = 0
