CA-55274: Assume only one checkpoint unless told otherwise.

diff -r 0d9d06651101 tools/libxc/xc_domain_restore.c
--- a/tools/libxc/xc_domain_restore.c
+++ b/tools/libxc/xc_domain_restore.c
@@ -1256,6 +1256,10 @@ int xc_domain_restore(xc_interface *xch,
 
     memset(ctx, 0, sizeof(*ctx));
 
+    /* Assume first checkpoint is last, unless we get a record
+     * at the head saying otherwise */
+    ctx->last_checkpoint = 1;
+
     ctxt = xc_hypercall_buffer_alloc(xch, ctxt, sizeof(*ctxt));
 
     if ( ctxt == NULL )
