# HG changeset patch
# Parent 12feba47764bda2bcea0f7af193c16f4b7fb8971

diff -r 12feba47764b tools/libxc/xc_domain_restore.c
--- a/tools/libxc/xc_domain_restore.c
+++ b/tools/libxc/xc_domain_restore.c
@@ -100,8 +100,8 @@ static ssize_t rdexact(xc_interface *xch
         if ( (len == -1) && ((errno == EINTR) || (errno == EAGAIN)) )
             continue;
         if ( len == 0 ) {
-            ERROR("0-length read");
-            errno = 0;
+            errno = EPIPE;
+            return -1;
         }
         if ( len <= 0 ) {
             ERROR("%s failed (read rc: %zd, errno: %d)", __func__, len, errno);
@@ -1664,7 +1664,7 @@ int xc_domain_restore(xc_interface *xch,
             pagebuf.nr_physpages = pagebuf.nr_pages = 0;
             pagebuf.compbuf_pos = pagebuf.compbuf_size = 0;
             if ( pagebuf_get_one(xch, ctx, &pagebuf, io_fd, dom) < 0 ) {
-                PERROR("Error when reading batch");
+	        /* pagebuf_get_one already returned a proper error */
                 goto out;
             }
         }
