diff -r 2ee07da1e4ba tools/libxc/xc_domain_restore.c
--- a/tools/libxc/xc_domain_restore.c	Mon Jan 13 10:37:28 2014 +0000
+++ b/tools/libxc/xc_domain_restore.c	Mon Jan 13 10:44:33 2014 +0000
@@ -1241,7 +1241,7 @@
     /* Now allocate a bunch of mfns for this batch */
     if ( nr_mfns )
     {
-        DPRINTF("Mapping order 0,  %d; first pfn %lx\n", nr_mfns, ctx->p2m_batch[0]);
+        DBGPRINTF("Mapping order 0,  %d; first pfn %lx\n", nr_mfns, ctx->p2m_batch[0]);
     
         if (!ctx->hvm && ctx->superpages)
             rc = alloc_superpage_mfns(xch, dom, ctx, nr_mfns);
