# HG changeset patch
# Parent 17c4cd134b82a5d6b2ce3e6bdf148544bac6cc4d

diff --git a/tools/libxc/xc_domain_restore.c b/tools/libxc/xc_domain_restore.c
--- a/tools/libxc/xc_domain_restore.c
+++ b/tools/libxc/xc_domain_restore.c
@@ -437,6 +437,13 @@
         return -1;
     }
 
+    /* suspend images from legacy xenopsd contain a newline and 32-bit int */
+    if ( *qbuf == '\n' && !memcmp(qbuf + 5, "QEVM", 4) ) {
+        DPRINTF("Detected legacy xenopsd signature; consuming...\n");
+        dlen -= 5;
+        memmove(qbuf, qbuf + 5, dlen);
+    }
+
     if ( memcmp(qbuf, "QEVM", 4) ) {
         ERROR("Invalid QEMU magic: 0x%08x", *(unsigned long*)qbuf);
         free(qbuf);
