diff -r 63dc54731878 -r cb3f0bf2479e tools/libxc/xc_private.c
--- a/tools/libxc/xc_private.c	Mon Mar 14 16:42:20 2011 +0000
+++ b/tools/libxc/xc_private.c	Mon Mar 14 17:55:42 2011 +0000
@@ -141,7 +141,7 @@ static struct xc_interface_core *xc_inte
     if (!xch->error_handler) {
         xch->error_handler = xch->error_handler_tofree =
             (xentoollog_logger*)
-            xtl_createlogger_stdiostream(stderr, XTL_PROGRESS, 0);
+            xtl_createlogger_stdiostream(stderr, XTL_DETAIL, 0);
         if (!xch->error_handler)
             goto err;
     }
