Needed for the xapi toolstack

diff -r 2808fcff033f tools/libxc/xc_dom_core.c
--- a/tools/libxc/xc_dom_core.c	Fri Jul 02 12:45:21 2010 +0100
+++ b/tools/libxc/xc_dom_core.c	Fri Jul 02 12:47:23 2010 +0100
@@ -793,6 +793,12 @@
     return -1;
 }
 
+/* Accessors */
+const char *xc_dom_get_native_protocol(struct xc_dom_image *dom)
+{
+    return dom->arch_hooks ? dom->arch_hooks->native_protocol : "";
+}
+
 /*
  * Local variables:
  * mode: C
diff -r 2808fcff033f tools/libxc/xenguest.h
--- a/tools/libxc/xenguest.h	Fri Jul 02 12:45:21 2010 +0100
+++ b/tools/libxc/xenguest.h	Fri Jul 02 12:47:23 2010 +0100
@@ -190,4 +190,8 @@
                       unsigned long max_mfn,
                       int prot,
                       unsigned long *mfn0);
+
+/* Accessors for struct xc_dom_image */
+const char *xc_dom_get_native_protocol(struct xc_dom_image *dom);
+
 #endif /* XENGUEST_H */
