# HG changeset patch
# Parent 1a960349a0bd9f85fd475ed426d6b956388f71e1

diff -r 1a960349a0bd mk/xen-4.4.spec.in
--- a/mk/xen-4.4.spec.in
+++ b/mk/xen-4.4.spec.in
@@ -220,6 +220,8 @@ rm -f %{buildroot}/usr/share/doc/xen/REA
 mkdir -p $RPM_BUILD_ROOT/opt/xensource/bin/
 cp -f %{buildroot}%{_sbindir}/oxenstored %{buildroot}/opt/xensource/bin/xenstored
 
+ln -s convert-legacy-stream %{buildroot}%{_libdir}/%{name}/bin/legacy.py
+
 # do not strip symbols some unstripped xen
 chmod -x %{buildroot}/boot/xen-syms-*
 
@@ -404,14 +406,19 @@ rm -rf $RPM_BUILD_ROOT
 %{python_sitearch}/pygrub-*.egg-info
 %{python_sitearch}/xen-*.egg-info
 #{python_sitearch}/xen/__init__.py*           - Must not duplicate xen-tools
+%{python_sitearch}/xen/streamv2/__init__.py*
+%{python_sitearch}/xen/streamv2/format.py*
 #{python_sitearch}/xen/lowlevel/__init__.py*  - Must not duplicate xen-tools
 %{python_sitearch}/xen/lowlevel/ptsname.so
 %{python_sitearch}/xen/lowlevel/xc.so
 /lib/udev/rules.d/90-xen-backend.rules
+%{_libdir}/%{name}/bin/convert-legacy-stream
+%{_libdir}/%{name}/bin/legacy.py*
 %{_libdir}/%{name}/bin/libxl-save-helper
 %{_libdir}/%{name}/bin/lsevtchn
 %{_libdir}/%{name}/bin/pygrub
 %{_libdir}/%{name}/bin/readnotes
+%{_libdir}/%{name}/bin/verify-stream-v2
 %{_libdir}/%{name}/bin/xc_restore
 %{_libdir}/%{name}/bin/xc_save
 %{_libdir}/%{name}/bin/xenconsole
