# HG changeset patch
# Parent 059ca801423704be1f4d3b6873a30dcd98fcf129

diff -r 059ca8014237 mk/xen-4.4.spec.in
--- a/mk/xen-4.4.spec.in
+++ b/mk/xen-4.4.spec.in
@@ -460,6 +460,9 @@ rm -rf $RPM_BUILD_ROOT
 %dir /var/lib/xen
 %dir /var/log/xen
 
+%{_libdir}/%{name}/bin/legacy.py*
+%{_libdir}/%{name}/bin/streamspec.py*
+
 %files dom0-libs
 %defattr(-,root,root,-)
 %{_libdir}/fs/ext2fs-lib/fsimage.so
diff -r 059ca8014237 tools/libxc/Makefile
--- a/tools/libxc/Makefile
+++ b/tools/libxc/Makefile
@@ -148,6 +148,10 @@ install: build
 	ln -sf libxenguest.so.$(MAJOR) $(DESTDIR)$(LIBDIR)/libxenguest.so
 	$(INSTALL_DATA) xenguest.h $(DESTDIR)$(INCLUDEDIR)
 
+	$(INSTALL_DIR) $(DESTDIR)$(PRIVATE_BINDIR)
+	$(INSTALL_PROG) saverestore/scripts/legacy.py $(DESTDIR)$(PRIVATE_BINDIR)
+	$(INSTALL_PROG) saverestore/scripts/streamspec.py $(DESTDIR)$(PRIVATE_BINDIR)
+
 .PHONY: TAGS
 TAGS:
 	etags -t *.c *.h
