# HG changeset patch
# Parent 67ca6028e59f6bcd00e54bcfeb85cf6e89fc05e8

diff -r 67ca6028e59f mk/xen-4.3.spec.in
--- a/mk/xen-4.3.spec.in
+++ b/mk/xen-4.3.spec.in
@@ -324,6 +324,10 @@ rm -rf $RPM_BUILD_ROOT
 %{_bindir}/xentrace
 %{_bindir}/xentrace_format
 %{_bindir}/xentrace_setsize
+%{_bindir}/list_domains
+%{_bindir}/raise_exception
+%{_bindir}/send_debug_keys
+%{_bindir}/xtl
 %{python_sitearch}/fsimage.so
 %{python_sitearch}/grub/ExtLinuxConf.py*
 %{python_sitearch}/grub/GrubConf.py*
@@ -442,6 +446,8 @@ rm -rf $RPM_BUILD_ROOT
 %{_libdir}/ocaml/stublibs/dllxeneventchn_stubs.so.owner
 %{_libdir}/ocaml/stublibs/dllxenmmap_stubs.so
 %{_libdir}/ocaml/stublibs/dllxenmmap_stubs.so.owner
+%{_libdir}/ocaml/stublibs/dllxentoollog_stubs.so
+%{_libdir}/ocaml/stublibs/dllxentoollog_stubs.so.owner
 %{_libdir}/ocaml/xenbus/META
 %{_libdir}/ocaml/xenbus/xenbus.cma
 %{_libdir}/ocaml/xenbus/xenbus.cmo
@@ -451,6 +457,8 @@ rm -rf $RPM_BUILD_ROOT
 %{_libdir}/ocaml/xeneventchn/xeneventchn.cma
 %{_libdir}/ocaml/xenmmap/META
 %{_libdir}/ocaml/xenmmap/xenmmap.cma
+%{_libdir}/ocaml/xentoollog/META
+%{_libdir}/ocaml/xentoollog/xentoollog.cma
 
 %files ocaml-devel
 %defattr(-,root,root,-)
@@ -474,6 +482,11 @@ rm -rf $RPM_BUILD_ROOT
 %{_libdir}/ocaml/xenmmap/xenmmap.cmi
 %{_libdir}/ocaml/xenmmap/xenmmap.cmx
 %{_libdir}/ocaml/xenmmap/xenmmap.cmxa
+%{_libdir}/ocaml/xentoollog/libxentoollog_stubs.a
+%{_libdir}/ocaml/xentoollog/xentoollog.a
+%{_libdir}/ocaml/xentoollog/xentoollog.cmi
+%{_libdir}/ocaml/xentoollog/xentoollog.cmx
+%{_libdir}/ocaml/xentoollog/xentoollog.cmxa
 
 %doc
 
