# HG changeset patch
# Parent d56b308d55d0542aa16ccc82a7f1f9add1809579

diff -r d56b308d55d0 mk/xen-4.4.spec.in
--- a/mk/xen-4.4.spec.in
+++ b/mk/xen-4.4.spec.in
@@ -218,6 +218,8 @@ rm -f %{buildroot}/usr/share/doc/xen/REA
 # Copy ocaml xenstored to legacy location until we can fix up the init scripts
 mkdir -p $RPM_BUILD_ROOT/opt/xensource/bin/
 cp -f %{buildroot}%{_sbindir}/oxenstored %{buildroot}/opt/xensource/bin/xenstored
+ln -s convert-legacy-stream %{buildroot}%{_libdir}/%{name}/bin/legacy.py
+
 # Nuke the xenstore and xenlight libraries
 rm -rf %{buildroot}%{_libdir}/ocaml/xenstore
 rm -rf %{buildroot}%{_libdir}/ocaml/xenlight
@@ -408,14 +410,19 @@ rm -rf $RPM_BUILD_ROOT
 %{python_sitearch}/pygrub-0.3-py2.6.egg-info
 %{python_sitearch}/xen-3.0-py2.6.egg-info
 %{python_sitearch}/xen/__init__.py*
+%{python_sitearch}/xen/streamv2/__init__.py*
+%{python_sitearch}/xen/streamv2/format.py*
 %{python_sitearch}/xen/lowlevel/__init__.py*
 %{python_sitearch}/xen/lowlevel/ptsname.so
 %{python_sitearch}/xen/lowlevel/xc.so
 /lib/udev/rules.d/90-xen-backend.rules
+%{_libdir}/%{name}/bin/convert-legacy-stream
+%{_libdir}/%{name}/bin/legacy.py*
 %{_libdir}/%{name}/bin/libxl-save-helper
 %{_libdir}/%{name}/bin/lsevtchn
 %{_libdir}/%{name}/bin/pygrub
 %{_libdir}/%{name}/bin/readnotes
+%{_libdir}/%{name}/bin/verify-stream-v2
 %{_libdir}/%{name}/bin/xc_restore
 %{_libdir}/%{name}/bin/xc_save
 %{_libdir}/%{name}/bin/xenconsole
