# HG changeset patch
# Parent e96ee52dda9a1c4b6cb0745e6d4906986fd44813

diff -r e96ee52dda9a mk/xen-4.4.spec.in
--- a/mk/xen-4.4.spec.in
+++ b/mk/xen-4.4.spec.in
@@ -513,6 +513,10 @@ rm -rf $RPM_BUILD_ROOT
 
 %files ocaml-libs
 %defattr(-,root,root,-)
+%{_libdir}/ocaml/gnt/META
+%{_libdir}/ocaml/gnt/gnt.cma
+%{_libdir}/ocaml/stublibs/dllgnt_stubs.so
+%{_libdir}/ocaml/stublibs/dllgnt_stubs.so.owner
 %{_libdir}/ocaml/stublibs/dllxenbus_stubs.so
 %{_libdir}/ocaml/stublibs/dllxenbus_stubs.so.owner
 %{_libdir}/ocaml/stublibs/dllxenctrl_stubs.so
@@ -544,6 +548,11 @@ rm -rf $RPM_BUILD_ROOT
 
 %files ocaml-devel
 %defattr(-,root,root,-)
+%{_libdir}/ocaml/gnt/libgnt_stubs.a
+%{_libdir}/ocaml/gnt/gnt.a
+%{_libdir}/ocaml/gnt/gnt.cmi
+%{_libdir}/ocaml/gnt/gnt.cmx
+%{_libdir}/ocaml/gnt/gnt.cmxa
 %{_libdir}/ocaml/xenbus/libxenbus_stubs.a
 %{_libdir}/ocaml/xenbus/xenbus.a
 %{_libdir}/ocaml/xenbus/xenbus.cmi
