# HG changeset patch
# Parent 2a57cf881f148310ddcf520df825f60a864ab7e9
diff -r 2a57cf881f14 -r 2561a8027348 tools/libxc/xc_domain_restore.c
--- a/tools/libxc/xc_domain_restore.c
+++ b/tools/libxc/xc_domain_restore.c
@@ -2110,6 +2110,9 @@ int xc_domain_restore(xc_interface *xch,
     free(mmu);
     free(ctx->p2m);
     free(pfn_type);
+    free(region_mfn);
+    free(ctx->p2m_batch);
+    pagebuf_free(&pagebuf);
     tailbuf_free(&tailbuf);
 
     /* discard cache for save file  */
diff -r 2a57cf881f14 -r 2561a8027348 tools/libxc/xc_domain_save.c
--- a/tools/libxc/xc_domain_save.c
+++ b/tools/libxc/xc_domain_save.c
@@ -139,6 +139,12 @@ static int outbuf_init(xc_interface *xch
     return 0;
 }
 
+static int outbuf_free(struct outbuf *ob)
+{
+    free(ob->buf);
+    return 0;
+}
+
 static inline int outbuf_write(xc_interface *xch,
                                struct outbuf* ob, void* buf, size_t len)
 {
@@ -2111,6 +2117,8 @@ int xc_domain_save(xc_interface *xch, in
     free(pfn_batch);
     free(pfn_err);
     free(to_fix);
+    free(hvm_buf);
+    outbuf_free(&ob_pagebuf);
 
     DPRINTF("Save exit rc=%d\n",rc);
 
