# HG changeset patch
# Parent 147f14103170ee98b1cb4519351524c358375ba6

diff -r 147f14103170 tools/firmware/hvmloader/pci.c
--- a/tools/firmware/hvmloader/pci.c
+++ b/tools/firmware/hvmloader/pci.c
@@ -129,7 +129,12 @@ void pci_setup(void)
             {
                 vga_devfn = devfn;
                 virtual_vga = VGA_pt;
-                if ( vendor_id == 0x8086 )
+                if ( vendor_id == 0x10DE )
+                {
+                    vga_devfn = devfn;
+                    virtual_vga = VGA_std;
+                }
+                else if ( vendor_id == 0x8086 )
                 {
                     igd_opregion_pgbase = mem_hole_alloc(IGD_OPREGION_PAGES);
                     /*
