# HG changeset patch
# Parent 6d167fad887b5cf00a76c62a1a8dd080562ae53e
diff -r 6d167fad887b -r 85e2239dfe3f tools/firmware/hvmloader/pci.c
--- a/tools/firmware/hvmloader/pci.c	Thu May 08 14:52:55 2014 +0100
+++ b/tools/firmware/hvmloader/pci.c	Thu May 08 14:56:31 2014 +0100
@@ -125,6 +125,11 @@ void pci_setup(void)
                 vga_devfn = devfn;
                 virtual_vga = VGA_cirrus;
             }
+            else if ( vendor_id == 0x10DE )
+            {
+                vga_devfn = devfn;
+                virtual_vga = VGA_std;
+            }
             else if ( virtual_vga == VGA_none )
             {
                 vga_devfn = devfn;
