# HG changeset patch
# Parent bd3892592d8cde3abe7bf6f3548c0d1bfdde2b8f

diff -r bd3892592d8c tools/firmware/hvmloader/acpi/dsdt.asl
--- a/tools/firmware/hvmloader/acpi/dsdt.asl
+++ b/tools/firmware/hvmloader/acpi/dsdt.asl
@@ -30,8 +30,8 @@ DefinitionBlock ("DSDT.aml", "DSDT", 2, 
     /* _S3 and _S4 are in separate SSDTs */
     Name (\_S5, Package (0x04)
     {
-        0x00,  /* PM1a_CNT.SLP_TYP */
-        0x00,  /* PM1b_CNT.SLP_TYP */
+        0x07,  /* PM1a_CNT.SLP_TYP */
+        0x07,  /* PM1b_CNT.SLP_TYP */
         0x00,  /* reserved */
         0x00   /* reserved */
     })
diff -r bd3892592d8c tools/firmware/hvmloader/acpi/ssdt_s3.asl
--- a/tools/firmware/hvmloader/acpi/ssdt_s3.asl
+++ b/tools/firmware/hvmloader/acpi/ssdt_s3.asl
@@ -23,8 +23,8 @@ DefinitionBlock ("SSDT_S3.aml", "SSDT", 
     /* Must match piix emulation */
     Name (\_S3, Package (0x04)
     {
-        0x01,  /* PM1a_CNT.SLP_TYP */
-        0x01,  /* PM1b_CNT.SLP_TYP */
+        0x05,  /* PM1a_CNT.SLP_TYP */
+        0x05,  /* PM1b_CNT.SLP_TYP */
         0x0,   /* reserved */
         0x0    /* reserved */
     })
diff -r bd3892592d8c tools/firmware/hvmloader/acpi/ssdt_s4.asl
--- a/tools/firmware/hvmloader/acpi/ssdt_s4.asl
+++ b/tools/firmware/hvmloader/acpi/ssdt_s4.asl
@@ -23,8 +23,8 @@ DefinitionBlock ("SSDT_S4.aml", "SSDT", 
     /* Must match piix emulation */
     Name (\_S4, Package (0x04)
     {
-        0x00,  /* PM1a_CNT.SLP_TYP */
-        0x00,  /* PM1b_CNT.SLP_TYP */
+        0x06,  /* PM1a_CNT.SLP_TYP */
+        0x06,  /* PM1b_CNT.SLP_TYP */
         0x00,  /* reserved */
         0x00   /* reserved */
     })
diff -r bd3892592d8c tools/firmware/hvmloader/hvmloader.c
--- a/tools/firmware/hvmloader/hvmloader.c
+++ b/tools/firmware/hvmloader/hvmloader.c
@@ -325,7 +325,7 @@ int main(void)
         struct xen_hvm_param p = {
             .domid = DOMID_SELF,
             .index = HVM_PARAM_ACPI_IOPORTS_LOCATION,
-            .value = 1,
+            .value = 0,
         };
 
         if ( bios->acpi_build_tables )
diff -r bd3892592d8c xen/arch/x86/hvm/hvm.c
--- a/xen/arch/x86/hvm/hvm.c
+++ b/xen/arch/x86/hvm/hvm.c
@@ -3910,7 +3910,7 @@ long do_hvm_op(unsigned long op, XEN_GUE
 
                 break;
             case HVM_PARAM_ACPI_IOPORTS_LOCATION:
-                rc = pmtimer_change_ioport(d, a.value);
+                rc = pmtimer_change_ioport(d, 0);
                 break;
             case HVM_PARAM_MEMORY_EVENT_CR0:
             case HVM_PARAM_MEMORY_EVENT_CR3:
diff -r bd3892592d8c xen/include/public/hvm/ioreq.h
--- a/xen/include/public/hvm/ioreq.h
+++ b/xen/include/public/hvm/ioreq.h
@@ -95,11 +95,11 @@ typedef struct buffered_iopage buffered_
 #define ACPI_GPE0_BLK_LEN_V0         0x08
 
 /* Version 1: Locations preferred by modern Qemu. */
-#define ACPI_PM1A_EVT_BLK_ADDRESS_V1 0xb000
-#define ACPI_PM1A_CNT_BLK_ADDRESS_V1 (ACPI_PM1A_EVT_BLK_ADDRESS_V1 + 0x04)
-#define ACPI_PM_TMR_BLK_ADDRESS_V1   (ACPI_PM1A_EVT_BLK_ADDRESS_V1 + 0x08)
-#define ACPI_GPE0_BLK_ADDRESS_V1     0xafe0
-#define ACPI_GPE0_BLK_LEN_V1         0x04
+#define ACPI_PM1A_EVT_BLK_ADDRESS_V1 0x1f40
+#define ACPI_PM1A_CNT_BLK_ADDRESS_V1 (ACPI_PM1A_EVT_BLK_ADDRESS_V0 + 0x04)
+#define ACPI_PM_TMR_BLK_ADDRESS_V1   (ACPI_PM1A_EVT_BLK_ADDRESS_V0 + 0x08)
+#define ACPI_GPE0_BLK_ADDRESS_V1     (ACPI_PM_TMR_BLK_ADDRESS_V0 + 0x20)
+#define ACPI_GPE0_BLK_LEN_V1         0x08
 
 /* Compatibility definitions for the default location (version 0). */
 #define ACPI_PM1A_EVT_BLK_ADDRESS    ACPI_PM1A_EVT_BLK_ADDRESS_V0
