# HG changeset patch
# Parent 1cb586efc70a7fb304561ebe690164bde7a6010c

diff -r 1cb586efc70a tools/pygrub/src/pygrub
--- a/tools/pygrub/src/pygrub	Tue Aug 12 11:14:22 2014 +0100
+++ b/tools/pygrub/src/pygrub	Tue Aug 12 14:59:18 2014 +0100
@@ -987,6 +987,11 @@ if __name__ == "__main__":
     if list_entries:
         sys.exit(0)
 
+    # XenMobile controller has a broken bootloader config, fix it up
+    if chosencfg["kernel"] and not fs.file_exists(chosencfg["kernel"]) \
+            and fs.file_exists("/kernel-xenserver"):
+        chosencfg["kernel"] = "/kernel-xenserver"
+
     # Did looping through partitions find us a kernel?
     if fs is None:
         raise RuntimeError, "Unable to find partition containing kernel"
