# HG changeset patch
# Parent 4bef01f20c0a2516d94e5e5bd78c1e5ccf6a52e5

diff -r 4bef01f20c0a tools/pygrub/src/pygrub
--- a/tools/pygrub/src/pygrub
+++ b/tools/pygrub/src/pygrub
@@ -697,7 +697,7 @@ if __name__ == "__main__":
     sel = None
     
     def usage():
-        print >> sys.stderr, "Usage: %s [-q|--quiet] [-i|--interactive] [-n|--not-really] [--output=] [--kernel=] [--ramdisk=] [--args=] [--entry=] [--output-directory=] [--output-format=sxp|simple|simple0] <image>" %(sys.argv[0],)
+        print >> sys.stderr, "Usage: %s [-q|--quiet] [-i|--interactive] [-n|--not-really] [--output=] [--kernel=] [--ramdisk=] [--args=] [--entry=] [--output-directory=] [--output-format=sxp|simple|simple0] [--default_args=] [--extra_args=] <image>" %(sys.argv[0],)
 
     def copy_from_image(fs, file_to_read, file_type, output_directory,
                         not_really):
@@ -734,8 +734,8 @@ if __name__ == "__main__":
         opts, args = getopt.gnu_getopt(sys.argv[1:], 'qinh::',
                                    ["quiet", "interactive", "not-really", "help", 
                                     "output=", "output-format=", "output-directory=",
-                                    "entry=", "kernel=", 
-                                    "ramdisk=", "args=", "isconfig", "debug"])
+                                    "entry=", "kernel=", "default_args=", "extra_args=",
+                                    "ramdisk=", "args=", "isconfig", "debug", "vm="])
     except getopt.GetoptError:
         usage()
         sys.exit(1)
@@ -753,6 +753,8 @@ if __name__ == "__main__":
     not_really = False
     output_format = "sxp"
     output_directory = "/var/run/xend/boot"
+    default_args = ""
+    extra_args = ""
 
     # what was passed in
     incfg = { "kernel": None, "ramdisk": None, "args": "" }
@@ -795,6 +797,10 @@ if __name__ == "__main__":
             output_format = a
         elif o in ("--output-directory",):
             output_directory = a
+        elif o in ("--default_args",):
+            default_args = a
+        elif o in ("--extra_args",):
+            extra_args = a
 
     if debug:
 	logging.basicConfig(level=logging.DEBUG)
@@ -854,6 +860,13 @@ if __name__ == "__main__":
     if not fs:
         raise RuntimeError, "Unable to find partition containing kernel"
 
+    # Append any extra arguments we were given
+    if extra_args:
+        if chosencfg["args"] == None:
+            chosencfg["args"] = extra_args
+        else:
+            chosencfg["args"] += " " + extra_args
+
     bootcfg["kernel"] = copy_from_image(fs, chosencfg["kernel"], "kernel",
                                         output_directory, not_really)
 
