# HG changeset patch
# Parent 7b2aa5592846c2a3ffbc1e20f051914500237b4d

diff -r 7b2aa5592846 tools/pygrub/src/pygrub
--- a/tools/pygrub/src/pygrub	Wed Sep 05 14:51:03 2012 +0100
+++ b/tools/pygrub/src/pygrub	Tue Sep 11 15:47:46 2012 +0100
@@ -897,7 +897,15 @@ if __name__ == "__main__":
     if len(part_offs) < 1:
         raise RuntimeError, "Disk has no partitions"
 
-    fs = fsimage.open(file, part_offs[0], bootfsoptions)
+    try:
+        fs = fsimage.open(file, part_offs[0], bootfsoptions)
+    except IOError:
+        # in this case we tried to detect a disk with a direct filesystem
+        # but probably this disk does not have a valid partition
+        if len(part_offs) == 1 and part_offs[0] == 0:
+            raise RuntimeError, "Bootable disk has no valid filesystem"
+        else:
+            raise
 
     # We always boot the "default" kernel if it exists, rather than
     # parsing the grub menu
