# HG changeset patch
# Parent 29be9b678ed259e4e273c5607c54eac906d03ec1

diff -r 29be9b678ed2 xen/arch/x86/time.c
--- a/xen/arch/x86/time.c
+++ b/xen/arch/x86/time.c
@@ -945,6 +945,7 @@ void domain_set_time_offset(struct domai
     d->time_offset_seconds = time_offset_seconds;
     if ( is_hvm_domain(d) )
         rtc_update_clock(d);
+    update_domain_wallclock_time(d);
 }
 
 int cpu_frequency_change(u64 freq)
