# HG changeset patch
# Parent 020a88bc27ca263f3d2ec244c963593a4e39b06e

diff -r 020a88bc27ca mk/xen-4.4.spec.in
--- a/mk/xen-4.4.spec.in	Wed Apr 09 09:13:35 2014 +0100
+++ b/mk/xen-4.4.spec.in	Wed Apr 09 09:51:40 2014 +0100
@@ -598,4 +598,8 @@ if [ -e %{_sysconfdir}/sysconfig/kernel 
   cat %{_sysconfdir}/sysconfig/kernel-xen >> %{_sysconfdir}/sysconfig/kernel
 fi
 
+# HACK - upstream udev rule interacts with the one in xenopsd
+%triggerin -n xen-dom0-tools -- xenopsd
+rm -f /lib/udev/rules.d/90-xen-backend.rules
+
 %changelog
