# HG changeset patch
# Parent 9ca75f4c84a8b2204fc80339c4b8e45ba4eda153

diff --git a/xen/arch/x86/nmi.c b/xen/arch/x86/nmi.c
--- a/xen/arch/x86/nmi.c
+++ b/xen/arch/x86/nmi.c
@@ -49,18 +49,10 @@
 
 static void __init parse_watchdog(char *s)
 {
-    switch ( parse_bool(s) )
-    {
-    case 0:
-        opt_watchdog = 0;
-        return;
-    case 1:
-        opt_watchdog = 1;
-        return;
-    }
+    opt_watchdog = !!parse_bool(s);
 
     if ( !strcmp(s, "force") )
-        watchdog_force = opt_watchdog = 1;
+        watchdog_force = 1;
 }
 custom_param("watchdog", parse_watchdog);
 
