diff -r 58cee359cbc7 tools/xcutils/xc_restore.c
--- a/tools/xcutils/xc_restore.c	Wed May 21 11:44:20 2014 +0100
+++ b/tools/xcutils/xc_restore.c	Wed May 21 14:23:59 2014 +0100
@@ -56,7 +56,7 @@ main(int argc, char **argv)
 
     ret = xc_domain_restore(xch, io_fd, domid, store_evtchn, &store_mfn, 0,
                             console_evtchn, &console_mfn, 0, hvm, pae, superpages,
-                            0, checkpointed, NULL, NULL);
+                            checkpointed, NULL);
 
     if ( ret == 0 )
     {
diff -r 58cee359cbc7 tools/xcutils/xc_save.c
--- a/tools/xcutils/xc_save.c	Wed May 21 11:44:20 2014 +0100
+++ b/tools/xcutils/xc_save.c	Wed May 21 14:23:59 2014 +0100
@@ -213,7 +213,7 @@ main(int argc, char **argv)
     callbacks.suspend = suspend;
     callbacks.switch_qemu_logdirty = switch_qemu_logdirty;
     ret = xc_domain_save(si.xch, io_fd, si.domid, maxit, max_f, si.flags, 
-                         &callbacks, !!(si.flags & XCFLAGS_HVM), 0);
+                         &callbacks, !!(si.flags & XCFLAGS_HVM));
 
     if (si.suspend_evtchn > 0)
 	 xc_suspend_evtchn_release(si.xch, si.xce, si.domid, si.suspend_evtchn);
