# HG changeset patch
# Parent 8bc223029d4b0878439422e2113c7484211bc777

diff -r 8bc223029d4b xen/arch/x86/traps.c
--- a/xen/arch/x86/traps.c
+++ b/xen/arch/x86/traps.c
@@ -195,10 +195,11 @@ static void show_guest_stack(struct vcpu
 static void show_trace(struct cpu_user_regs *regs)
 {
     unsigned long *stack = ESP_BEFORE_EXCEPTION(regs), addr;
+    unsigned stack_pos = 0;
 
     printk("Xen call trace:\n   ");
 
-    printk("[<%p>]", _p(regs->eip));
+    printk("   [<%p>]", _p(regs->eip));
     print_symbol(" %s\n   ", regs->eip);
 
     while ( ((long)stack & (STACK_SIZE-BYTES_PER_LONG)) != 0 )
@@ -206,9 +207,10 @@ static void show_trace(struct cpu_user_r
         addr = *stack++;
         if ( is_active_kernel_text(addr) )
         {
-            printk("[<%p>]", _p(addr));
+            printk("%3u[<%p>]", stack_pos, _p(addr));
             print_symbol(" %s\n   ", addr);
         }
+        ++stack_pos;
     }
 
     printk("\n");
