# HG changeset patch
# Parent 36c5249e37e9462efe968e6a97a6cf38bd6d4b80

diff -r 36c5249e37e9 xen/arch/x86/traps.c
--- a/xen/arch/x86/traps.c	Mon Aug 06 09:03:37 2012 +0100
+++ b/xen/arch/x86/traps.c	Mon Aug 06 09:06:00 2012 +0100
@@ -199,10 +199,11 @@ static void show_guest_stack(struct vcpu
 static void show_trace(struct cpu_user_regs *regs)
 {
     unsigned long *stack = ESP_BEFORE_EXCEPTION(regs), addr;
+    unsigned stack_pos = 0;
 
     printk("Xen call trace:\n   ");
 
-    printk("[<%p>]", _p(regs->eip));
+    printk("   [<%p>]", _p(regs->eip));
     print_symbol(" %s\n   ", regs->eip);
 
     while ( ((long)stack & (STACK_SIZE-BYTES_PER_LONG)) != 0 )
@@ -210,9 +211,10 @@ static void show_trace(struct cpu_user_r
         addr = *stack++;
         if ( is_kernel_text(addr) || is_kernel_inittext(addr) )
         {
-            printk("[<%p>]", _p(addr));
+            printk("%3u[<%p>]", stack_pos, _p(addr));
             print_symbol(" %s\n   ", addr);
         }
+        ++stack_pos;
     }
 
     printk("\n");
