# HG changeset patch
# Parent 0bff3bcbec6c7f5d63596dc8e739b284e88055d5

diff -r 0bff3bcbec6c xen/arch/x86/hvm/hvm.c
--- a/xen/arch/x86/hvm/hvm.c
+++ b/xen/arch/x86/hvm/hvm.c
@@ -4456,6 +4456,10 @@ long do_hvm_op(unsigned long op, XEN_GUE
         gdprintk(XENLOG_INFO, "Legacy windows driver quirks enabled.\n");
         break;
 
+    case HVMOP_process_dying:
+        /* Reduced to functional NOP */
+        break;
+
     default:
     {
         gdprintk(XENLOG_DEBUG, "Bad HVM op %ld.\n", op);
diff -r 0bff3bcbec6c xen/include/public/hvm/hvm_op.h
--- a/xen/include/public/hvm/hvm_op.h
+++ b/xen/include/public/hvm/hvm_op.h
@@ -272,6 +272,9 @@ DEFINE_XEN_GUEST_HANDLE(xen_hvm_inject_m
 
 #endif /* defined(__XEN__) || defined(__XEN_TOOLS__) */
 
+/* Legacy NOP: Hint from PV drivers for process destruction. */
+#define HVMOP_process_dying        0x102
+
 #define HVMOP_set_driver_version 0x103
 struct xen_hvm_set_driver_version {
     uint32_t build;
