Allow 32-bit PV drivers to work on 64-bit Xen.

diff -r 6c4caa7f08ec xen/arch/x86/hvm/hvm.c
--- a/xen/arch/x86/hvm/hvm.c
+++ b/xen/arch/x86/hvm/hvm.c
@@ -3973,6 +3973,20 @@ long do_hvm_op(unsigned long op, XEN_GUE
                 if ( a.value > 1 )
                     rc = -EINVAL;
                 break;
+            case HVM_PARAM_32BIT:
+                /* Not used any more, as this has been fixed correctly.
+                 *
+                 * It was introduced originally to hack around the issue of
+                 * the drivers incorrectly not rewriting the hypercall page on
+                 * S4resume.
+                 *
+                 * "Legacy" windows drivers unconditionally make this
+                 * hypercall and ignore the return value.
+                 *
+                 * The Tampa and Clearwater non-legacy drivers still make this
+                 * hypercall, but will BSOD with 0x0000DEAD if the call fails.
+                 */
+                break;
             }
 
             if ( rc == 0 ) 
diff -r 6c4caa7f08ec xen/include/public/hvm/params.h
--- a/xen/include/public/hvm/params.h
+++ b/xen/include/public/hvm/params.h
@@ -56,6 +56,10 @@
 
 #if defined(__i386__) || defined(__x86_64__)
 
+/* The guest is using 32 bit PV drivers.  Apply suitable translations
+   to shared info and hypercall arguments. */
+#define HVM_PARAM_32BIT        8
+
 /* Expose Viridian interfaces to this HVM guest? */
 #define HVM_PARAM_VIRIDIAN     9
 
