# HG changeset patch
# Parent ba298e2f16e4da9116846da843daa335b36315a7

diff -r ba298e2f16e4 xen/common/kexec.c
--- a/xen/common/kexec.c
+++ b/xen/common/kexec.c
@@ -66,7 +66,7 @@ static struct {
 
 /* Low crashinfo mode.  Start as INVALID so serveral codepaths can set up
  * defaults without needing to know the state of the others. */
-enum low_crashinfo low_crashinfo_mode = LOW_CRASHINFO_INVALID;
+enum low_crashinfo low_crashinfo_mode = LOW_CRASHINFO_MIN;
 
 /* This value is only considered if low_crash_mode is set to MIN or ALL, so
  * setting a default here is safe. Default to 4GB.  This is because the current
diff -r ba298e2f16e4 xen/drivers/char/console.c
--- a/xen/drivers/char/console.c
+++ b/xen/drivers/char/console.c
@@ -47,14 +47,14 @@ static bool_t __read_mostly opt_console_
 boolean_param("console_to_ring", opt_console_to_ring);
 
 /* console_timestamps: include a timestamp prefix on every Xen console line. */
-static bool_t __read_mostly opt_console_timestamps;
+static bool_t __read_mostly opt_console_timestamps = 1;
 boolean_param("console_timestamps", opt_console_timestamps);
 
-/* conring_size: allows a large console ring than default (16kB). */
+/* conring_size: allows a large console ring than default (64kB). */
 static uint32_t __initdata opt_conring_size;
 size_param("conring_size", opt_conring_size);
 
-#define _CONRING_SIZE 16384
+#define _CONRING_SIZE 65536
 #define CONRING_IDX_MASK(i) ((i)&(conring_size-1))
 static char __initdata _conring[_CONRING_SIZE];
 static char *__read_mostly conring = _conring;
