# HG changeset patch
# Parent 0896f518c628b2dcd0261a6cd8d29caec517f266

diff -r 0896f518c628 xen/arch/x86/cpu/amd.c
--- a/xen/arch/x86/cpu/amd.c
+++ b/xen/arch/x86/cpu/amd.c
@@ -440,7 +440,7 @@ static void __devinit amd_get_topology(s
 static void __devinit init_amd(struct cpuinfo_x86 *c)
 {
 	u32 l, h;
-
+	static bool_t automatic_fma4_hiding = 0;
 	unsigned long long value;
 
 	/* Disable TLB flush filter by setting HWCR.FFDIS on K8
@@ -632,8 +632,32 @@ static void __devinit init_amd(struct cp
 	if ((smp_processor_id() == 1) && c1_ramping_may_cause_clock_drift(c))
 		disable_c1_ramping();
 
+	/* Bugfix for Ubuntu 12.04 installer.  XSAVE is not enabled by default
+	 * on XenSever, but glibc in 12.04 incorrectly tried to AVX
+	 * instructions, and dies with an illegal opcode exception.  If the
+	 * user has not explicitly requested XSAVE support, mask out FMA4 to
+	 * prevent glibc from falling over. */
+	if ( !use_xsave && (c == &boot_cpu_data) &&
+	     (c->x86 == 0x15) && (c->x86_model <= 0x02) &&
+	     !(opt_cpuid_mask_ext_ecx & cpufeat_mask(X86_FEATURE_FMA4)) )
+	{
+		printk(KERN_DEBUG "CPU: Masking FMA4 CPUID feature\n");
+		automatic_fma4_hiding = 1;
+		opt_cpuid_mask_ext_ecx &= ~cpufeat_mask(X86_FEATURE_FMA4);
+	}
+
 	set_cpuidmask(c);
 
+	/*
+	 * Put FMA4 feature back into mask variables IFF it was not masked on
+	 * command line, so XAPI allows Rolling Pool Upgrade to work
+	 */
+	if ( automatic_fma4_hiding )
+	{
+		c->boot_ext_ecx   |= cpufeat_mask(X86_FEATURE_FMA4);
+		c->masked_ext_ecx |= cpufeat_mask(X86_FEATURE_FMA4);
+	}
+
 	check_syscfg_dram_mod_en();
 }
 
diff -r 0896f518c628 xen/arch/x86/cpu/common.c
--- a/xen/arch/x86/cpu/common.c
+++ b/xen/arch/x86/cpu/common.c
@@ -15,7 +15,7 @@
 
 #include "cpu.h"
 
-static bool_t __cpuinitdata use_xsave = 0;
+bool_t __cpuinitdata use_xsave = 0;
 boolean_param("xsave", use_xsave);
 
 bool_t __devinitdata opt_arat = 1;
diff -r 0896f518c628 xen/arch/x86/cpu/cpu.h
--- a/xen/arch/x86/cpu/cpu.h
+++ b/xen/arch/x86/cpu/cpu.h
@@ -15,6 +15,7 @@ extern bool_t opt_arat;
 extern unsigned int opt_cpuid_mask_ecx, opt_cpuid_mask_edx;
 extern unsigned int opt_cpuid_mask_xsave_eax;
 extern unsigned int opt_cpuid_mask_ext_ecx, opt_cpuid_mask_ext_edx;
+extern bool_t use_xsave;
 
 extern int get_model_name(struct cpuinfo_x86 *c);
 extern void display_cacheinfo(struct cpuinfo_x86 *c);
