diff -r fe72c4d2ada2 xen/arch/x86/hvm/hvm.c
--- a/xen/arch/x86/hvm/hvm.c	Fri Feb 18 17:37:02 2011 +0000
+++ b/xen/arch/x86/hvm/hvm.c	Fri Feb 18 17:37:52 2011 +0000
@@ -3374,6 +3374,10 @@
         {
             switch ( a.index )
             {
+            case HVM_PARAM_CONSOLE_PFN:
+            case HVM_PARAM_CONSOLE_EVTCHN:
+                rc=-EINVAL;
+                goto param_fail;
             case HVM_PARAM_ACPI_S_STATE:
                 a.value = d->arch.hvm_domain.is_s3_suspended ? 3 : 0;
                 break;
