# HG changeset patch
# Parent 658dabc66c636863cd43bcc35781e648c5c53322

diff -r 658dabc66c63 tools/xenguest/xenguest.c
--- a/tools/xenguest/xenguest.c
+++ b/tools/xenguest/xenguest.c
@@ -492,9 +492,6 @@ int main(int argc, char * const argv[])
     xc_interface *xch;
     xs_handle * xsh;
 
-    /* Force migration v2 all the time. */
-    setenv("XG_MIGRATION_V2", "", 1);
-
     {   /* Conjoin the command line into a single string for logging */
         size_t sum, s;
         int i;
