# HG changeset patch
# Parent 8eee47ebf75403f838e8060d12ae369d53f313a0

diff --git a/tools/ocaml/xenstored/store.ml b/tools/ocaml/xenstored/store.ml
--- a/tools/ocaml/xenstored/store.ml
+++ b/tools/ocaml/xenstored/store.ml
@@ -379,7 +379,8 @@ let set_node store path node orig_quota 
 
 let write store perm path value =
 	let owner = Node.get_owner (get_deepest_existing_node store path) in
-	Quota.check store.quota owner (String.length value);
+	if not (path_exists store path) then
+		Quota.check store.quota owner (String.length value);
 	let root, node_created = path_write store perm path value in
 	store.root <- root;
 	if node_created
