# HG changeset patch
# Parent 34dd7b26a4ab723406b4fc3bcb94406a398219b4

diff -r 34dd7b26a4ab tools/xenstore/compat/xs.h
--- a/tools/xenstore/compat/xs.h
+++ b/tools/xenstore/compat/xs.h
@@ -1,2 +1,1 @@
-#warning xs.h is deprecated use xenstore.h instead
 #include <xenstore.h>
diff -r 34dd7b26a4ab tools/xenstore/compat/xs_lib.h
--- a/tools/xenstore/compat/xs_lib.h
+++ b/tools/xenstore/compat/xs_lib.h
@@ -1,2 +1,1 @@
-#warning xs_lib.h is deprecated use xenstore_lib.h instead
 #include <xenstore_lib.h>
