From d1d9ca85c856edd84ed57eb3f0e38d40d162c91f Mon Sep 17 00:00:00 2001
From: Sergey Dyasli <sergey.dyasli@citrix.com>
Date: Tue, 9 Apr 2019 15:21:12 +0100
Subject: [PATCH] microcode: remove panic() calls

Signed-off-by: Sergey Dyasli <sergey.dyasli@citrix.com>
diff --git a/xen/arch/x86/microcode.c b/xen/arch/x86/microcode.c
index 73dfdc4c42..e35a999026 100644
--- a/xen/arch/x86/microcode.c
+++ b/xen/arch/x86/microcode.c
@@ -360,8 +360,9 @@ static int do_microcode_update(void *unused)
         {
             if ( atomic_read(&cpu_out) > finished )
                 continue;
-            panic("Timeout during sequential microcode update (finished %d/%d)",
-                  finished, cpu_nr);
+            printk("Timeout during sequential microcode update "
+                   "(finished %d/%d)", finished, cpu_nr);
+            return -EBUSY;
         }
     }
 
@@ -394,8 +395,9 @@ static int do_microcode_update(void *unused)
                            MICROCODE_UPDATE_TIMEOUT_US) && !error )
         {
             error = true;
-            panic("Timeout when finishing updating microcode (finished %d/%d)",
-                  finished, cpu_nr);
+            printk("Timeout when finishing updating microcode (finished %d/%d)",
+                   finished, cpu_nr);
+            return -EBUSY;
         }
 
         finished = atomic_read(&cpu_out);
