From 79bf6dc46d8751956b82f372efdf51655c4fe39a Mon Sep 17 00:00:00 2001
From: Sergey Dyasli <sergey.dyasli@citrix.com>
Date: Wed, 7 Feb 2018 11:29:22 +0000
Subject: [PATCH] XPTI: show status during boot

Signed-off-by: Sergey Dyasli <sergey.dyasli@citrix.com>
diff --git a/xen/arch/x86/smpboot.c b/xen/arch/x86/smpboot.c
index d0ebb0eecc..82f1c98574 100644
--- a/xen/arch/x86/smpboot.c
+++ b/xen/arch/x86/smpboot.c
@@ -1038,6 +1038,10 @@ void __init smp_prepare_cpus(unsigned int max_cpus)
     if ( opt_xpti < 0 )
         opt_xpti = boot_cpu_data.x86_vendor != X86_VENDOR_AMD;
 
+    printk("XPTI status: %s for Dom0, %s for all DomU PV domains\n",
+            opt_xpti && opt_xpti_dom0 ? "enabled" : "disabled",
+            opt_xpti ? "enabled" : "disabled");
+
     rc = setup_cpu_root_pgt(0);
     if ( rc )
         panic("Error %d setting up PV root page table\n", rc);
