From c90c3015f535ddfc6ec8723ffe0ce867677e1322 Mon Sep 17 00:00:00 2001
From: Sergey Dyasli <sergey.dyasli@citrix.com>
Date: Thu, 5 Apr 2018 14:30:59 +0100
Subject: [PATCH] Revert "x86/HVM: don't give the wrong impression of WRMSR
 succeeding"

This reverts commit b56a0cdeaf91347a2c738dc5dd0350420e0e7201.
diff --git a/xen/arch/x86/hvm/svm/svm.c b/xen/arch/x86/hvm/svm/svm.c
index 645106d5b7..e048137fbf 100644
--- a/xen/arch/x86/hvm/svm/svm.c
+++ b/xen/arch/x86/hvm/svm/svm.c
@@ -2003,13 +2003,6 @@ static int svm_msr_write_intercept(unsigned int msr, uint64_t msr_content)
             result = X86EMUL_RETRY;
             break;
         case 0:
-            /*
-             * Match up with the RDMSR side for now; ultimately this entire
-             * case block should go away.
-             */
-            if ( rdmsr_safe(msr, msr_content) == 0 )
-                break;
-            goto gpf;
         case 1:
             break;
         default:
diff --git a/xen/arch/x86/hvm/vmx/vmx.c b/xen/arch/x86/hvm/vmx/vmx.c
index 8f10cff203..e5eacd917e 100644
--- a/xen/arch/x86/hvm/vmx/vmx.c
+++ b/xen/arch/x86/hvm/vmx/vmx.c
@@ -3157,13 +3157,6 @@ static int vmx_msr_write_intercept(unsigned int msr, uint64_t msr_content)
                     case -ERESTART:
                         return X86EMUL_RETRY;
                     case 0:
-                        /*
-                         * Match up with the RDMSR side for now; ultimately this
-                         * entire case block should go away.
-                         */
-                        if ( rdmsr_safe(msr, msr_content) == 0 )
-                            break;
-                        goto gp_fault;
                     case 1:
                         break;
                     default:
