diff --git a/xen/common/memory.c b/xen/common/memory.c
index eea4103..875255a 100644
--- a/xen/common/memory.c
+++ b/xen/common/memory.c
@@ -832,7 +832,15 @@ long do_memory_op(unsigned long cmd, XEN_GUEST_HANDLE_PARAM(void) arg)
 
         if ( op == XENMEM_populate_physmap
              && (reservation.mem_flags & XENMEMF_populate_on_demand) )
+        {
+            if ( !is_hardware_domain(current->domain) )
+            {
+                rcu_unlock_domain(d);
+                return start_extent;
+            }
+
             args.memflags |= MEMF_populate_on_demand;
+        }
 
         if ( xsm_memory_adjust_reservation(XSM_TARGET, current->domain, d) )
         {
