From 06fd2b89a0d5dda39f304e1b420bc452cc1d6854 Mon Sep 17 00:00:00 2001
From: Sergey Dyasli <sergey.dyasli@citrix.com>
Date: Thu, 31 Jan 2019 09:58:23 +0000
Subject: [PATCH] Disable invpcid support

Signed-off-by: Sergey Dyasli <sergey.dyasli@citrix.com>
diff --git a/xen/arch/x86/setup.c b/xen/arch/x86/setup.c
index 43d9911b20..d84fcdb675 100644
--- a/xen/arch/x86/setup.c
+++ b/xen/arch/x86/setup.c
@@ -66,7 +66,7 @@ int8_t __read_mostly opt_smt = -1;
 boolean_param("smt", opt_smt);
 
 /* opt_invpcid: If false, don't use INVPCID instruction even if available. */
-static bool __initdata opt_invpcid = true;
+static bool __initdata opt_invpcid = false;
 boolean_param("invpcid", opt_invpcid);
 bool __read_mostly use_invpcid;
 
