From 57c3f2b89d0453aa2f81c17b0a31c182428e026d Mon Sep 17 00:00:00 2001
From: Sergey Dyasli <sergey.dyasli@citrix.com>
Date: Tue, 26 Mar 2019 11:20:10 +0000
Subject: [PATCH 2/2] x86: add xen-spec-ctrl utility

As a wrapper around XEN_SYSCTL_spec_ctrl

Signed-off-by: Sergey Dyasli <sergey.dyasli@citrix.com>
diff --git a/tools/misc/Makefile b/tools/misc/Makefile
index 63947bf..e914c91 100644
--- a/tools/misc/Makefile
+++ b/tools/misc/Makefile
@@ -23,6 +23,7 @@ INSTALL_SBIN-$(CONFIG_X86)     += xen-hvmctx
 INSTALL_SBIN-$(CONFIG_X86)     += xen-lowmemd
 INSTALL_SBIN-$(CONFIG_X86)     += xen-mfndump
 INSTALL_SBIN-$(CONFIG_X86)     += xen-ucode
+INSTALL_SBIN-$(CONFIG_X86)     += xen-spec-ctrl
 INSTALL_SBIN                   += xencov
 INSTALL_SBIN                   += xenlockprof
 INSTALL_SBIN                   += xenperf
@@ -117,4 +118,7 @@ xencov: xencov.o
 xen-ucode: xen-ucode.o
 	$(CC) $(LDFLAGS) -o $@ $< $(LDLIBS_libxenctrl) $(APPEND_LDFLAGS)
 
+xen-spec-ctrl: xen-spec-ctrl.o
+	$(CC) $(LDFLAGS) -o $@ $< $(LDLIBS_libxenctrl) $(APPEND_LDFLAGS)
+
 -include $(DEPS_INCLUDE)
diff --git a/tools/misc/xen-spec-ctrl.c b/tools/misc/xen-spec-ctrl.c
new file mode 100644
index 0000000..86489b7
--- /dev/null
+++ b/tools/misc/xen-spec-ctrl.c
@@ -0,0 +1,63 @@
+#define _GNU_SOURCE
+
+#include <stdio.h>
+#include <err.h>
+#include <errno.h>
+#include <string.h>
+#include <xenctrl.h>
+
+void show_help(void)
+{
+    fprintf(stderr,
+            "xen-spec-ctrl: Xen speculation control tool\n"
+            "Usage: xen-spec-ctrl update\n");
+}
+
+int main(int argc, char *argv[])
+{
+    int ret;
+    struct xen_sysctl sysctl = {0};
+    xc_interface *xch;
+    char *str = "Mitigations have been updated! "
+                "Check xen-cpuid output for the details.";
+
+    if ( argc < 2 || strcmp(argv[1], "update") != 0 )
+    {
+        show_help();
+        return 1;
+    }
+
+    xch = xc_interface_open(NULL, NULL, 0);
+    if ( xch == NULL )
+        err(1, "xc_interface_open");
+
+    sysctl.interface_version = XEN_SYSCTL_INTERFACE_VERSION;
+    sysctl.cmd = XEN_SYSCTL_spec_ctrl;
+    sysctl.u.spec_ctrl.op = XENPF_spec_ctrl_update;
+
+    ret = xc_sysctl(xch, &sysctl);
+    if ( ret != 0 )
+    {
+        switch (errno)
+        {
+            case ENOEXEC:
+                str = "No new H/W features have been found. "
+                      "Did you forget to update the microcode with xen-ucode?";
+                break;
+
+            case EFAULT:
+                str = "Some error has occured!";
+                break;
+
+            default:
+                str = strerror(errno);
+                break;
+        }
+    }
+
+    fprintf(stderr, "Status: %s\n", str);
+
+    xc_interface_close(xch);
+
+    return 0;
+}
