From 417b793fd58cd75e42b2e3f0eba64a660dbfee32 Mon Sep 17 00:00:00 2001
From: Sergey Dyasli <sergey.dyasli@citrix.com>
Date: Wed, 20 Mar 2019 15:59:03 +0000
Subject: [PATCH 7/8] microcode: update raw cpuid policy after a successful
 ucode update

To be upstreamed.

Signed-off-by: Sergey Dyasli <sergey.dyasli@citrix.com>
diff --git a/xen/arch/x86/cpuid.c b/xen/arch/x86/cpuid.c
index b40dd4d163..c143da4cc3 100644
--- a/xen/arch/x86/cpuid.c
+++ b/xen/arch/x86/cpuid.c
@@ -270,7 +270,7 @@ static void recalculate_misc(struct cpuid_policy *p)
     }
 }
 
-static void __init calculate_raw_policy(void)
+void cpuid_calculate_raw_policy(void)
 {
     struct cpuid_policy *p = &raw_cpuid_policy;
     unsigned int i;
@@ -521,7 +521,7 @@ static void __init calculate_hvm_max_policy(void)
 
 void __init init_guest_cpuid(void)
 {
-    calculate_raw_policy();
+    cpuid_calculate_raw_policy();
     calculate_host_policy();
     calculate_pv_max_policy();
     calculate_hvm_max_policy();
diff --git a/xen/arch/x86/microcode.c b/xen/arch/x86/microcode.c
index d8fd1e4553..ef0907d4df 100644
--- a/xen/arch/x86/microcode.c
+++ b/xen/arch/x86/microcode.c
@@ -503,6 +503,9 @@ int microcode_update(XEN_GUEST_HANDLE_PARAM(const_void) buf, unsigned long len,
             microcode_ops->free_patch(curr_ucode);
         curr_ucode = new_ucode;
         new_ucode = NULL;
+
+        /* Refresh RAW Cpuid policy */
+        cpuid_calculate_raw_policy();
     }
     else
     {
diff --git a/xen/include/asm-x86/cpuid.h b/xen/include/asm-x86/cpuid.h
index 4113a5edee..fc9356eeea 100644
--- a/xen/include/asm-x86/cpuid.h
+++ b/xen/include/asm-x86/cpuid.h
@@ -269,6 +269,8 @@ static inline void cpuid_featureset_to_policy(
 extern struct cpuid_policy raw_cpuid_policy, host_cpuid_policy,
     pv_max_cpuid_policy, hvm_max_cpuid_policy;
 
+void cpuid_calculate_raw_policy(void);
+
 /* Check that all previously present features are still available. */
 bool recheck_cpu_features(unsigned int cpu);
 
