From 417b793fd58cd75e42b2e3f0eba64a660dbfee32 Mon Sep 17 00:00:00 2001
From: Sergey Dyasli <sergey.dyasli@citrix.com>
Date: Wed, 20 Mar 2019 15:59:03 +0000
Subject: [PATCH 7/8] microcode: update raw cpuid policy after a successful
 ucode update

To be upstreamed.

Signed-off-by: Sergey Dyasli <sergey.dyasli@citrix.com>
diff --git a/xen/arch/x86/cpuid.c b/xen/arch/x86/cpuid.c
index 99778bb658..baed5c2e5c 100644
--- a/xen/arch/x86/cpuid.c
+++ b/xen/arch/x86/cpuid.c
@@ -109,7 +109,7 @@ static void __init sanitise_featureset(uint32_t *fs)
                           (fs[FEATURESET_e1d] & ~CPUID_COMMON_1D_FEATURES));
 }
 
-static void __init calculate_raw_featureset(void)
+void calculate_raw_featureset(void)
 {
     unsigned int max, tmp;
 
diff --git a/xen/arch/x86/microcode.c b/xen/arch/x86/microcode.c
index 33886c86b1..8c31543f18 100644
--- a/xen/arch/x86/microcode.c
+++ b/xen/arch/x86/microcode.c
@@ -504,6 +504,9 @@ int microcode_update(XEN_GUEST_HANDLE_PARAM(const_void) buf, unsigned long len,
             microcode_ops->free_patch(curr_ucode);
         curr_ucode = new_ucode;
         new_ucode = NULL;
+
+        /* Refresh RAW Cpuid policy */
+        calculate_raw_featureset();
     }
     else
     {
diff --git a/xen/include/asm-x86/cpuid.h b/xen/include/asm-x86/cpuid.h
index 36a797a3eb..1b95ab91a3 100644
--- a/xen/include/asm-x86/cpuid.h
+++ b/xen/include/asm-x86/cpuid.h
@@ -30,6 +30,7 @@ extern uint32_t raw_featureset[FSCAPINTS];
 extern uint32_t pv_featureset[FSCAPINTS];
 extern uint32_t hvm_featureset[FSCAPINTS];
 
+void calculate_raw_featureset(void);
 void calculate_featuresets(void);
 
 const uint32_t *lookup_deep_deps(uint32_t feature);
