From 4223bfb5a20a0ed7341f2472a845846df87a90a9 Mon Sep 17 00:00:00 2001
From: Daniel Kiper <daniel.kiper@oracle.com>
Date: Fri, 30 Jan 2015 18:54:11 +0100
Subject: [PATCH 07/18] efi: run EFI specific code on EFI platform only

Signed-off-by: Daniel Kiper <daniel.kiper@oracle.com>
diff --git a/xen/common/efi/boot.c b/xen/common/efi/boot.c
index 1e168d3..d765b11 100644
--- a/xen/common/efi/boot.c
+++ b/xen/common/efi/boot.c
@@ -1246,6 +1246,11 @@ void __init efi_init_memory(void)
     } *extra, *extra_head = NULL;
 #endif
 
+#ifndef CONFIG_ARM /* TODO - disabled until implemented on ARM */
+    if ( !efi_platform )
+        return;
+#endif
+
     printk(XENLOG_INFO "EFI memory map:%s\n",
            map_bs ? " (mapping BootServices)" : "");
     for ( i = 0; i < efi_memmap_size; i += efi_mdesc_size )
diff --git a/xen/common/efi/runtime.c b/xen/common/efi/runtime.c
index 093c3e4..b31555f 100644
--- a/xen/common/efi/runtime.c
+++ b/xen/common/efi/runtime.c
@@ -197,6 +197,9 @@ int efi_get_info(uint32_t idx, union xenpf_efi_info *info)
 {
     unsigned int i, n;
 
+    if ( !efi_platform )
+        return -ENOSYS;
+
     switch ( idx )
     {
     case XEN_FW_EFI_VERSION:
@@ -332,6 +335,9 @@ int efi_runtime_call(struct xenpf_efi_runtime_call *op)
     EFI_STATUS status = EFI_NOT_STARTED;
     int rc = 0;
 
+    if ( !efi_platform )
+        return -ENOSYS;
+
     switch ( op->function )
     {
     case XEN_EFI_get_time:
