From 417b793fd58cd75e42b2e3f0eba64a660dbfee32 Mon Sep 17 00:00:00 2001
From: Sergey Dyasli <sergey.dyasli@citrix.com>
Date: Wed, 20 Mar 2019 15:59:03 +0000
Subject: [PATCH 7/8] microcode: update raw cpuid policy after a successful
 ucode update

To be upstreamed.

Signed-off-by: Sergey Dyasli <sergey.dyasli@citrix.com>
diff --git a/xen/arch/x86/cpuid.c b/xen/arch/x86/cpuid.c
index acba0f7583..e3aaf842a0 100644
--- a/xen/arch/x86/cpuid.c
+++ b/xen/arch/x86/cpuid.c
@@ -281,7 +281,7 @@ static void recalculate_misc(struct cpuid_policy *p)
     }
 }
 
-static void __init calculate_raw_policy(void)
+void cpuid_calculate_raw_policy(void)
 {
     struct cpuid_policy *p = &raw_cpuid_policy;
 
@@ -439,7 +439,7 @@ static void __init calculate_hvm_max_policy(void)
 
 void __init init_guest_cpuid(void)
 {
-    calculate_raw_policy();
+    cpuid_calculate_raw_policy();
     calculate_host_policy();
     calculate_pv_max_policy();
     calculate_hvm_max_policy();
diff --git a/xen/arch/x86/microcode.c b/xen/arch/x86/microcode.c
index cd16a81b9a..e7b7ba14c8 100644
--- a/xen/arch/x86/microcode.c
+++ b/xen/arch/x86/microcode.c
@@ -618,6 +618,9 @@ int microcode_update(XEN_GUEST_HANDLE_PARAM(const_void) buf, unsigned long len,
         spin_lock(&microcode_mutex);
         microcode_update_cache(patch);
         spin_unlock(&microcode_mutex);
+
+        /* Refresh RAW Cpuid policy */
+        cpuid_calculate_raw_policy();
     }
     else
         microcode_free_patch(patch);
diff --git a/xen/include/asm-x86/cpuid.h b/xen/include/asm-x86/cpuid.h
index 1b00e832d6..6c0c403788 100644
--- a/xen/include/asm-x86/cpuid.h
+++ b/xen/include/asm-x86/cpuid.h
@@ -53,6 +53,8 @@ extern struct cpuid_policy raw_cpuid_policy, host_cpuid_policy,
 
 extern const struct cpu_policy system_policies[];
 
+void cpuid_calculate_raw_policy(void);
+
 /* Check that all previously present features are still available. */
 bool recheck_cpu_features(unsigned int cpu);
 
