From 4f7007d0b1eac377569b1a25ead874db571b6732 Mon Sep 17 00:00:00 2001
From: Sergey Dyasli <sergey.dyasli@citrix.com>
Date: Fri, 22 Mar 2019 16:45:46 +0000
Subject: [PATCH 8/8] microcode: collect_cpu_info() during
 early_microcode_init()

Upstreaming in progress.

Signed-off-by: Sergey Dyasli <sergey.dyasli@citrix.com>
diff --git a/xen/arch/x86/microcode.c b/xen/arch/x86/microcode.c
index ef0907d4df..cac95878c7 100644
--- a/xen/arch/x86/microcode.c
+++ b/xen/arch/x86/microcode.c
@@ -619,6 +619,8 @@ int __init early_microcode_init(void)
 
     if ( microcode_ops )
     {
+        microcode_ops->collect_cpu_info(&this_cpu(cpu_sig));
+
         if ( ucode_mod.mod_end || ucode_blob.size )
             rc = early_microcode_parse_and_update_cpu();
     }
