From 7aae2bb712b35c3905d6503cf1b7b451e3f62c7b Mon Sep 17 00:00:00 2001
From: Andrew Cooper <andrew.cooper3@citrix.com>
Date: Wed, 8 Aug 2018 13:52:36 +0100
Subject: [PATCH] x86/shadow: dom0-force-shadow for debugging purposes

Signed-off-by: Andrew Cooper <andrew.cooper3@citrix.com>
diff --git a/xen/arch/x86/setup.c b/xen/arch/x86/setup.c
index 219655f..3cec937 100644
--- a/xen/arch/x86/setup.c
+++ b/xen/arch/x86/setup.c
@@ -521,6 +521,10 @@ static inline bool_t using_2M_mapping(void)
            !l1_table_offset((unsigned long)__2M_rwdata_end);
 }
 
+#include <asm/shadow.h>
+static bool_t opt_dom0_force_shadow;
+boolean_param("dom0-force-shadow", opt_dom0_force_shadow);
+
 static void noinline init_done(void)
 {
     void *va;
@@ -531,6 +535,12 @@ static void noinline init_done(void)
     /* MUST be done prior to removing .init data. */
     unregister_init_virtual_region();
 
+    if ( opt_dom0_force_shadow )
+    {
+        hardware_domain->arch.pv_domain.check_l1tf = 1;
+        pv_l1tf_check_pte(hardware_domain, 1, 0xdead000);
+    }
+
     domain_unpause_by_systemcontroller(hardware_domain);
 
     /* Zero the .init code and data. */
