From d8cdad8ab25171027371f90eab8be3bbf25fa751 Mon Sep 17 00:00:00 2001
From: Ross Lagerwall <ross.lagerwall@citrix.com>
Date: Wed, 10 Feb 2016 21:01:09 -0500
Subject: [PATCH 20/26] xsplice: Prevent duplicate payloads to be loaded.

Signed-off-by: Ross Lagerwall <ross.lagerwall@citrix.com>
Signed-off-by: Konrad Rzeszutek Wilk <konrad.wilk@oracle.com>
---
 xen/common/xsplice.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/xen/common/xsplice.c b/xen/common/xsplice.c
index 90f4ffd..e2915a2 100644
--- a/xen/common/xsplice.c
+++ b/xen/common/xsplice.c
@@ -728,6 +728,23 @@ static int find_special_sections(struct payload *payload,
         if ( xen_build_id_check(&payload->id.p, &payload->id.len, n) )
             return -EINVAL;
     }
+    /* Make sure it is not a duplicate. */
+    if ( payload->id.len )
+    {
+        struct payload *data;
+
+        spin_lock(&payload_lock);
+        list_for_each_entry ( data, &payload_list, list )
+        {
+            if ( data != payload && data->id.len &&
+                 !memcmp(data->id.p, payload->id.p, data->id.len) )
+            {
+                spin_unlock(&payload_lock);
+                return -EEXIST;
+            }
+        }
+        spin_unlock(&payload_lock);
+    }
 
     sec = xsplice_elf_sec_by_name(elf, ".xsplice.depends");
     {
-- 
2.4.3

