diff --git a/xen/include/asm-x86/p2m.h b/xen/include/asm-x86/p2m.h
index 2cf73ca..e4d7f87 100644
--- a/xen/include/asm-x86/p2m.h
+++ b/xen/include/asm-x86/p2m.h
@@ -170,7 +170,8 @@ typedef unsigned int p2m_query_t;
 
 #define p2m_is_any_ram(_t)  (p2m_to_mask(_t) &                   \
                              (P2M_RAM_TYPES | P2M_GRANT_TYPES |  \
-                              p2m_to_mask(p2m_map_foreign)))
+                              p2m_to_mask(p2m_map_foreign) |    \
+			      p2m_to_mask(p2m_mmio_write_dm)) )
 
 /* Per-p2m-table state */
 struct p2m_domain {
