From af23e79f46583abc18e8cffdec0b69297cfbdff0 Mon Sep 17 00:00:00 2001
From: Jan Beulich <jbeulich@suse.com>
Date: Wed, 10 Jun 2015 12:01:35 +0200
Subject: [PATCH] x86/EFI: fix EFI_MEMORY_WP handling

Signed-off-by: Jan Beulich <jbeulich@suse.com>
Reviewed-by: Andrew Cooper <andrew.cooper3@citrix.com>
diff --git a/xen/common/efi/boot.c b/xen/common/efi/boot.c
index 77b4487..8aa4419 100644
--- a/xen/common/efi/boot.c
+++ b/xen/common/efi/boot.c
@@ -1195,7 +1195,7 @@ void __init efi_init_memory(void)
         }
 
         if ( desc->Attribute & EFI_MEMORY_WP )
-            prot &= _PAGE_RW;
+            prot &= ~_PAGE_RW;
         if ( desc->Attribute & EFI_MEMORY_XP )
             prot |= _PAGE_NX_BIT;
 
