From 319e431e40139e27d871cbe53976b423301a6af0 Mon Sep 17 00:00:00 2001
From: Andrew Cooper <andrew.cooper3@citrix.com>
Date: Thu, 25 Jun 2015 14:56:26 +0200
Subject: [PATCH] common/vsprintf: special-case DOMID_IDLE handling for %pv

Prints IDLEv0 as opposed to d32767v0

Signed-off-by: Andrew Cooper <andrew.cooper3@citrix.com>
diff --git a/xen/common/vsprintf.c b/xen/common/vsprintf.c
index 065cc42..51b5e4e 100644
--- a/xen/common/vsprintf.c
+++ b/xen/common/vsprintf.c
@@ -336,9 +336,14 @@ static char *pointer(char *str, char *end, const char **fmt_ptr,
         const struct vcpu *v = arg;
 
         ++*fmt_ptr;
-        if ( str < end )
-            *str = 'd';
-        str = number(str + 1, end, v->domain->domain_id, 10, -1, -1, 0);
+        if ( unlikely(v->domain->domain_id == DOMID_IDLE) )
+            str = string(str, end, "IDLE", -1, -1, 0);
+        else
+        {
+            if ( str < end )
+                *str = 'd';
+            str = number(str + 1, end, v->domain->domain_id, 10, -1, -1, 0);
+        }
         if ( str < end )
             *str = 'v';
         return number(str + 1, end, v->vcpu_id, 10, -1, -1, 0);
