From e318fa314550b1978525aef30c325715a41aa0b7 Mon Sep 17 00:00:00 2001
From: Wei Liu <wei.liu2@citrix.com>
Date: Thu, 23 Feb 2017 15:18:20 +0000
Subject: [PATCH] libs/devicemodel: initialise op_bufs in xendevicemodel_xcall

To avoid freeing uninitialised buffer when taking the first error exit
path.

Signed-off-by: Wei Liu <wei.liu2@citrix.com>
Acked-by: Ian Jackson <ian.jackson@eu.citrix.com>
---
 tools/libs/devicemodel/core.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/tools/libs/devicemodel/core.c b/tools/libs/devicemodel/core.c
index 504543c..19ebef6 100644
--- a/tools/libs/devicemodel/core.c
+++ b/tools/libs/devicemodel/core.c
@@ -79,7 +79,7 @@ int xendevicemodel_xcall(xendevicemodel_handle *dmod,
 {
     int ret = -1;
     void **xcall_bufs;
-    xen_dm_op_buf_t *op_bufs;
+    xen_dm_op_buf_t *op_bufs = NULL;
     unsigned int i;
 
     xcall_bufs = calloc(nr_bufs, sizeof(*xcall_bufs));
-- 
2.7.4

