From aba97792a8c5bb20886cc0969b4e43d6470b9aba Mon Sep 17 00:00:00 2001
From: Jan Beulich <jbeulich@suse.com>
Date: Mon, 18 Sep 2017 12:27:06 +0200
Subject: [PATCH] x86/cpuidle: add new CPU families

Bring code up-to-date with SDM version 063.

Signed-off-by: Jan Beulich <jbeulich@suse.com>
Acked-by: Andrew Cooper <andrew.cooper3@citrix.com>
Acked-by: Kevin Tian <kevin.tian@intel.com>
diff --git a/xen/arch/x86/acpi/cpu_idle.c b/xen/arch/x86/acpi/cpu_idle.c
index 244667993d..057bdfedde 100644
--- a/xen/arch/x86/acpi/cpu_idle.c
+++ b/xen/arch/x86/acpi/cpu_idle.c
@@ -164,7 +164,13 @@ static void do_get_hw_residencies(void *arg)
     case 0x56:
     /* Skylake */
     case 0x4E:
+    case 0x55:
     case 0x5E:
+    /* Cannon Lake */
+    case 0x66:
+    /* Kaby Lake */
+    case 0x8E:
+    case 0x9E:
         GET_PC2_RES(hw_res->pc2);
         GET_CC7_RES(hw_res->cc7);
         /* fall through */
@@ -183,8 +189,10 @@ static void do_get_hw_residencies(void *arg)
         GET_CC3_RES(hw_res->cc3);
         GET_CC6_RES(hw_res->cc6);
         break;
-    /* next gen Xeon Phi */
+    /* Xeon Phi Knights Landing */
     case 0x57:
+    /* Xeon Phi Knights Mill */
+    case 0x85:
         GET_CC3_RES(hw_res->mc0); /* abusing GET_CC3_RES */
         GET_CC6_RES(hw_res->mc6); /* abusing GET_CC6_RES */
         GET_PC2_RES(hw_res->pc2);
@@ -215,6 +223,8 @@ static void do_get_hw_residencies(void *arg)
     /* Goldmont */
     case 0x5C:
     case 0x5F:
+    /* Goldmont Plus */
+    case 0x7A:
         GET_PC2_RES(hw_res->pc2);
         GET_PC3_RES(hw_res->pc3);
         GET_PC6_RES(hw_res->pc6);
