diff --git a/tools/tests/mce-test/tools/Makefile b/tools/tests/mce-test/tools/Makefile
index 5ee001f..aba7177 100644
--- a/tools/tests/mce-test/tools/Makefile
+++ b/tools/tests/mce-test/tools/Makefile
@@ -3,9 +3,11 @@ include $(XEN_ROOT)/tools/Rules.mk
 
 CFLAGS += -Werror
 CFLAGS += $(CFLAGS_libxenctrl)
+# xen-mceinj.c incorrectly use libxc internals
+CFLAGS += -I$(XEN_ROOT)/tools/libxc
 CFLAGS += $(CFLAGS_libxenguest)
-CFLAGS += $(CFLAGS_libxenstore) 
-CFLAGS += $(CFLAGS_xeninclude) 
+CFLAGS += $(CFLAGS_libxenstore)
+CFLAGS += $(CFLAGS_xeninclude)
 
 .PHONY: all
 all: xen-mceinj
