diff --git a/xen/Makefile b/xen/Makefile
index 9d91ca7..85b340f 100644
--- a/xen/Makefile
+++ b/xen/Makefile
@@ -116,6 +116,7 @@ delete-unfresh-files:
 		rm -f include/xen/compile.h; \
 	fi
 
+.banner: XEN_FULLVERSION := $$(echo $(XEN_FULLVERSION) | sed "s/-xs.*//")
 .banner: Makefile
 	@if which figlet >/dev/null 2>&1 ; then \
 		echo " Xen $(XEN_FULLVERSION)" | figlet -f tools/xen.flf > $@.tmp; \
