diff --git a/xen/common/domctl.c b/xen/common/domctl.c
index 16d4559..60d767b 100644
--- a/xen/common/domctl.c
+++ b/xen/common/domctl.c
@@ -1034,13 +1034,27 @@ long do_domctl(XEN_GUEST_HANDLE_PARAM(xen_domctl_t) u_domctl)
             break;
 
         ret = -EPERM;
-        if ( !iomem_access_permitted(current->domain, mfn, mfn_end) ||
-             !iomem_access_permitted(d, mfn, mfn_end) )
+        if ( !iomem_access_permitted(current->domain, mfn, mfn_end) )
+        {
+            gdprintk(XENLOG_INFO, "dbg: !perm(dom%d, %lx, %lx)\n",
+                     current->domain->domain_id, mfn, mfn_end);
+            break;
+        }
+
+        if ( !iomem_access_permitted(d, mfn, mfn_end) )
+        {
+            gdprintk(XENLOG_INFO, "dbg: !perm(dom%d, %lx, %lx)\n",
+                     d->domain_id, mfn, mfn_end);
             break;
+        }
 
         ret = xsm_iomem_mapping(XSM_HOOK, d, mfn, mfn_end, add);
         if ( ret )
+        {
+            gdprintk(XENLOG_INFO, "dbg: !xsm(dom%d, %lx, %lx, %d)\n",
+                     d->domain_id, mfn, mfn_end, add);
             break;
+        }
 
         if ( add )
         {
