diff --git a/xen/arch/x86/usercopy.c b/xen/arch/x86/usercopy.c
index 4cc78f5..4ffd869 100644
--- a/xen/arch/x86/usercopy.c
+++ b/xen/arch/x86/usercopy.c
@@ -45,6 +45,10 @@ unsigned long __copy_to_user_ll(void __user *to, const void *from, unsigned n)
         : "memory" );
     clac();
 
+    if (current->domain->domain_id == 0 && __n) {
+        guest_cpu_user_regs()->r9 = read_cr2();
+    }
+
     return __n;
 }
 
@@ -90,6 +94,10 @@ __copy_from_user_ll(void *to, const void __user *from, unsigned n)
         : "memory" );
     clac();
 
+    if (current->domain->domain_id == 0 && __n) {
+        guest_cpu_user_regs()->r9 = read_cr2();
+    }
+
     return __n;
 }
 
