diff --git a/xen/arch/x86/mm.c b/xen/arch/x86/mm.c
index d39dd25..4cf1ef1 100644
--- a/xen/arch/x86/mm.c
+++ b/xen/arch/x86/mm.c
@@ -2545,6 +2545,23 @@ static int _put_final_page_type(struct page_info *page, unsigned long type,
             dec_linear_uses(page);
             dec_linear_entries(ptpg);
         }
+#ifndef NDEBUG
+        if ( unlikely(!(!page->linear_pt_count || page_get_owner(page)->is_dying)) )
+        {
+            unsigned int i;
+            unsigned long *ptr = _p(page);
+
+            printk("*** %s({%p, %"PRI_mfn"}, %lx, %d, {%p, %"PRI_mfn"})\n",
+                   __func__, page, __page_to_mfn(page), type, preemptible,
+                   ptpg, ptpg ? __page_to_mfn(ptpg) : -1ul);
+            printk("*** owner d%d, dying %u, tlbtime %#x\n",
+                   page_get_owner(page)->domain_id,
+                   page_get_owner(page)->is_dying,
+                   tlbflush_current_time());
+            for ( i = 0; _p(&ptr[i]) < _p(page + 1); ++i )
+                printk("*** [%02u]: %016lx\n", i * 8, ptr[i]);
+        }
+#endif
         ASSERT(!page->linear_pt_count || page_get_owner(page)->is_dying);
         /*
          * Record TLB information for flush later. We do not stamp page tables
