diff --git a/xen/tools/scmversion b/xen/tools/scmversion
index b6ce049..08a64db 100755
--- a/xen/tools/scmversion
+++ b/xen/tools/scmversion
@@ -41,6 +41,15 @@ scm_version()
 	if test -d .git && head=`git rev-parse --verify --short HEAD 2>/dev/null`; then
 		date=`git show -s --pretty="%ad" HEAD`
 
+		if test -d .git/patches/.git; then
+			# Guilt patch queue - Grab the revisions separately
+			head=`git rev-parse --verify --short=12 qparent`
+			pqhead=`git --git-dir=.git/patches/.git rev-parse --verify --short=12 HEAD`
+
+			printf '%s, pq %s' "$head" "$pqhead"
+			return
+		fi
+
 		printf '%s %s%s' "$date" git: $head
 
 		# Is this git on svn?
@@ -62,15 +71,11 @@ scm_version()
 
 	# Check for mercurial and a mercurial repo.
 	if test -d .hg && hgid=`hg id 2>/dev/null`; then
-		id=`printf '%s' "$hgid" | sed 's/[+ ].*//'`
-		date=`hg parents --template "{date|date}"`
-		printf '%s %s%s' "$date" hg: "$id"
-
-		# Are there uncommitted changes?
-		# These are represented by + after the changeset id.
-		case "$hgid" in
-			*+|*+\ *) printf '%s' -dirty ;;
-		esac
+		TEMPLATE="{rev}:{node|short}"
+		BASE_CSET_STR="$(hg log -r qparent --template $TEMPLATE)"
+		PQ_CSET_STR="$(hg log -R .hg/patches -r tip --template $TEMPLATE)"
+
+		printf '%s, pq %s' "$BASE_CSET_STR" "$PQ_CSET_STR"
 
 		# All done with mercurial
 		return
