diff --git a/misc/coverity/model.c b/misc/coverity/model.c
index bd62566..70fbc95 100644
--- a/misc/coverity/model.c
+++ b/misc/coverity/model.c
@@ -121,11 +121,11 @@ void libxl__ctx_unlock(libxl_ctx *ctx)
 }
 
 /*
- * Coverity doesn't understand __builtin_unreachable(), which causes it to
- * incorrectly find issues based on continuing execution along the false
- * branch of an ASSERT().
+ * Coverity doesn't understand our unreachable() macro, which causes it to
+ * incorrectly find issues based on continuing execution along unreachable
+ * paths.
  */
-void __builtin_unreachable(void)
+void unreachable(void)
 {
     __coverity_panic__();
 }
