diff --git a/xen/arch/x86/x86_emulate/x86_emulate.c b/xen/arch/x86/x86_emulate/x86_emulate.c
index 9a9d3c4..c0a9b68 100644
--- a/xen/arch/x86/x86_emulate/x86_emulate.c
+++ b/xen/arch/x86/x86_emulate/x86_emulate.c
@@ -1688,8 +1688,8 @@ x86_emulate(
  done_prefixes:
 
     /* %{e,c,s,d}s overrides are ignored in 64bit mode. */
-    if ( mode_64bit() && override_seg < x86_seg_fs )
-        override_seg = x86_seg_none;
+    if ( mode_64bit() && (unsigned int)override_seg < x86_seg_fs )
+        override_seg = -1;
 
     if ( rex_prefix & REX_W )
         op_bytes = 8;
