# HG changeset patch
# Parent 4def7d6b0ac5a0c05ebbb308b6f4252f41d3704d

diff --git a/tools/libxc/xc_cpuid_x86.c b/tools/libxc/xc_cpuid_x86.c
index 84b9df2..ee61a61 100644
--- a/tools/libxc/xc_cpuid_x86.c
+++ b/tools/libxc/xc_cpuid_x86.c
@@ -73,7 +73,7 @@ static void xc_cpuid_brand_get(char *str)
 static void amd_xc_cpuid_policy(
     xc_interface *xch, domid_t domid,
     const unsigned int *input, unsigned int *regs,
-    int is_pae, int is_nestedhvm)
+    int is_pae, int is_nestedhvm, uint64_t xfeature_mask)
 {
     switch ( input[0] )
     {
@@ -102,7 +102,8 @@ static void amd_xc_cpuid_policy(
                     bitmaskof(X86_FEATURE_OSVW) |
                     bitmaskof(X86_FEATURE_XOP) |
                     bitmaskof(X86_FEATURE_LWP) |
-                    bitmaskof(X86_FEATURE_FMA4) |
+                    ((xfeature_mask != 0) ?
+                     (bitmaskof(X86_FEATURE_FMA4)) : 0) |
                     bitmaskof(X86_FEATURE_TBM) |
                     bitmaskof(X86_FEATURE_DBEXT));
         regs[3] &= (0x0183f3ff | /* features shared with 0x00000001:EDX */
@@ -275,7 +276,6 @@ static void xc_cpuid_hvm_policy(
         regs[2] &= (bitmaskof(X86_FEATURE_XMM3) |
                     bitmaskof(X86_FEATURE_PCLMULQDQ) |
                     bitmaskof(X86_FEATURE_SSSE3) |
-                    bitmaskof(X86_FEATURE_FMA) |
                     bitmaskof(X86_FEATURE_CX16) |
                     bitmaskof(X86_FEATURE_PCID) |
                     bitmaskof(X86_FEATURE_SSE4_1) |
@@ -283,10 +283,11 @@ static void xc_cpuid_hvm_policy(
                     bitmaskof(X86_FEATURE_MOVBE)  |
                     bitmaskof(X86_FEATURE_POPCNT) |
                     bitmaskof(X86_FEATURE_AES) |
-                    bitmaskof(X86_FEATURE_F16C) |
                     bitmaskof(X86_FEATURE_RDRAND) |
                     ((xfeature_mask != 0) ?
                      (bitmaskof(X86_FEATURE_AVX) |
+                      bitmaskof(X86_FEATURE_F16C) |
+                      bitmaskof(X86_FEATURE_FMA) |
                       bitmaskof(X86_FEATURE_XSAVE)) : 0));
 
         regs[2] |= (bitmaskof(X86_FEATURE_HYPERVISOR) |
@@ -330,7 +331,8 @@ static void xc_cpuid_hvm_policy(
             regs[1] &= (bitmaskof(X86_FEATURE_TSC_ADJUST) |
                         bitmaskof(X86_FEATURE_BMI1) |
                         bitmaskof(X86_FEATURE_HLE)  |
-                        bitmaskof(X86_FEATURE_AVX2) |
+			((xfeature_mask != 0) ?
+			 (bitmaskof(X86_FEATURE_AVX2)) : 0) |
                         bitmaskof(X86_FEATURE_SMEP) |
                         bitmaskof(X86_FEATURE_BMI2) |
                         bitmaskof(X86_FEATURE_ERMS) |
@@ -396,7 +398,7 @@ static void xc_cpuid_hvm_policy(
 
     xc_cpuid_brand_get(brand);
     if ( strstr(brand, "AMD") )
-        amd_xc_cpuid_policy(xch, domid, input, regs, is_pae, is_nestedhvm);
+        amd_xc_cpuid_policy(xch, domid, input, regs, is_pae, is_nestedhvm, xfeature_mask);
     else
         intel_xc_cpuid_policy(xch, domid, input, regs, is_pae, is_nestedhvm);
 
@@ -471,7 +473,8 @@ static void xc_cpuid_pv_policy(
         if ( input[1] == 0 )
             regs[1] &= (bitmaskof(X86_FEATURE_BMI1) |
                         bitmaskof(X86_FEATURE_HLE)  |
-                        bitmaskof(X86_FEATURE_AVX2) |
+			((xfeature_mask != 0) ?
+			 (bitmaskof(X86_FEATURE_AVX2)) : 0) |
                         bitmaskof(X86_FEATURE_BMI2) |
                         bitmaskof(X86_FEATURE_ERMS) |
                         bitmaskof(X86_FEATURE_RTM)  |
