# HG changeset patch
# Parent 4def7d6b0ac5a0c05ebbb308b6f4252f41d3704d

diff --git a/tools/libxc/xc_cpuid_x86.c b/tools/libxc/xc_cpuid_x86.c
index 2d0d82e..6ccdd2d 100644
--- a/tools/libxc/xc_cpuid_x86.c
+++ b/tools/libxc/xc_cpuid_x86.c
@@ -173,7 +173,8 @@ static void amd_xc_cpuid_policy(xc_interface *xch,
                     bitmaskof(X86_FEATURE_OSVW) |
                     bitmaskof(X86_FEATURE_XOP) |
                     bitmaskof(X86_FEATURE_LWP) |
-                    bitmaskof(X86_FEATURE_FMA4) |
+                    ((info->xfeature_mask != 0) ?
+                     (bitmaskof(X86_FEATURE_FMA4)) : 0) |
                     bitmaskof(X86_FEATURE_TBM) |
                     bitmaskof(X86_FEATURE_DBEXT));
         regs[3] &= (0x0183f3ff | /* features shared with 0x00000001:EDX */
@@ -329,7 +330,6 @@ static void xc_cpuid_hvm_policy(xc_interface *xch,
         regs[2] &= (bitmaskof(X86_FEATURE_XMM3) |
                     bitmaskof(X86_FEATURE_PCLMULQDQ) |
                     bitmaskof(X86_FEATURE_SSSE3) |
-                    bitmaskof(X86_FEATURE_FMA) |
                     bitmaskof(X86_FEATURE_CX16) |
                     bitmaskof(X86_FEATURE_PCID) |
                     bitmaskof(X86_FEATURE_SSE4_1) |
@@ -337,10 +337,11 @@ static void xc_cpuid_hvm_policy(xc_interface *xch,
                     bitmaskof(X86_FEATURE_MOVBE)  |
                     bitmaskof(X86_FEATURE_POPCNT) |
                     bitmaskof(X86_FEATURE_AES) |
-                    bitmaskof(X86_FEATURE_F16C) |
                     bitmaskof(X86_FEATURE_RDRAND) |
                     ((info->xfeature_mask != 0) ?
                      (bitmaskof(X86_FEATURE_AVX) |
+                      bitmaskof(X86_FEATURE_F16C) |
+                      bitmaskof(X86_FEATURE_FMA) |
                       bitmaskof(X86_FEATURE_XSAVE)) : 0));
 
         regs[2] |= (bitmaskof(X86_FEATURE_HYPERVISOR) |
@@ -385,7 +386,8 @@ static void xc_cpuid_hvm_policy(xc_interface *xch,
             regs[1] &= (bitmaskof(X86_FEATURE_TSC_ADJUST) |
                         bitmaskof(X86_FEATURE_BMI1) |
                         bitmaskof(X86_FEATURE_HLE)  |
-                        bitmaskof(X86_FEATURE_AVX2) |
+			((info->xfeature_mask != 0) ?
+			 (bitmaskof(X86_FEATURE_AVX2)) : 0) |
                         bitmaskof(X86_FEATURE_SMEP) |
                         bitmaskof(X86_FEATURE_BMI2) |
                         bitmaskof(X86_FEATURE_ERMS) |
@@ -508,7 +510,8 @@ static void xc_cpuid_pv_policy(xc_interface *xch,
         if ( input[1] == 0 )
             regs[1] &= (bitmaskof(X86_FEATURE_BMI1) |
                         bitmaskof(X86_FEATURE_HLE)  |
-                        bitmaskof(X86_FEATURE_AVX2) |
+			((info->xfeature_mask != 0) ?
+			 (bitmaskof(X86_FEATURE_AVX2)) : 0) |
                         bitmaskof(X86_FEATURE_BMI2) |
                         bitmaskof(X86_FEATURE_ERMS) |
                         bitmaskof(X86_FEATURE_RTM)  |
