diff --git a/xen/arch/x86/cpuid.c b/xen/arch/x86/cpuid.c
index af6889d..4fde717 100644
--- a/xen/arch/x86/cpuid.c
+++ b/xen/arch/x86/cpuid.c
@@ -139,6 +139,9 @@ static void calculate_hvm_featureset(void)
     /* Unconditionally claim to be able to set the hypervisor bit. */
     __set_bit(X86_FEATURE_HYPERVISOR, hvm_featureset);
 
+    /* The fixed apic id layout is incompatible with hyperthreading. */
+    __clear_bit(X86_FEATURE_HT, hvm_featureset);
+
     /*
      * On AMD, PV guests are entirely unable to use 'sysenter' as Xen runs in
      * long mode (and init_amd() has cleared it out of host capabilities), but
