Needed for the xapi toolstack

diff --git a/tools/libxc/include/xenguest.h b/tools/libxc/include/xenguest.h
index 55755cf..cbccf7d 100644
--- a/tools/libxc/include/xenguest.h
+++ b/tools/libxc/include/xenguest.h
@@ -335,4 +335,8 @@ xen_pfn_t *xc_map_m2p(xc_interface *xch,
                       unsigned long max_mfn,
                       int prot,
                       unsigned long *mfn0);
+
+/* Accessors for struct xc_dom_image */
+const char *xc_dom_get_native_protocol(struct xc_dom_image *dom);
+
 #endif /* XENGUEST_H */
diff --git a/tools/libxc/xc_dom_core.c b/tools/libxc/xc_dom_core.c
index baa62a1..f7f65ef 100644
--- a/tools/libxc/xc_dom_core.c
+++ b/tools/libxc/xc_dom_core.c
@@ -1032,6 +1032,12 @@ int xc_dom_build_image(struct xc_dom_image *dom)
     return -1;
 }
 
+/* Accessors */
+const char *xc_dom_get_native_protocol(struct xc_dom_image *dom)
+{
+    return dom->arch_hooks ? dom->arch_hooks->native_protocol : "";
+}
+
 /*
  * Local variables:
  * mode: C
