tools/libxc: Helper function for XEN_SYSCTL_consoleringsize.

Signed-off-by: Andrew Cooper <andrew.cooper3@citrix.com>

diff --git a/tools/libxc/include/xenctrl.h b/tools/libxc/include/xenctrl.h
index 2aa86eb..0ec535a 100644
--- a/tools/libxc/include/xenctrl.h
+++ b/tools/libxc/include/xenctrl.h
@@ -1215,6 +1215,7 @@ int xc_readconsolering(xc_interface *xch,
                        char *buffer,
                        unsigned int *pnr_chars,
                        int clear, int incremental, uint32_t *pindex);
+int xc_consoleringsize(xc_interface *xch, uint64_t *psize);
 
 int xc_send_debug_keys(xc_interface *xch, char *keys);
 
diff --git a/tools/libxc/xc_misc.c b/tools/libxc/xc_misc.c
index e253a58..704f4cd 100644
--- a/tools/libxc/xc_misc.c
+++ b/tools/libxc/xc_misc.c
@@ -139,6 +139,26 @@ int xc_readconsolering(xc_interface *xch,
     return ret;
 }
 
+int xc_consoleringsize(xc_interface *xch, uint64_t *psize)
+{
+    int ret = -1;
+    DECLARE_SYSCTL;
+
+    if ( !psize )
+    {
+        errno = EINVAL;
+        return ret;
+    }
+
+    sysctl.cmd = XEN_SYSCTL_consoleringsize;
+    ret = do_sysctl(xch, &sysctl);
+
+    if ( !ret )
+        *psize = sysctl.u.consoleringsize.size;
+
+    return ret;
+}
+
 int xc_send_debug_keys(xc_interface *xch, char *keys)
 {
     int ret, len = strlen(keys);
