diff --git a/mk/xen.spec.in b/mk/xen.spec.in
index 33150ca..8c0df30 100644
--- a/mk/xen.spec.in
+++ b/mk/xen.spec.in
@@ -454,6 +454,7 @@ rm -rf $RPM_BUILD_ROOT
 %{_sbindir}/xen-mceinj
 %{_sbindir}/xen-mfndump
 %exclude %{_sbindir}/xen-ringwatch
+%{_sbindir}/xen-vmdebug
 %{_sbindir}/xenbaked
 %{_sbindir}/xenconsoled
 %{_sbindir}/xencov
diff --git a/tools/vmdebug/hvmparam.c b/tools/vmdebug/hvmparam.c
index 68d947d..ed9dfa9 100644
--- a/tools/vmdebug/hvmparam.c
+++ b/tools/vmdebug/hvmparam.c
@@ -27,6 +27,7 @@ static hvmparam_desc_t params[HVM_NR_PARAMS] =
     PARAM(IOREQ_PFN, "Device Model IO Request frame"),
     PARAM(BUFIOREQ_PFN, "Device Model Buffered IO Request frame"),
     PARAM(BUFIOREQ_EVTCHN, "Device Model Buffered IO Request event channel"),
+    PARAM(NX_ENABLED, "(XS) Pagetable NX bit available?"),
     PARAM(VIRIDIAN, "Windows Viridian enlightenments"),
     PARAM(TIMER_MODE, "Timer tick delivery settings"),
     PARAM(HPET_ENABLED, "HPET available?"),
