diff --git a/mk/xen.spec.in b/mk/xen.spec.in
index 250620e..eb0ff5c 100644
--- a/mk/xen.spec.in
+++ b/mk/xen.spec.in
@@ -222,6 +222,7 @@ rm -f %{buildroot}/usr/share/doc/xen/README.xenmon
 # Copy ocaml xenstored to legacy location until we can fix up the init scripts
 mkdir -p $RPM_BUILD_ROOT/opt/xensource/bin/
 cp -f %{buildroot}%{_sbindir}/oxenstored %{buildroot}/opt/xensource/bin/xenstored
+ln -s convert-legacy-stream %{buildroot}%{_libexecdir}/%{name}/bin/legacy.py
 
 %clean
 rm -rf $RPM_BUILD_ROOT
@@ -409,11 +410,16 @@ rm -rf $RPM_BUILD_ROOT
 #{python_sitearch}/xen/__init__.py*           - Must not duplicate xen-tools
 #{python_sitearch}/xen/lowlevel/__init__.py*  - Must not duplicate xen-tools
 %{python_sitearch}/xen/lowlevel/xc.so
+%{python_sitearch}/xen/streamv2/__init__.py*
+%{python_sitearch}/xen/streamv2/format.py*
 /lib/udev/rules.d/90-xen-backend.rules
+%{_libexecdir}/%{name}/bin/convert-legacy-stream
+%{_libexecdir}/%{name}/bin/legacy.py*
 %{_libexecdir}/%{name}/bin/libxl-save-helper
 %{_libexecdir}/%{name}/bin/lsevtchn
 %{_libexecdir}/%{name}/bin/pygrub
 %{_libexecdir}/%{name}/bin/readnotes
+%{_libexecdir}/%{name}/bin/verify-stream-v2
 %{_libexecdir}/%{name}/bin/xen-init-dom0
 %{_libexecdir}/%{name}/bin/xenconsole
 %{_libexecdir}/%{name}/bin/xenctx
