diff --git a/mk/xen.spec.in b/mk/xen.spec.in
index 64ec038..5b98eab 100644
--- a/mk/xen.spec.in
+++ b/mk/xen.spec.in
@@ -270,6 +270,7 @@ rm -rf $RPM_BUILD_ROOT
 %{_includedir}/%{name}/arch-arm.h
 %{_includedir}/%{name}/arch-arm/hvm/save.h
 %{_includedir}/%{name}/arch-x86/cpuid.h
+%{_includedir}/%{name}/arch-x86/featureset.h
 %{_includedir}/%{name}/arch-x86/hvm/save.h
 %{_includedir}/%{name}/arch-x86/pmu.h
 %{_includedir}/%{name}/arch-x86/xen-mca.h
@@ -407,6 +408,7 @@ rm -rf $RPM_BUILD_ROOT
 %exclude %{_sysconfdir}/%{name}/xlexample.pvlinux
 %config %{_sysconfdir}/xen/xl.conf
 %{_bindir}/pygrub
+%{_bindir}/xen-cpuid
 %{_bindir}/xen-detect
 %{_bindir}/xenalyze
 %{_bindir}/xencons
