diff --git a/mk/xen.spec.in b/mk/xen.spec.in
index 1d69bfa..ef9c00c 100644
--- a/mk/xen.spec.in
+++ b/mk/xen.spec.in
@@ -518,6 +518,10 @@ rm -rf $RPM_BUILD_ROOT
 
 %files ocaml-libs
 %defattr(-,root,root,-)
+%{_libdir}/ocaml/gnt/META
+%{_libdir}/ocaml/gnt/gnt.cma
+%{_libdir}/ocaml/stublibs/dllgnt_stubs.so
+%{_libdir}/ocaml/stublibs/dllgnt_stubs.so.owner
 %{_libdir}/ocaml/stublibs/dllxenbus_stubs.so
 %{_libdir}/ocaml/stublibs/dllxenbus_stubs.so.owner
 %{_libdir}/ocaml/stublibs/dllxenctrl_stubs.so
@@ -549,6 +553,11 @@ rm -rf $RPM_BUILD_ROOT
 
 %files ocaml-devel
 %defattr(-,root,root,-)
+%{_libdir}/ocaml/gnt/gnt.a
+%{_libdir}/ocaml/gnt/gnt.cmi
+%{_libdir}/ocaml/gnt/gnt.cmx
+%{_libdir}/ocaml/gnt/gnt.cmxa
+%{_libdir}/ocaml/gnt/libgnt_stubs.a
 %{_libdir}/ocaml/xenbus/libxenbus_stubs.a
 %{_libdir}/ocaml/xenbus/xenbus.a
 %{_libdir}/ocaml/xenbus/xenbus.cmi
