diff --git a/mk/xen.spec.in b/mk/xen.spec.in
index ccc8100..f0d5695 100644
--- a/mk/xen.spec.in
+++ b/mk/xen.spec.in
@@ -269,6 +269,7 @@ rm -rf $RPM_BUILD_ROOT
 %{_includedir}/%{name}/arch-arm.h
 %{_includedir}/%{name}/arch-arm/hvm/save.h
 %{_includedir}/%{name}/arch-x86/cpuid.h
+%{_includedir}/%{name}/arch-x86/cpufeatureset.h
 %{_includedir}/%{name}/arch-x86/hvm/save.h
 %{_includedir}/%{name}/arch-x86/pmu.h
 %{_includedir}/%{name}/arch-x86/xen-mca.h
@@ -421,6 +422,7 @@ rm -rf $RPM_BUILD_ROOT
 %exclude %{_sysconfdir}/%{name}/xlexample.pvlinux
 %config %{_sysconfdir}/xen/xl.conf
 %{_bindir}/pygrub
+%{_bindir}/xen-cpuid
 %{_bindir}/xen-detect
 %{_bindir}/xenalyze
 %{_bindir}/xencons
