diff --git a/tools/xenguest/xenguest.c b/tools/xenguest/xenguest.c
index 826a1c6..2bf9ea9 100644
--- a/tools/xenguest/xenguest.c
+++ b/tools/xenguest/xenguest.c
@@ -82,6 +82,7 @@ static const char *opt_features = NULL;
 static int opt_flags = 0;
 static int opt_mem_max_mib = -1;
 static int opt_mem_start_mib = -1;
+static int hvm_mode = true;
 
 static const char *xg_mode_names[] = {
     [XG_MODE_SAVE] = "save",
@@ -616,7 +617,7 @@ void do_migrate_live(struct threadargs *args)
 
    send_return(args->cli, NULL);
 
-   emu_stub_xc_domain_save(opt_fd, (void*) &args->cli, opt_flags, true);
+   emu_stub_xc_domain_save(opt_fd, (void*) &args->cli, opt_flags, hvm_mode);
    xg_info(" Finished, send compleate");
    if (progress_cli.num >= 0)
        send_event_migrate_completed(progress_cli);
@@ -631,7 +632,7 @@ void do_migrate_nonlive(struct threadargs *args)
 
    send_return(args->cli, NULL);
 
-   emu_stub_xc_domain_save(opt_fd, (void*) &args->cli, opt_flags, true);
+   emu_stub_xc_domain_save(opt_fd, (void*) &args->cli, opt_flags, hvm_mode);
    xg_info(" Finished, send compleate");
    if (progress_cli.num >= 0)
        send_event_migrate_completed(progress_cli);
@@ -667,7 +668,7 @@ void do_cmd_restore(struct threadargs *args)
     }
     send_return(args->cli, NULL);
 
-    stub_xc_domain_restore(opt_fd, opt_store_port, opt_console_port, true,
+    stub_xc_domain_restore(opt_fd, opt_store_port, opt_console_port, hvm_mode,
                            &store_mfn, &console_mfn);
 
     xg_info("Restore compleat, send result");
@@ -694,10 +695,11 @@ int str_lookup(const char* table[], char cmp[]) {
 enum arg_keys {
      unknown    = -1,
      store_port = 0,
-     console_port
+     console_port,
+     mode_pv
 };
 
-const char* setable_args[] = {"store_port", "console_port", NULL};
+const char* setable_args[] = {"store_port", "console_port","pv", NULL};
 
 
 void do_cmd_set_args(struct threadargs *args)
@@ -745,6 +747,9 @@ void do_cmd_set_args(struct threadargs *args)
             }
             xg_info("console port = %d", opt_console_port);
        break;
+       case mode_pv:
+            hvm_mode = false;
+       break;
        default:
            xg_err("Unknown arg: %s", s_key);
        }
