diff --git a/xen/drivers/passthrough/iommu.c b/xen/drivers/passthrough/iommu.c
index 135ff9ec2c..df3a23f418 100644
--- a/xen/drivers/passthrough/iommu.c
+++ b/xen/drivers/passthrough/iommu.c
@@ -271,9 +271,6 @@ int iommu_map(struct domain *d, unsigned long gfn, unsigned long mfn,
     if ( !iommu_enabled || !hd->platform_ops )
         return 0;
 
-    ASSERT(IS_ALIGNED(gfn, (1ul << page_order)));
-    ASSERT(IS_ALIGNED(mfn, (1ul << page_order)));
-
     for ( i = 0; i < (1ul << page_order); i++ )
     {
         rc = hd->platform_ops->map_page(d, gfn + i, mfn + i, flags,
@@ -328,8 +325,6 @@ int iommu_unmap(struct domain *d, unsigned long gfn,
     if ( !iommu_enabled || !hd->platform_ops )
         return 0;
 
-    ASSERT(IS_ALIGNED(gfn, (1ul << page_order)));
-
     for ( i = 0; i < (1ul << page_order); i++ )
     {
         int err = hd->platform_ops->unmap_page(d, gfn + i, flush_flags);
