diff --git a/xen/common/pv_iommu.c b/xen/common/pv_iommu.c
index c784f6f..6aedd9d 100644
--- a/xen/common/pv_iommu.c
+++ b/xen/common/pv_iommu.c
@@ -796,6 +796,10 @@ ret_t do_iommu_op(XEN_GUEST_HANDLE_PARAM(void) arg, unsigned int count)
     struct pv_iommu_op op;
     struct domain *d = current->domain;
 
+    /* Restrict to HW domain until upstream approved */
+    if ( !is_hardware_domain(d) )
+        return -ENODEV;
+
     if ( (int)count < 0 )
         return -EINVAL;
 
