diff --git a/tools/libxc/include/xenctrl.h b/tools/libxc/include/xenctrl.h
index 22f394953b..2d24a9e694 100644
--- a/tools/libxc/include/xenctrl.h
+++ b/tools/libxc/include/xenctrl.h
@@ -2041,13 +2041,6 @@ int xc_monitor_descriptor_access(xc_interface *xch, uint32_t domain_id,
                                  bool enable);
 int xc_monitor_guest_request(xc_interface *xch, uint32_t domain_id,
                              bool enable, bool sync, bool allow_userspace);
-/*
- * Disables page-walk mem_access events by emulating. If the
- * emulation can not be performed then a VM_EVENT_REASON_EMUL_UNIMPLEMENTED
- * event will be issued.
- */
-int xc_monitor_inguest_pagefault(xc_interface *xch, uint32_t domain_id,
-                                 bool disable);
 int xc_monitor_debug_exceptions(xc_interface *xch, uint32_t domain_id,
                                 bool enable, bool sync);
 int xc_monitor_cpuid(xc_interface *xch, uint32_t domain_id, bool enable);
diff --git a/tools/libxc/xc_monitor.c b/tools/libxc/xc_monitor.c
index 4ac823e775..0233b87b3f 100644
--- a/tools/libxc/xc_monitor.c
+++ b/tools/libxc/xc_monitor.c
@@ -163,20 +163,6 @@ int xc_monitor_guest_request(xc_interface *xch, uint32_t domain_id, bool enable,
     return do_domctl(xch, &domctl);
 }
 
-int xc_monitor_inguest_pagefault(xc_interface *xch, uint32_t domain_id,
-                                bool disable)
-{
-    DECLARE_DOMCTL;
-
-    domctl.cmd = XEN_DOMCTL_monitor_op;
-    domctl.domain = domain_id;
-    domctl.u.monitor_op.op = disable ? XEN_DOMCTL_MONITOR_OP_ENABLE
-                                    : XEN_DOMCTL_MONITOR_OP_DISABLE;
-    domctl.u.monitor_op.event = XEN_DOMCTL_MONITOR_EVENT_INGUEST_PAGEFAULT;
-
-    return do_domctl(xch, &domctl);
-}
-
 int xc_monitor_emulate_each_rep(xc_interface *xch, uint32_t domain_id,
                                 bool enable)
 {
diff --git a/tools/tests/xen-access/xen-access.c b/tools/tests/xen-access/xen-access.c
index 6aaee16d67..8c32bfbc3f 100644
--- a/tools/tests/xen-access/xen-access.c
+++ b/tools/tests/xen-access/xen-access.c
@@ -360,7 +360,7 @@ void usage(char* progname)
 {
     fprintf(stderr, "Usage: %s [-m] <domain_id> write|exec", progname);
 #if defined(__i386__) || defined(__x86_64__)
-            fprintf(stderr, "|breakpoint|altp2m_write|altp2m_exec|debug|cpuid|desc_access|write_ctrlreg_cr4|altp2m_write_no_gpt");
+            fprintf(stderr, "|breakpoint|altp2m_write|altp2m_exec|debug|cpuid|desc_access|write_ctrlreg_cr4");
 #elif defined(__arm__) || defined(__aarch64__)
             fprintf(stderr, "|privcall");
 #endif
@@ -393,7 +393,6 @@ int main(int argc, char *argv[])
     int cpuid = 0;
     int desc_access = 0;
     int write_ctrlreg_cr4 = 0;
-    int altp2m_write_no_gpt = 0;
     uint16_t altp2m_view_id = 0;
 
     char* progname = argv[0];
@@ -452,13 +451,6 @@ int main(int argc, char *argv[])
         altp2m = 1;
         memaccess = 1;
     }
-    else if ( !strcmp(argv[0], "altp2m_write_no_gpt") )
-    {
-        default_access = XENMEM_access_rw;
-        altp2m_write_no_gpt = 1;
-        memaccess = 1;
-        altp2m = 1;
-    }
     else if ( !strcmp(argv[0], "debug") )
     {
         debug = 1;
@@ -519,22 +511,6 @@ int main(int argc, char *argv[])
         xen_pfn_t gfn = 0;
         unsigned long perm_set = 0;
 
-        if( altp2m_write_no_gpt )
-        {
-            rc = xc_monitor_inguest_pagefault(xch, domain_id, 1);
-            if ( rc < 0 )
-            {
-                ERROR("Error %d setting inguest pagefault\n", rc);
-                goto exit;
-            }
-            rc = xc_monitor_emul_unimplemented(xch, domain_id, 1);
-            if ( rc < 0 )
-            {
-                ERROR("Error %d failed to enable emul unimplemented\n", rc);
-                goto exit;
-            }
-        }
-
         rc = xc_altp2m_set_domain_state( xch, domain_id, 1 );
         if ( rc < 0 )
         {
@@ -881,16 +857,6 @@ int main(int argc, char *argv[])
                        req.u.write_ctrlreg.old_value,
                        req.u.write_ctrlreg.new_value);
                 break;
-            case VM_EVENT_REASON_EMUL_UNIMPLEMENTED:
-                if ( altp2m_write_no_gpt && req.flags & VM_EVENT_FLAG_ALTERNATE_P2M )
-                {
-                    DPRINTF("\tSwitching back to default view!\n");
-
-                    rsp.flags |= (VM_EVENT_FLAG_ALTERNATE_P2M |
-                                  VM_EVENT_FLAG_TOGGLE_SINGLESTEP);
-                    rsp.altp2m_idx = 0;
-                }
-                break;
             default:
                 fprintf(stderr, "UNKNOWN REASON CODE %d\n", req.reason);
             }
diff --git a/xen/arch/x86/mm/mem_access.c b/xen/arch/x86/mm/mem_access.c
index 6346e8b35b..e63b7248ac 100644
--- a/xen/arch/x86/mm/mem_access.c
+++ b/xen/arch/x86/mm/mem_access.c
@@ -30,7 +30,6 @@
 #include <public/vm_event.h>
 #include <asm/p2m.h>
 #include <asm/altp2m.h>
-#include <asm/hvm/emulate.h>
 #include <asm/vm_event.h>
 #include <asm/hvm/hvm.h>
 
@@ -411,21 +410,6 @@ bool p2m_mem_access_check(paddr_t gpa, unsigned long gla,
         }
     }
 
-    /*
-     * Try to avoid sending a mem event. Suppress events caused by page-walks
-     * by emulating but still checking mem_access violations.
-     */
-    if ( vm_event_check_ring(d->vm_event_monitor) &&
-         d->arch.monitor.inguest_pagefault_disabled &&
-         npfec.kind == npfec_kind_in_gpt )
-    {
-        v->arch.vm_event->send_event = true;
-        hvm_emulate_one_vm_event(EMUL_KIND_NORMAL, TRAP_invalid_op, X86_EVENT_NO_EC);
-        v->arch.vm_event->send_event = false;
-
-        return true;
-    }
-
     if ( opt_introspection_extn &&
          vm_event_check_ring(d->vm_event_monitor) &&
          hvm_funcs.exited_by_nested_pagefault &&
diff --git a/xen/arch/x86/monitor.c b/xen/arch/x86/monitor.c
index 3c42e21906..3fb6531f6e 100644
--- a/xen/arch/x86/monitor.c
+++ b/xen/arch/x86/monitor.c
@@ -242,19 +242,6 @@ int arch_monitor_domctl_event(struct domain *d,
         break;
     }
 
-    case XEN_DOMCTL_MONITOR_EVENT_INGUEST_PAGEFAULT:
-    {
-        bool old_status = ad->monitor.inguest_pagefault_disabled;
-
-        if ( unlikely(old_status == requested_status) )
-            return -EEXIST;
-
-        domain_pause(d);
-        ad->monitor.inguest_pagefault_disabled = requested_status;
-        domain_unpause(d);
-        break;
-    }
-
     case XEN_DOMCTL_MONITOR_EVENT_DESC_ACCESS:
     {
         bool old_status = ad->monitor.descriptor_access_enabled;
diff --git a/xen/include/asm-x86/domain.h b/xen/include/asm-x86/domain.h
index e3d1c4f22a..a4fe2d2e2d 100644
--- a/xen/include/asm-x86/domain.h
+++ b/xen/include/asm-x86/domain.h
@@ -406,11 +406,6 @@ struct arch_domain
         unsigned int descriptor_access_enabled                             : 1;
         unsigned int guest_request_userspace_enabled                       : 1;
         unsigned int emul_unimplemented_enabled                            : 1;
-        /*
-         * By default all events are sent.
-         * This is used to filter out pagefaults.
-         */
-        unsigned int inguest_pagefault_disabled                            : 1;
         struct monitor_msr_bitmap *msr_bitmap;
         uint64_t write_ctrlreg_mask[4];
     } monitor;
diff --git a/xen/include/asm-x86/monitor.h b/xen/include/asm-x86/monitor.h
index 4afb0665e8..8e396f36a0 100644
--- a/xen/include/asm-x86/monitor.h
+++ b/xen/include/asm-x86/monitor.h
@@ -84,8 +84,7 @@ static inline uint32_t arch_monitor_get_capabilities(struct domain *d)
                     (1U << XEN_DOMCTL_MONITOR_EVENT_CPUID) |
                     (1U << XEN_DOMCTL_MONITOR_EVENT_DEBUG_EXCEPTION) |
                     (1U << XEN_DOMCTL_MONITOR_EVENT_WRITE_CTRLREG) |
-                    (1U << XEN_DOMCTL_MONITOR_EVENT_EMUL_UNIMPLEMENTED) |
-                    (1U << XEN_DOMCTL_MONITOR_EVENT_INGUEST_PAGEFAULT));
+                    (1U << XEN_DOMCTL_MONITOR_EVENT_EMUL_UNIMPLEMENTED));
 
     if ( hvm_is_singlestep_supported() )
         capabilities |= (1U << XEN_DOMCTL_MONITOR_EVENT_SINGLESTEP);
diff --git a/xen/include/public/domctl.h b/xen/include/public/domctl.h
index ee0ecb5f7c..af7799e7f9 100644
--- a/xen/include/public/domctl.h
+++ b/xen/include/public/domctl.h
@@ -1032,8 +1032,6 @@ struct xen_domctl_psr_cmt_op {
 #define XEN_DOMCTL_MONITOR_EVENT_INTERRUPT             8
 #define XEN_DOMCTL_MONITOR_EVENT_DESC_ACCESS           9
 #define XEN_DOMCTL_MONITOR_EVENT_EMUL_UNIMPLEMENTED    10
-/* Enabled by default */
-#define XEN_DOMCTL_MONITOR_EVENT_INGUEST_PAGEFAULT     11
 
 struct xen_domctl_monitor_op {
     uint32_t op; /* XEN_DOMCTL_MONITOR_OP_* */
