diff --git a/tools/libxc/xc_cpuid_x86.c b/tools/libxc/xc_cpuid_x86.c
index 810377c..1cfc634 100644
--- a/tools/libxc/xc_cpuid_x86.c
+++ b/tools/libxc/xc_cpuid_x86.c
@@ -308,14 +308,6 @@ static void amd_xc_cpuid_policy(xc_interface *xch,
             regs[0] = DEF_MAX_AMDEXT;
         break;
 
-    case 0x80000008:
-        /*
-         * ECX[15:12] is ApicIdCoreSize: ECX[7:0] is NumberOfCores (minus one).
-         * Update to reflect vLAPIC_ID = vCPU_ID * 2.
-         */
-        regs[2] = ((regs[2] & 0xf000u) + 1) | ((regs[2] & 0xffu) << 1) | 1u;
-        break;
-
     case 0x8000000a: {
         if ( !info->nestedhvm )
         {
@@ -355,12 +347,7 @@ static void intel_xc_cpuid_policy(xc_interface *xch,
     switch ( input[0] )
     {
     case 0x00000004:
-        /*
-         * EAX[31:26] is Maximum Cores Per Package (minus one).
-         * Update to reflect vLAPIC_ID = vCPU_ID * 2.
-         */
-        regs[0] = (((regs[0] & 0x7c000000u) << 1) | 0x04000000u |
-                   (regs[0] & 0x3ffu));
+        regs[0] &= 0x3ffu;
         regs[3] &= 0x3ffu;
         break;
 
@@ -372,11 +359,6 @@ static void intel_xc_cpuid_policy(xc_interface *xch,
     case 0x80000005:
         regs[0] = regs[1] = regs[2] = 0;
         break;
-
-    case 0x80000008:
-        /* Mask AMD Number of Cores information. */
-        regs[2] = 0;
-        break;
     }
 }
 
@@ -471,12 +453,6 @@ static void xc_cpuid_hvm_policy(xc_interface *xch,
         break;
 
     case 0x00000001:
-        /*
-         * EBX[23:16] is Maximum Logical Processors Per Package.
-         * Update to reflect vLAPIC_ID = vCPU_ID * 2.
-         */
-        regs[1] = (regs[1] & 0x0000ffffu) | ((regs[1] & 0x007f0000u) << 1);
-
         regs[2] = info->featureset[featureword_of(X86_FEATURE_XMM3)];
         regs[3] = info->featureset[featureword_of(X86_FEATURE_FPU)];
         break;
@@ -519,7 +495,7 @@ static void xc_cpuid_hvm_policy(xc_interface *xch,
 
     case 0x80000008:
         regs[0] &= 0x0000ffffu;
-        regs[1] = regs[3] = 0;
+        regs[1] = regs[2] = regs[3] = 0;
         break;
 
     case 0x00000002: /* Intel cache info (dumped by AMD policy) */
