diff --git a/xen/common/efi/boot.c b/xen/common/efi/boot.c
index c5725ca..a021d0b 100644
--- a/xen/common/efi/boot.c
+++ b/xen/common/efi/boot.c
@@ -1266,11 +1266,9 @@ void __init efi_init_memory(void)
         else if ( desc->Attribute & (EFI_MEMORY_UC | EFI_MEMORY_UCE) )
             prot |= _PAGE_PWT | _PAGE_PCD | MAP_SMALL_PAGES;
         else
-        {
+            /* Treat unknown cachability as writeback */
             printk(XENLOG_ERR "Unknown cachability for MFNs %#lx-%#lx\n",
                    smfn, emfn - 1);
-            continue;
-        }
 
         if ( desc->Attribute & EFI_MEMORY_WP )
             prot &= _PAGE_RW;
