diff --git a/tools/libxc/saverestore/restore_x86_hvm.c b/tools/libxc/saverestore/restore_x86_hvm.c
index c65f9ac..594dcfb 100644
--- a/tools/libxc/saverestore/restore_x86_hvm.c
+++ b/tools/libxc/saverestore/restore_x86_hvm.c
@@ -78,7 +78,11 @@ static int handle_hvm_params(struct xc_sr_context *ctx, struct xc_sr_record *rec
             break;
         }
 
-        rc = xc_set_hvm_param(xch, ctx->domid, entry->index, entry->value);
+        rc = xc_hvm_param_set(xch, ctx->domid, entry->index, entry->value);
+
+        IPRINTF("DBG: Set HVM param %02"PRId64" = 0x%016"PRIx64" (%d)",
+                entry->index, entry->value, rc);
+
         if ( rc < 0 )
         {
             PERROR("set HVM param %"PRId64" = 0x%016"PRIx64,
diff --git a/tools/xenguest/xenguest_stubs.c b/tools/xenguest/xenguest_stubs.c
index f9ac109..9b18cd2 100644
--- a/tools/xenguest/xenguest_stubs.c
+++ b/tools/xenguest/xenguest_stubs.c
@@ -657,6 +657,30 @@ int stub_xc_domain_restore(xc_interface *xch, xs_handle *xsh, int fd, int domid,
         r = set_genid(xch, xsh, domid);
         if (r)
             exit(1);
+
+        /*
+         * ### GROSS HACK ###
+         *
+         * Attempt to reinstate lost console pfn when migrating from versions
+         * of XenServer which turned access of HVM_PARAM_CONSOLE_{PFN,EVTCHN}
+         * into -EINVAL in a failed attempt to be more performant.
+         */
+        if ( console_evtchn && *console_mfn == 0 && *store_mfn != 0 )
+        {
+            *console_mfn = *store_mfn + 1;
+
+            info("Reinstating lost CONSOLE_PFN at %lx\n", *console_mfn);
+
+            r = xc_clear_domain_page(xch, domid, *console_mfn);
+            if ( r )
+                failwith_oss_xc(xch, "clear console page");
+            r = xc_hvm_param_set(xch, domid, HVM_PARAM_CONSOLE_PFN, *console_mfn);
+            if ( r )
+                failwith_oss_xc(xch, "set console pfn");
+            r = xc_hvm_param_set(xch, domid, HVM_PARAM_CONSOLE_EVTCHN, console_evtchn);
+            if ( r )
+                failwith_oss_xc(xch, "set console pfn");
+        }
     }
 
     return 0;
diff --git a/xen/arch/x86/hvm/hvm.c b/xen/arch/x86/hvm/hvm.c
index 89e6496..d0626c3 100644
--- a/xen/arch/x86/hvm/hvm.c
+++ b/xen/arch/x86/hvm/hvm.c
@@ -5790,6 +5790,16 @@ long do_hvm_op(unsigned long op, XEN_GUEST_HANDLE_PARAM(void) arg)
             rc = __copy_to_guest(arg, &a, 1) ? -EFAULT : 0;
         }
 
+        if ( (a.index == HVM_PARAM_STORE_PFN) ||
+             (a.index == HVM_PARAM_STORE_EVTCHN) ||
+             (a.index == HVM_PARAM_CONSOLE_PFN) ||
+             (a.index == HVM_PARAM_CONSOLE_EVTCHN) )
+        {
+            printk("DBG: %cet dom%u param %u = 0x%016"PRIx64"\n",
+                   op == HVMOP_set_param ? 's' : 'g',
+                   d->domain_id, a.index, a.value);
+        }
+
         HVM_DBG_LOG(DBG_LEVEL_HCALL, "%s param %u = %"PRIx64,
                     op == HVMOP_set_param ? "set" : "get",
                     a.index, a.value);
