diff --git a/xen/arch/x86/domain.c b/xen/arch/x86/domain.c
index 9369932..56619cc 100644
--- a/xen/arch/x86/domain.c
+++ b/xen/arch/x86/domain.c
@@ -73,6 +73,9 @@ void (*dead_idle) (void) __read_mostly = default_dead_idle;
 static void paravirt_ctxt_switch_from(struct vcpu *v);
 static void paravirt_ctxt_switch_to(struct vcpu *v);
 
+static bool_t __read_mostly opt_force_invariant_tsc;
+boolean_param("force_invariant_tsc", opt_force_invariant_tsc);
+
 static void default_idle(void)
 {
     local_irq_disable();
@@ -2250,7 +2253,8 @@ void domain_cpuid(
              * emulated, or the domain cannot migrate to other hosts.
              */
             if ( (input == 0x80000007) && /* Advanced Power Management */
-                 !d->disable_migrate && !d->arch.vtsc )
+                 !d->disable_migrate && !d->arch.vtsc &&
+                 !opt_force_invariant_tsc )
                 *edx &= ~cpufeat_mask(X86_FEATURE_ITSC);
 
             return;
