diff --git a/xen/arch/x86/hvm/hvm.c b/xen/arch/x86/hvm/hvm.c
index d6cc806..2fe8223 100644
--- a/xen/arch/x86/hvm/hvm.c
+++ b/xen/arch/x86/hvm/hvm.c
@@ -6122,6 +6122,10 @@ long do_hvm_op(unsigned long op, XEN_GUEST_HANDLE_PARAM(void) arg)
         gdprintk(XENLOG_INFO, "Legacy windows driver quirks enabled. (HVMOP_set_driver_version)\n");
         break;
 
+    case HVMOP_process_dying:
+        /* Reduced to functional NOP */
+        break;
+
     default:
     {
         gdprintk(XENLOG_DEBUG, "Bad HVM op %ld.\n", op);
diff --git a/xen/include/public/hvm/hvm_op.h b/xen/include/public/hvm/hvm_op.h
index 955a94e..962f467 100644
--- a/xen/include/public/hvm/hvm_op.h
+++ b/xen/include/public/hvm/hvm_op.h
@@ -369,6 +369,9 @@ DEFINE_XEN_GUEST_HANDLE(xen_hvm_set_ioreq_server_state_t);
 
 #endif /* defined(__XEN__) || defined(__XEN_TOOLS__) */
 
+/* Legacy NOP: Hint from PV drivers for process destruction. */
+#define HVMOP_process_dying        0x102
+
 #define HVMOP_set_driver_version 0x103
 struct xen_hvm_set_driver_version {
     uint32_t build;
