diff --git a/xen/arch/x86/hvm/hvm.c b/xen/arch/x86/hvm/hvm.c
index 521e4d9..52ca198 100644
--- a/xen/arch/x86/hvm/hvm.c
+++ b/xen/arch/x86/hvm/hvm.c
@@ -5741,6 +5741,16 @@ long do_hvm_op(unsigned long op, XEN_GUEST_HANDLE_PARAM(void) arg)
         {
             switch ( a.index )
             {
+            case HVM_PARAM_CONSOLE_PFN:
+            case HVM_PARAM_CONSOLE_EVTCHN:
+                if ( d->arch.hvm_domain.params[HVM_PARAM_CONSOLE_PFN] == 0 &&
+                     d->arch.hvm_domain.params[HVM_PARAM_CONSOLE_EVTCHN] == 0 )
+                {
+                    rc = -EINVAL;
+                    goto param_fail;
+                }
+                a.value = d->arch.hvm_domain.params[a.index];
+                break;
             case HVM_PARAM_ACPI_S_STATE:
                 a.value = d->arch.hvm_domain.is_s3_suspended ? 3 : 0;
                 break;
