diff --git a/xen/arch/x86/hvm/hvm.c b/xen/arch/x86/hvm/hvm.c
index 526d0b6..dd7ea13 100644
--- a/xen/arch/x86/hvm/hvm.c
+++ b/xen/arch/x86/hvm/hvm.c
@@ -5516,6 +5516,18 @@ long do_hvm_op(unsigned long op, XEN_GUEST_HANDLE_PARAM(void) arg)
             gdprintk(XENLOG_INFO, "Legacy windows driver quirks enabled. "
                      "(HVMOP_set_driver_version)\n");
             goto legacy_win_complete;
+
+        case HVMOP_process_dying:
+            /*
+             * Reduced to functional NOP, but must return success for certain
+             * versions of the legacy drivers.
+             *
+             * Xen-4.5 introduced the HVMOP_op_mask which truncates this op
+             * into HVMOP_set_pci_intx_level.  As HVMOP_set_pci_intx_level is
+             * ineligible for continuations, we can still distinguish legacy
+             * drivers.
+             */
+            goto legacy_win_complete;
         }
     }
 
diff --git a/xen/include/public/hvm/hvm_op.h b/xen/include/public/hvm/hvm_op.h
index a28f684..2bd5ba1 100644
--- a/xen/include/public/hvm/hvm_op.h
+++ b/xen/include/public/hvm/hvm_op.h
@@ -388,6 +388,9 @@ DEFINE_XEN_GUEST_HANDLE(xen_hvm_evtchn_upcall_vector_t);
 
 #endif /* defined(__i386__) || defined(__x86_64__) */
 
+/* Legacy NOP: Hint from PV drivers for process destruction. */
+#define HVMOP_process_dying        0x102
+
 #define HVMOP_set_driver_version 0x103
 struct xen_hvm_set_driver_version {
     uint32_t build;
