Allow 32-bit PV drivers to work on 64-bit Xen.

diff --git a/xen/arch/x86/hvm/hvm.c b/xen/arch/x86/hvm/hvm.c
index ab75b40..d6cc806 100644
--- a/xen/arch/x86/hvm/hvm.c
+++ b/xen/arch/x86/hvm/hvm.c
@@ -5533,6 +5533,21 @@ long do_hvm_op(unsigned long op, XEN_GUEST_HANDLE_PARAM(void) arg)
                 if ( a.value > HVMPTM_one_missed_tick_pending )
                     rc = -EINVAL;
                 break;
+            case HVM_PARAM_32BIT:
+                /* Not used any more, as the underlying bugs which caused this
+                 * param to be created have been fixed properly upstream.
+                 *
+                 * It served two purposes.  First to correctly latch the 32bit
+                 * shinfo, and to force a resychronisation of the domain
+                 * wallclock, as qemu has likely just stepped it on resume.
+                 *
+                 * "Legacy" windows drivers unconditionally make this
+                 * hypercall and ignore the return value.
+                 *
+                 * The Tampa and Clearwater non-legacy drivers still make this
+                 * hypercall, but will BSOD with 0x0000DEAD if the call fails.
+                 */
+                break;
             case HVM_PARAM_VIRIDIAN:
                 /* This should only ever be set once by the tools and read by the guest. */
                 rc = -EPERM;
diff --git a/xen/include/public/hvm/params.h b/xen/include/public/hvm/params.h
index 3c51072..403bfd0 100644
--- a/xen/include/public/hvm/params.h
+++ b/xen/include/public/hvm/params.h
@@ -56,6 +56,10 @@
 
 #if defined(__i386__) || defined(__x86_64__)
 
+/* The guest is using 32 bit PV drivers.  Apply suitable translations
+   to shared info and hypercall arguments. */
+#define HVM_PARAM_32BIT        8
+
 /*
  * Viridian enlightenments
  *
