diff --git a/docs/misc/xen-command-line.pandoc b/docs/misc/xen-command-line.pandoc
index 098173d..7ecc65b 100644
--- a/docs/misc/xen-command-line.pandoc
+++ b/docs/misc/xen-command-line.pandoc
@@ -1269,6 +1269,14 @@ wait descriptor timed out', try increasing this value.
 **WARNING: This command line option is deprecated, and superseded by
 _dom0-iommu=map-inclusive_ - using both options in combination is undefined.**
 
+### introspection_extn
+> '= <boolean>'
+
+> Default: 'false'
+
+This parameter enables/disables the use of extra introspection extensions in
+addition to the existing ones.
+
 ### irq_ratelimit (x86)
 > `= <integer>`
 
diff --git a/xen/arch/x86/hvm/hvm.c b/xen/arch/x86/hvm/hvm.c
index 633191c..42cc044 100644
--- a/xen/arch/x86/hvm/hvm.c
+++ b/xen/arch/x86/hvm/hvm.c
@@ -115,6 +115,9 @@ static const char __initconst warning_hvm_fep[] =
 static bool_t __initdata opt_altp2m_enabled = 0;
 boolean_param("altp2m", opt_altp2m_enabled);
 
+bool __read_mostly opt_introspection_extn = false;
+boolean_param("introspection_extn", opt_introspection_extn);
+
 static int cpu_callback(
     struct notifier_block *nfb, unsigned long action, void *hcpu)
 {
diff --git a/xen/include/asm-x86/hvm/hvm.h b/xen/include/asm-x86/hvm/hvm.h
index c4ab317..88cd391 100644
--- a/xen/include/asm-x86/hvm/hvm.h
+++ b/xen/include/asm-x86/hvm/hvm.h
@@ -40,6 +40,13 @@ extern bool opt_pod_enabled;
 #define opt_pod_enabled false
 #endif
 
+/* introspection extensions commandline argument value */
+#ifdef CONFIG_HVM
+extern bool opt_introspection_extn;
+#else
+#define opt_introspection_extn false
+#endif
+
 /* Interrupt acknowledgement sources. */
 enum hvm_intsrc {
     hvm_intsrc_none,
