diff --git a/tools/flask/policy/modules/xen.if b/tools/flask/policy/modules/xen.if
index 202ac0c..13f893c 100644
--- a/tools/flask/policy/modules/xen.if
+++ b/tools/flask/policy/modules/xen.if
@@ -84,7 +84,8 @@ define(`create_domain_build_label', `
 define(`manage_domain', `
 	allow $1 $2:domain { getdomaininfo getvcpuinfo getaffinity
 			getaddrsize pause unpause trigger shutdown destroy
-			setaffinity setdomainmaxmem getscheduler resume };
+			setaffinity setdomainmaxmem getscheduler resume
+			setpodtarget };
     allow $1 $2:domain2 set_vnumainfo;
 ')
 
