diff --git a/xen/arch/x86/cpu/vpmu_intel.c b/xen/arch/x86/cpu/vpmu_intel.c
index 80eecc7..4240dc1 100644
--- a/xen/arch/x86/cpu/vpmu_intel.c
+++ b/xen/arch/x86/cpu/vpmu_intel.c
@@ -726,7 +726,7 @@ static int core2_vpmu_do_rdmsr(unsigned int msr, uint64_t *msr_content)
         /* Extension for BTS */
         if ( vpmu_is_set(vpmu, VPMU_CPU_HAS_BTS) )
             *msr_content &= ~MSR_IA32_MISC_ENABLE_BTS_UNAVAIL;
-        *msr_content |= MSR_IA32_MISC_ENABLE_PEBS_UNAVAIL;
+        /* *msr_content |= MSR_IA32_MISC_ENABLE_PEBS_UNAVAIL; */
     }
 
     return 0;
