diff --git a/xen/common/domctl.c b/xen/common/domctl.c
index bbcc74de8d..3ef6b4dd52 100644
--- a/xen/common/domctl.c
+++ b/xen/common/domctl.c
@@ -1071,6 +1071,13 @@ long do_domctl(XEN_GUEST_HANDLE_PARAM(xen_domctl_t) u_domctl)
         copyback = 1;
         break;
 
+    case XEN_DOMCTL_set_privileged:
+        if ( opt_introspection_extn )
+            d->is_privileged = 1;
+        else
+            ret = -ENOSYS;
+        break;
+
     default:
         ret = arch_do_domctl(op, d, u_domctl);
         break;
diff --git a/xen/include/public/domctl.h b/xen/include/public/domctl.h
index 691f03e0a2..140d09610f 100644
--- a/xen/include/public/domctl.h
+++ b/xen/include/public/domctl.h
@@ -1250,6 +1250,7 @@ struct xen_domctl {
 #define XEN_DOMCTL_get_cpu_policy                82
 #define XEN_DOMCTL_set_cpu_policy                83
 #define XEN_DOMCTL_get_runstate_info             98
+#define XEN_DOMCTL_set_privileged                99
 #define XEN_DOMCTL_gdbsx_guestmemio            1000
 #define XEN_DOMCTL_gdbsx_pausevcpu             1001
 #define XEN_DOMCTL_gdbsx_unpausevcpu           1002
