diff --git a/xen/common/domctl.c b/xen/common/domctl.c
index 480a391..276864f 100644
--- a/xen/common/domctl.c
+++ b/xen/common/domctl.c
@@ -1211,6 +1211,13 @@ long do_domctl(XEN_GUEST_HANDLE_PARAM(xen_domctl_t) u_domctl)
         break;
     }
 
+    case XEN_DOMCTL_set_privileged:
+        if ( opt_introspection_extn )
+            d->is_privileged = 1;
+        else
+            ret = -ENOSYS;
+        break;
+
     default:
         ret = arch_do_domctl(op, d, u_domctl);
         break;
diff --git a/xen/include/public/domctl.h b/xen/include/public/domctl.h
index f7fc66d..ab8170c 100644
--- a/xen/include/public/domctl.h
+++ b/xen/include/public/domctl.h
@@ -1193,6 +1193,7 @@ struct xen_domctl {
 #define XEN_DOMCTL_psr_cat_op                    78
 #define XEN_DOMCTL_soft_reset                    79
 #define XEN_DOMCTL_get_runstate_info             98
+#define XEN_DOMCTL_set_privileged                99
 #define XEN_DOMCTL_gdbsx_guestmemio            1000
 #define XEN_DOMCTL_gdbsx_pausevcpu             1001
 #define XEN_DOMCTL_gdbsx_unpausevcpu           1002
