diff --git a/tools/xenguest/xenguest_stubs.c b/tools/xenguest/xenguest_stubs.c
index 66dcb43..4ba6384 100644
--- a/tools/xenguest/xenguest_stubs.c
+++ b/tools/xenguest/xenguest_stubs.c
@@ -1104,7 +1104,6 @@ int stub_xc_domain_save(int fd, int max_iters, int max_factors,
     migration_safety_checks();
 
     r = xc_domain_save(xch, fd, domid,
-                       max_iters, max_factors,
                        flags, &callbacks, hvm, 0, -1);
     if (r)
         failwith_oss_xc("xc_domain_save");
