diff --git a/tools/xenguest/xenguest_stubs.c b/tools/xenguest/xenguest_stubs.c
index b5695a2..5f45352 100644
--- a/tools/xenguest/xenguest_stubs.c
+++ b/tools/xenguest/xenguest_stubs.c
@@ -1110,7 +1110,6 @@ int emu_stub_xc_domain_save(int fd, int max_iters, int max_factors,
     migration_safety_checks();
 
     r = xc_domain_save(xch, fd, domid,
-                       max_iters, max_factors,
                        flags, &callbacks, hvm, 0, -1);
     if (r)
         failwith_oss_xc("xc_domain_save");
@@ -1133,7 +1132,6 @@ int stub_xc_domain_save(int fd, int max_iters, int max_factors,
     migration_safety_checks();
 
     r = xc_domain_save(xch, fd, domid,
-                       max_iters, max_factors,
                        flags, &callbacks, hvm, 0, -1);
     if (r)
         failwith_oss_xc("xc_domain_save");
