diff --git a/tools/xenguest/xenguest_stubs.c b/tools/xenguest/xenguest_stubs.c
index 5f45352..6db72a2 100644
--- a/tools/xenguest/xenguest_stubs.c
+++ b/tools/xenguest/xenguest_stubs.c
@@ -1105,6 +1105,7 @@ int emu_stub_xc_domain_save(int fd, int max_iters, int max_factors,
             .suspend = emu_suspend_callback,
             .switch_qemu_logdirty = switch_qemu_logdirty,
             .data = NULL,
+            .precopy_policy = NULL,
         };
 
     migration_safety_checks();
