xsplice: Make the replace command work

diff --git a/xen/common/xsplice.c b/xen/common/xsplice.c
index 7b9d050..b44239d 100644
--- a/xen/common/xsplice.c
+++ b/xen/common/xsplice.c
@@ -1169,9 +1169,9 @@ static int xsplice_do_wait(atomic_t *counter, s_time_t timeout,
 static void xsplice_do_single(unsigned int total_cpus)
 {
     nmi_callback_t saved_nmi_callback;
-    struct payload *data, *tmp;
+    struct payload *data, *tmp, *data2;
     s_time_t timeout;
-    int rc;
+    int rc = 0;
 
     data = xsplice_work.data;
     timeout = xsplice_work.timeout + NOW();
@@ -1207,11 +1207,11 @@ static void xsplice_do_single(unsigned int total_cpus)
             data->state = XSPLICE_STATE_CHECKED;
         break;
     case XSPLICE_ACTION_REPLACE:
-        list_for_each_entry_safe_reverse ( data, tmp, &applied_list, list )
+        list_for_each_entry_safe_reverse ( data2, tmp, &applied_list, applied_list )
         {
-            data->rc = revert_payload(data);
-            if ( data->rc == 0 )
-                data->state = XSPLICE_STATE_CHECKED;
+            data2->rc = revert_payload(data2);
+            if ( data2->rc == 0 )
+                data2->state = XSPLICE_STATE_CHECKED;
             else
             {
                 rc = -EINVAL;
